<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Define Genomic Regions Based on Gene Annotations — defineRegions • extraChIPs</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Define Genomic Regions Based on Gene Annotations — defineRegions"><meta property="og:description" content="Use gene, transcript and exon annotations to define genomic regions"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_signal_sliding.html">Differential Signal Using Sliding Windows</a>
    </li>
    <li>
      <a href="../articles/differential_signal_fixed.html">Differential Signal Using Fixed-Width Windows</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/smped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Define Genomic Regions Based on Gene Annotations</h1>
    <small class="dont-index">Source: <a href="https://github.com/smped/extraChIPs/blob/HEAD/R/defineRegions.R" class="external-link"><code>R/defineRegions.R</code></a></small>
    <div class="hidden name"><code>defineRegions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use gene, transcript and exon annotations to define genomic regions</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">defineRegions</span><span class="op">(</span></span>
<span>  <span class="va">genes</span>,</span>
<span>  <span class="va">transcripts</span>,</span>
<span>  <span class="va">exons</span>,</span>
<span>  promoter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2500</span>, <span class="fl">500</span><span class="op">)</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  intron <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  proximal <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>, <span class="st">"transcript_id"</span>, <span class="st">"transcript_name"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>genes, transcripts, exons</dt>
<dd><p>GRanges objects defining each level of
annotation</p></dd>


<dt>promoter</dt>
<dd><p>Numeric vector defining upstream and/or downstream distances
for a promoter. Passing a single value will define a symmetrical promoter
The first value represents the upstream range</p></dd>


<dt>upstream</dt>
<dd><p>The distance from a TSS defining an upstream promoter</p></dd>


<dt>intron</dt>
<dd><p>logical(1) Separate gene bodies into introns and exons. If
<code>intron = FALSE</code> gene bodies will simply be defined as gene bodies</p></dd>


<dt>proximal</dt>
<dd><p>Distance from a gene to be considered a proximal intergenic
region. If set to 0, intergenic regions will simply be considered as
uniformly intergenic</p></dd>


<dt>cols</dt>
<dd><p>Column names to be retained from the supplied annotations</p></dd>


<dt>...</dt>
<dd><p>Not used</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A GRangesList</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Using GRanges annotated as genes, transcripts and exons this function will
define ranges uniquely assigned to a region type using a hierarchical
process. By default, these region types will be (in order) 1) Promoters, 2)
Upstream Promoters, 3) Exons, 4) Introns, 5) Proximal Intergenic and 6)
Distal Intergenic.</p>
<p>Setting <code>intron = FALSE</code> will replace introns and exons with a generic "Gene
Body" annotation.
Setting <code>proximal = 0</code> will return all intergenic regions (not previously
annotated as promoters or upstream promoters) to an "Intergenic" category</p>
<p>Notably, once a region has been defined, it is excluded from all subsquent
candidate regions.</p>
<p>Any columns matching the names provided in cols will be returned, and it is
assumed that the gene/transcript/exon ranges will contain informative columns
in the <code>mcols()</code> element.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Define two exons for two transcripts</span></span></span>
<span class="r-in"><span><span class="va">sq</span> <span class="op">&lt;-</span> <span class="fu">Seqinfo</span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chr1"</span>, seqlengths <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:20001-21000"</span>, <span class="st">"chr1:29001-29950"</span>, <span class="st">"chr1:22001-23000"</span>, <span class="st">"chr1:29001-30000"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="va">e</span>, seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene_id <span class="op">=</span> <span class="st">"Gene1"</span>,</span></span>
<span class="r-in"><span>  transcript_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Trans1"</span>, <span class="st">"Trans1"</span>, <span class="st">"Trans2"</span>, <span class="st">"Trans2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Define the transcript ranges</span></span></span>
<span class="r-in"><span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu">endoapply</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">e</span>, <span class="va">e</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span>, <span class="va">range</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">t</span><span class="op">$</span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="st">"Gene1"</span></span></span>
<span class="r-in"><span><span class="va">t</span><span class="op">$</span><span class="va">transcript_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Summarise to gene level</span></span></span>
<span class="r-in"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="reduceMC.html">reduceMC</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">g</span><span class="op">$</span><span class="va">transcript_id</span> <span class="op">&lt;-</span> <span class="cn">NA_character_</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Now annotate the regions</span></span></span>
<span class="r-in"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">defineRegions</span><span class="op">(</span>genes <span class="op">=</span> <span class="va">g</span>, transcripts <span class="op">=</span> <span class="va">t</span>, exons <span class="op">=</span> <span class="va">e</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 9 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       seqnames      ranges strand |                 region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     distal_intergenic     chr1     1-10000      * |     Intergenic (&gt;10kb)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   proximal_intergenic     chr1 10001-15000      * |     Intergenic (&lt;10kb)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     upstream_promoter     chr1 15001-17500      * | Upstream Promoter (-..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              promoter     chr1 17501-22500      * |  Promoter (-2500/+500)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  exon     chr1 22501-23000      * |                   Exon</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                intron     chr1 23001-29000      * |                 Intron</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  exon     chr1 29001-30000      * |                   Exon</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   proximal_intergenic     chr1 30001-40000      * |     Intergenic (&lt;10kb)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     distal_intergenic     chr1 40001-50000      * |     Intergenic (&gt;10kb)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           gene_id   transcript_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       &lt;character&gt; &lt;CharacterList&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     distal_intergenic        &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   proximal_intergenic       Gene1            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     upstream_promoter       Gene1   Trans1,Trans2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              promoter       Gene1   Trans1,Trans2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  exon       Gene1          Trans2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                intron       Gene1            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  exon       Gene1   Trans1,Trans2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   proximal_intergenic       Gene1            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     distal_intergenic        &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Alternatively, collpse gene body and intergenic ranges</span></span></span>
<span class="r-in"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">defineRegions</span><span class="op">(</span></span></span>
<span class="r-in"><span>  genes <span class="op">=</span> <span class="va">g</span>, transcripts <span class="op">=</span> <span class="va">t</span>, exons <span class="op">=</span> <span class="va">e</span>, intron <span class="op">=</span> <span class="cn">FALSE</span>, proximal <span class="op">=</span> <span class="fl">0</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 5 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     seqnames      ranges strand |                 region</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                        &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          intergenic     chr1     1-15000      * |             Intergenic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   upstream_promoter     chr1 15001-17500      * | Upstream Promoter (-..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            promoter     chr1 17501-22500      * |  Promoter (-2500/+500)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           gene_body     chr1 22501-30000      * |              Gene Body</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          intergenic     chr1 30001-50000      * |             Intergenic</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         gene_id   transcript_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     &lt;character&gt; &lt;CharacterList&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          intergenic        &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   upstream_promoter       Gene1   Trans1,Trans2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            promoter       Gene1   Trans1,Trans2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           gene_body       Gene1            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          intergenic        &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stephen Pederson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

