<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>extraChIPs: A package for enabling and extending ChIP-Seq analysis — extraChIPs-package • extraChIPs</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: A package for enabling and extending ChIP-Seq analysis — extraChIPs-package"><meta property="og:description" content="The package provides three categories of important functions:
Range-based, Visualisation and Convenience functions, with most centred
around GenomicRanges objects"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_binding.html">Differential Binding</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/steveped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>extraChIPs: A package for enabling and extending ChIP-Seq analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/steveped/extraChIPs/blob/HEAD/R/extraChIPs.R" class="external-link"><code>R/extraChIPs.R</code></a></small>
    <div class="hidden name"><code>extraChIPs-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The package provides three categories of important functions:
Range-based, Visualisation and Convenience functions, with most centred
around GenomicRanges objects</p>
    </div>


    <div id="range-based-functions">
    <h2>Range-based Functions</h2>
    


<p>Many of the range-based functions included in this package have a focus on
retaining the <code>mcols</code> information whilst manipulating the ranges, such as
<code><a href="reduceMC.html">reduceMC()</a></code> which not only reduces the Ranges, but collapses the <code>mcols</code>
into vectors or <a href="https://rdrr.io/pkg/IRanges/man/CompressedList-class.html" class="external-link">IRanges::CompressedList</a> objects.
Key function from this group are:</p><ul><li><p><code><a href="reduceMC.html">reduceMC()</a></code>, <code><a href="setoptsMC-methods.html">setdiffMC()</a></code>, <code><a href="setoptsMC-methods.html">intersectMC()</a></code>, <code><a href="setoptsMC-methods.html">unionMC()</a></code>, <code><a href="distinctMC.html">distinctMC()</a></code>
and <code><a href="chopMC.html">chopMC()</a></code></p></li>
<li><p><code><a href="bestOverlap-methods.html">bestOverlap()</a></code> and <code><a href="propOverlap-methods.html">propOverlap()</a></code> provide simple output easily able to
be added as a column within the <code>mcols</code> element</p></li>
<li><p><code><a href="as_tibble.html">as_tibble()</a></code> coerces a GRanges object to a <a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble::tibble</a>.</p></li>
<li><p><code><a href="colToRanges-methods.html">colToRanges()</a></code> enables parsing of a single column to a GRanges object,
setting all other columns as the <code>mcols</code> element.</p></li>
<li><p><code><a href="stitchRanges.html">stitchRanges()</a></code> merges nearby ranges setting barrier ranges which cannot
be crossed when merging</p></li>
<li><p><code><a href="partitionRanges-methods.html">partitionRanges()</a></code> break apart one set of ranges by another</p></li>
<li><p><code><a href="dualFilter.html">dualFilter()</a></code> filters ranges from sliding windows using a guide set of
reference ranges where signal is confidently expected</p></li>
<li><p><code><a href="mergeByCol-methods.html">mergeByCol()</a></code> merges overlapping ranges, as produced by sliding windows</p></li>
<li><p><code><a href="mapByFeature.html">mapByFeature()</a></code> is able to map a set of GRanges to the most appropriate
gene, using any optional combination of promoters, enhancers and HiC
interactions</p></li>
<li><p><code><a href="grlToSE-methods.html">grlToSE()</a></code> takes selected columns from a GRangesList and sets them as
assays within a <a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment::RangedSummarizedExperiment</a> object.
Used for combining peak intensities or results across multiple ChIP targets.</p></li>
</ul></div>
    <div id="visualisation-functions">
    <h2>Visualisation Functions</h2>
    

<ul><li><p><code><a href="plotHFGC.html">plotHFGC()</a></code> is a wrapper to Gviz plotting functions, able to take any
combination of HiC, Features, Genes and Coverage (i.e. BigWig) and plot a
specified range.</p></li>
<li><p><code><a href="plotProfileHeatmap-methods.html">plotProfileHeatmap()</a></code> plots the average signal around a set of ranges,
as prepared by <code><a href="getProfileData-methods.html">getProfileData()</a></code></p></li>
<li><p><code><a href="plotPie-methods.html">plotPie()</a></code> emables simple comparison across multuple annotation columns
within a GRanges object.</p></li>
<li><p><code><a href="plotAssayDensities-methods.html">plotAssayDensities()</a></code>, <code><a href="plotAssayPCA-methods.html">plotAssayPCA()</a></code> and <code><a href="plotAssayRle-methods.html">plotAssayRle()</a></code> provide
simple interfaces to plotting key values from a
<a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment::RangedSummarizedExperiment</a>.</p></li>
</ul></div>
    <div id="convenience-functions">
    <h2>Convenience Functions</h2>
    

<ul><li><p><code><a href="collapseGenes.html">collapseGenes()</a></code> prints a vector of genes for an rmarkdown document,
using italics.</p></li>
<li><p><code><a href="importPeaks.html">importPeaks()</a></code> imports large numbers of broadPeak or narrowPeak files</p></li>
<li><p><code><a href="voomWeightsFromCPM.html">voomWeightsFromCPM()</a></code> allows creation of an limma::EList object
as would be created from counts by <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code>, but using
<code><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">edgeR::cpm()</a></code> values as input.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Stephen Pederson</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/steveped" class="external-link">Stephen
Pederson</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

