<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ extraChIPs</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_signal_sliding.html">Differential Signal Using Sliding Windows</a>
    </li>
    <li>
      <a href="../articles/differential_signal_fixed.html">Differential Signal Using Fixed-Width Windows</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/smped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="overview">Overview <a href="#overview" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="extraChIPs-package.html">extraChIPs-package</a></code> </p>
        </td>
        <td><p>extraChIPs: A package for enabling and extending ChIP-Seq analysis</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="range-based-operations">Range-Based Operations <a href="#range-based-operations" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="makeConsensus.html">makeConsensus()</a></code> </p>
        </td>
        <td><p>Make a set of consensus peaks</p></td>
      </tr><tr><td>
          <p><code><a href="mapGrlCols.html">mapGrlCols()</a></code> </p>
        </td>
        <td><p>Collapse a GRangesList adding multiple columns from each element</p></td>
      </tr><tr><td>
          <p><code><a href="chopMC.html">chopMC()</a></code> </p>
        </td>
        <td><p>Keep unique ranges and collapse mcols</p></td>
      </tr><tr><td>
          <p><code><a href="distinctMC.html">distinctMC()</a></code> </p>
        </td>
        <td><p>Keep distinct ranges and mcols</p></td>
      </tr><tr><td>
          <p><code><a href="reduceMC.html">reduceMC()</a></code> </p>
        </td>
        <td><p>Reduce ranges retaining mcols</p></td>
      </tr><tr><td>
          <p><code><a href="setoptsMC-methods.html">setdiffMC()</a></code> <code><a href="setoptsMC-methods.html">intersectMC()</a></code> <code><a href="setoptsMC-methods.html">unionMC()</a></code> </p>
        </td>
        <td><p>Perform set operations retaining mcols</p></td>
      </tr><tr><td>
          <p><code><a href="bestOverlap-methods.html">bestOverlap()</a></code> </p>
        </td>
        <td><p>Find the best overlap between GRanges</p></td>
      </tr><tr><td>
          <p><code><a href="propOverlap-methods.html">propOverlap()</a></code> </p>
        </td>
        <td><p>Find the proportions of an overlapping range</p></td>
      </tr><tr><td>
          <p><code><a href="colToRanges-methods.html">colToRanges()</a></code> </p>
        </td>
        <td><p>Coerce a column to a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="defineRegions.html">defineRegions()</a></code> </p>
        </td>
        <td><p>Define Genomic Regions Based on Gene Annotations</p></td>
      </tr><tr><td>
          <p><code><a href="partitionRanges-methods.html">partitionRanges()</a></code> </p>
        </td>
        <td><p>Partition a set of Genomic Ranges</p></td>
      </tr><tr><td>
          <p><code><a href="stitchRanges.html">stitchRanges()</a></code> </p>
        </td>
        <td><p>Stitch Ranges within a given distance</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="functions-for-differential-signal-analysis">Functions For Differential Signal Analysis <a href="#functions-for-differential-signal-analysis" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="dualFilter.html">dualFilter()</a></code> </p>
        </td>
        <td><p>Apply two filters to sliding windows</p></td>
      </tr><tr><td>
          <p><code><a href="importPeaks.html">importPeaks()</a></code> </p>
        </td>
        <td><p>Import peaks</p></td>
      </tr><tr><td>
          <p><code><a href="fitAssayDiff-methods.html">fitAssayDiff()</a></code> </p>
        </td>
        <td><p>Detect Differential ChIP Signal</p></td>
      </tr><tr><td>
          <p><code><a href="getProfileData-methods.html">getProfileData()</a></code> </p>
        </td>
        <td><p>Get Profile Data surrounding specified ranges</p></td>
      </tr><tr><td>
          <p><code><a href="grlToSE-methods.html">grlToSE()</a></code> </p>
        </td>
        <td><p>Set columns from a GRangesList as Assays in a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="mapByFeature.html">mapByFeature()</a></code> </p>
        </td>
        <td><p>Map Genomic Ranges to genes using defined features</p></td>
      </tr><tr><td>
          <p><code><a href="mergeByCol-methods.html">mergeByCol()</a></code> </p>
        </td>
        <td><p>Merge sliding windows using a specified column</p></td>
      </tr><tr><td>
          <p><code><a href="mergeByHMP-methods.html">mergeByHMP()</a></code> </p>
        </td>
        <td><p>Merge Sliding Windows using the Harmonic Mean P</p></td>
      </tr><tr><td>
          <p><code><a href="mergeBySig-methods.html">mergeBySig()</a></code> </p>
        </td>
        <td><p>Merge overlapping ranges based on p-values</p></td>
      </tr><tr><td>
          <p><code><a href="voomWeightsFromCPM.html">voomWeightsFromCPM()</a></code> </p>
        </td>
        <td><p>Estimate voom precision weights directly From CPM values</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="visualisation">Visualisation <a href="#visualisation" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plotAssayDensities-methods.html">plotAssayDensities()</a></code> </p>
        </td>
        <td><p>Plot Densities for any assay within a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="plotAssayHeatmap-methods.html">plotAssayHeatmap()</a></code> </p>
        </td>
        <td><p>Draw a heatmap from a single SummarizedExperiment assay</p></td>
      </tr><tr><td>
          <p><code><a href="plotAssayPCA-methods.html">plotAssayPCA()</a></code> </p>
        </td>
        <td><p>Plot PCA For any assay within a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="plotAssayRle-methods.html">plotAssayRle()</a></code> </p>
        </td>
        <td><p>Plot RLE for a given assay within a SummarizedExperiment</p></td>
      </tr><tr><td>
          <p><code><a href="plotHFGC.html">plotHFGC()</a></code> </p>
        </td>
        <td><p>Plot a Genomic Region showing HiC, Features, Genes and Coverage</p></td>
      </tr><tr><td>
          <p><code><a href="plotOverlaps-methods.html">plotOverlaps()</a></code> </p>
        </td>
        <td><p>Plot Overlaps Between List Elements</p></td>
      </tr><tr><td>
          <p><code><a href="plotPairwise.html">plotPairwise()</a></code> </p>
        </td>
        <td><p>Plot Pairwise Values from a GRangeList</p></td>
      </tr><tr><td>
          <p><code><a href="plotPie-methods.html">plotPie()</a></code> </p>
        </td>
        <td><p>Draw Pie Graphs based on one or more columns</p></td>
      </tr><tr><td>
          <p><code><a href="plotProfileHeatmap-methods.html">plotProfileHeatmap()</a></code> </p>
        </td>
        <td><p>Draw a coverage Profile Heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotSplitDonut-methods.html">plotSplitDonut()</a></code> </p>
        </td>
        <td><p>Draw Two-Level Donut Charts</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="coercion-convenience-functions">Coercion &amp; Convenience Functions <a href="#coercion-convenience-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="as_tibble.html">as_tibble(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="as_tibble.html">as_tibble(<i>&lt;GenomicRanges&gt;</i>)</a></code> <code><a href="as_tibble.html">as_tibble(<i>&lt;Seqinfo&gt;</i>)</a></code> <code><a href="as_tibble.html">as_tibble(<i>&lt;GInteractions&gt;</i>)</a></code> <code><a href="as_tibble.html">as_tibble(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="as_tibble.html">as_tibble(<i>&lt;TopTags&gt;</i>)</a></code> </p>
        </td>
        <td><p>Convert to a tibble</p></td>
      </tr><tr><td>
          <p><code><a href="collapseGenes.html">collapseGenes()</a></code> </p>
        </td>
        <td><p>Collapse a vector of gene names</p></td>
      </tr><tr><td>
          <p><code><a href="addDiffStatus-methods.html">addDiffStatus()</a></code> </p>
        </td>
        <td><p>Add a status column</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data-objects">Data Objects <a href="#data-objects" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cytobands.html">grch37.cytobands</a></code> <code><a href="cytobands.html">grch38.cytobands</a></code> </p>
        </td>
        <td><p>Cytogenetic bands</p></td>
      </tr><tr><td>
          <p><code><a href="ex_datasets.html">ex_trans</a></code> <code><a href="ex_datasets.html">ex_genes</a></code> <code><a href="ex_datasets.html">ex_prom</a></code> <code><a href="ex_datasets.html">ex_hic</a></code> </p>
        </td>
        <td><p>Datasets for an example region</p></td>
      </tr><tr><td>
          <p><code><a href="fixed_width_datasets.html">se</a></code> <code><a href="fixed_width_datasets.html">peaks</a></code> </p>
        </td>
        <td><p>Datasets for the Fixed-Width Vignette</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Stephen Pederson.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

