<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Differential Binding Analysis • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Differential Binding Analysis">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_binding.html">Differential Binding</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/steveped/extraChIPs/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Differential Binding Analysis</h1>
                        <h4 data-toc-skip class="author">Steve Pederson</h4>
            <address class="author_afil">
      Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, University of Adelaide<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/steveped/extraChIPs/blob/HEAD/vignettes/differential_binding.Rmd" class="external-link"><code>vignettes/differential_binding.Rmd</code></a></small>
      <div class="hidden name"><code>differential_binding.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="differential-binding-and-chip-seq-analysis">Differential Binding and ChIP-Seq Analysis<a class="anchor" aria-label="anchor" href="#differential-binding-and-chip-seq-analysis"></a>
</h2>
<p>The <a href="https://github.com/steveped/GRAVI" class="external-link">GRAVI</a> workflow, for which this package is designed, uses sliding windows for differential binding analysis in a manner similar to the package <code>csaw</code>, but also incorporating <code>macs2</code> peaks. The workflow itself extends to integrating multiple IP targets and external data sources, and as such, this package introduces a handful of functions to enable these analyses.</p>
<p>The majority of examples below use extremely simplified datasets to provide general guidance on using the functions. Some results may appear trivial as a result, but will hopefully prove far more useful in a true experimental context.</p>
<div class="section level3">
<h3 id="sliding-windows">Sliding Windows<a class="anchor" aria-label="anchor" href="#sliding-windows"></a>
</h3>
<p>The starting point for differential binding analysis using sliding windows is to define windows, then count reads within each window using the bam files. Commonly one or IP input/control samples is also produced during a ChIP-Seq experiment. The example files provided here contain a small subset of reads from chromosome 10 across two experimental and one input sample.</p>
<p>The approach taken below is to define a set of sliding windows, using the capabilities of <code>csaw</code>, but to then use <code>macs</code> peaks to define regions of most likely signal. First we can define our windows and count the alignments using existing tools. In the following, we’ll use a sliding window of 180bp and a step size of 60bp, meaning each nucleotide is covered by 3 windows.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csaw</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata"</span>, <span class="st">"bam"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1.bam"</span>, <span class="st">"ex2.bam"</span>, <span class="st">"input.bam"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"extraChIPs"</span>
<span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1"</span>, <span class="st">"ex2"</span>, <span class="st">"input"</span><span class="op">)</span>
<span class="va">rp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/readParam.html" class="external-link">readParam</a></span><span class="op">(</span>
  pe <span class="op">=</span> <span class="st">"none"</span>,
  dedup <span class="op">=</span> <span class="cn">TRUE</span>,
  restrict <span class="op">=</span> <span class="st">"chr10"</span>
<span class="op">)</span>
<span class="va">wincounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/windowCounts.html" class="external-link">windowCounts</a></span><span class="op">(</span>
  bam.files <span class="op">=</span> <span class="va">bfl</span>,
  spacing <span class="op">=</span> <span class="fl">60</span>,
  width <span class="op">=</span> <span class="fl">180</span>,
  ext <span class="op">=</span> <span class="fl">200</span>,
  filter <span class="op">=</span> <span class="fl">1</span>,
  param <span class="op">=</span> <span class="va">rp</span>,
  BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">bpparam</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>This produces a <code>RangesSummarizedExperiment</code> with windows included which passed the minimum threshold of 1 total read. As we’ve only counted reads within a very small window, the complete library sizes will be highly inaccurate. The true library sizes can be added here noting that <em>this step is not normally</em> <em>required</em>, but given these values are essential for accurate CPM values, they will be added here.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wincounts</span><span class="op">$</span><span class="va">totals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">964076L</span>, <span class="fl">989543L</span>, <span class="fl">1172179L</span><span class="op">)</span></code></pre></div>
<p>We can also add some key information to the <code>colData</code> element of this object, which will also be propagated to all downstream objects.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wincounts</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span>
<span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"treat"</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 3 rows and 6 columns</span>
<span class="co">##                    bam.files    totals       ext      rlen      sample    treat</span>
<span class="co">##                  &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;factor&gt;</span>
<span class="co">## ex1   /tmp/Rtmps0u37i/temp..    964076       200        73         ex1    ctrl </span>
<span class="co">## ex2   /tmp/Rtmps0u37i/temp..    989543       200        74         ex2    treat</span>
<span class="co">## input /tmp/Rtmps0u37i/temp..   1172179       200        74       input    NA</span></code></pre>
<p>A density plot can be simply drawn of these counts, with the vast majority of windows receiving very low counts, due to the nature of transcription factor binding, where long stretches are unbound. The windows with higher counts tend to be associated with the samples targeting a transcription factor (TF), as seen in the two treatment group samples.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/steveped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">wincounts</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, trans <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-densities-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="filtering-of-sliding-windows">Filtering of Sliding Windows<a class="anchor" aria-label="anchor" href="#filtering-of-sliding-windows"></a>
</h3>
<p>After counting all reads in the sliding windows, the next step is to discard windows for which counts are unlikely to represent TF binding. The package <code>extraChIPs</code> uses a set of consensus peaks to automatically set a threshold based on 1) counts strongly above the counts from the input sample, and 2) the windows with the overall highest signal. Thresholds are determined such that <code>q = 0.5</code> of the retained windows overlap on of the supplied consensus peaks. Higher values for <code>q</code> will return more windows, however many of these will tend to only marginally overlap a peak. Experience has shown that values such as <code>q = 0.5</code> tend to return a considerable proportion of windows containing true TF binding signal.</p>
<p>First we can load the peaks, supplied here as a simple <code>bed</code> file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"peaks.bed.gz"</span>, package <span class="op">=</span> <span class="st">"extraChIPs"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></code></pre></div>
<p>The we can pass these to the function <code><a href="../reference/dualFilter-methods.html">dualFilter()</a></code> which utilises the strategy described above. On large datasets, this can be quite time-consuming, as can the initial counting step. Due to the small example dataset, a more inclusive threshold for <code>q</code> will be used here.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filtcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dualFilter-methods.html">dualFilter</a></span><span class="op">(</span>
  x <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span><span class="op">)</span><span class="op">]</span>,
  bg <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span><span class="op">)</span><span class="op">]</span>, 
  ref <span class="op">=</span> <span class="va">peaks</span>,
  q <span class="op">=</span> <span class="fl">0.8</span> <span class="co"># Better to use q = 0.5 on real data</span>
<span class="op">)</span></code></pre></div>
<p>The returned object will by default contain <code>counts</code> and <code>logCPM</code> assays, with the complete library sizes used for the calculation of <code>logCPM</code> values.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plotcpm-1.png" width="960"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotAssayPCA-methods.html">plotAssayPCA</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, label <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plotcpm-2.png" width="960"></p>
<p>Whilst the initial set of counts contained 595 windows, these have now been reduced to 97 windows. Similarly, the input sample is no longer included in the data object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 595   3</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 97  2</span></code></pre>
<p>The <code>rowData</code> element of the returned object will contain a logical column indicating where each specific retained window overlapped one of the supplied consensus peaks.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 97 ranges and 1 metadata column:</span>
<span class="co">##        seqnames              ranges strand | overlaps_ref</span>
<span class="co">##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;logical&gt;</span>
<span class="co">##    [1]    chr10 103865521-103865700      * |         TRUE</span>
<span class="co">##    [2]    chr10 103865581-103865760      * |         TRUE</span>
<span class="co">##    [3]    chr10 103865641-103865820      * |         TRUE</span>
<span class="co">##    [4]    chr10 103865701-103865880      * |         TRUE</span>
<span class="co">##    [5]    chr10 103874161-103874340      * |        FALSE</span>
<span class="co">##    ...      ...                 ...    ... .          ...</span>
<span class="co">##   [93]    chr10 103900141-103900320      * |         TRUE</span>
<span class="co">##   [94]    chr10 103900201-103900380      * |         TRUE</span>
<span class="co">##   [95]    chr10 103900261-103900440      * |         TRUE</span>
<span class="co">##   [96]    chr10 103900321-103900500      * |         TRUE</span>
<span class="co">##   [97]    chr10 103900381-103900560      * |         TRUE</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">overlaps_ref</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 0.8247423</span></code></pre>
</div>
<div class="section level3">
<h3 id="using-voom">Using Voom<a class="anchor" aria-label="anchor" href="#using-voom"></a>
</h3>
<p>Multiple approaches are available for analysis of differential binding, and given the small example dataset, only a brief example of conventional results will be used. <code>extraChIPs</code> does provide a simple coercion function to covert <code>logCPM</code> to a <code>voom</code> object, which requires the relationship between library sizes and <code>logCPM</code> values to be intact. Whist this will not be discussed further here should this be a viable approach for an analysis, the following code may prove helpful.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voomWeightsFromCPM.html">voomWeightsFromCPM</a></span><span class="op">(</span>
  cpm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>, <span class="st">"logCPM"</span><span class="op">)</span>, 
  lib.size <span class="op">=</span> <span class="va">filtcounts</span><span class="op">$</span><span class="va">totals</span>, 
  isLogCPM <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-windows">Merging Windows<a class="anchor" aria-label="anchor" href="#merging-windows"></a>
</h3>
<p>After an analysis has been performed, common values contained in the output may be estimated signal (<code>logCPM</code>), estimated change (<code>logFC</code>) with both raw and adjusted p-values. Given the dependency of neighbouring windows, any adjusted p-values will not be appropriate and a merging of overlapping windows will be performed.</p>
<p>For our example dataset we’ll add these manually, however this is just for demonstration purposes for the eventual merging of windows.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">logCPM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>,<span class="st">"logCPM"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;-</span> <span class="fu">rowDiffs</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>,<span class="st">"logCPM"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">PValue</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Chisquare.html" class="external-link">pchisq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">logFC</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Now we have some example values, we can merge any overlapping windows using <code><a href="../reference/mergeByCol-methods.html">mergeByCol()</a></code>. During this process, overlapping ranges are merged into a single range with representative values taken from one of te initial sliding windows. The recommended approach for retaining statistical independence between windows is to choose the window with the largest signal as representative of the entire merged window.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeByCol-methods.html">mergeByCol</a></span><span class="op">(</span><span class="va">filtcounts</span>, col <span class="op">=</span> <span class="st">"logCPM"</span>, pval <span class="op">=</span> <span class="st">"PValue"</span><span class="op">)</span>
<span class="va">res_gr</span><span class="op">$</span><span class="va">overlaps_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">overlapsAny</a></span><span class="op">(</span><span class="va">res_gr</span>, <span class="va">peaks</span><span class="op">)</span></code></pre></div>
<p>A <code>GRanges</code> object is returned with representative values for each merged window. The <code>mcol</code> <code>keyval_range</code> provides the original range from which the representative values were taken. A column with adjusted p-values will also be added if <code>p_adj_method</code> is not set to “none”.</p>
</div>
<div class="section level3">
<h3 id="mapping-of-windows-to-genes">Mapping of Windows To Genes<a class="anchor" aria-label="anchor" href="#mapping-of-windows-to-genes"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="visualisation-of-results">Visualisation of Results<a class="anchor" aria-label="anchor" href="#visualisation-of-results"></a>
</h2>
<div class="section level3">
<h3 id="profile-heatmaps">Profile Heatmaps<a class="anchor" aria-label="anchor" href="#profile-heatmaps"></a>
</h3>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata"</span>, <span class="st">"bigwig"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1.bw"</span>, <span class="st">"ex2.bw"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"extraChIPs"</span>
<span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFileList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1"</span>, <span class="st">"ex2"</span><span class="op">)</span><span class="op">)</span>
<span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getProfileData-methods.html">getProfileData</a></span><span class="op">(</span><span class="va">bwfl</span>, <span class="va">res_gr</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap</a></span><span class="op">(</span>
  <span class="va">pd</span>, <span class="st">"profile_data"</span>, 
  facetY <span class="op">=</span> <span class="st">"overlaps_ref"</span>,
  labeller <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/as_labeller.html" class="external-link">as_labeller</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
      <span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"Consensus Peak"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"No Consensus Peak"</span>,
      <span class="st">"ex1"</span> <span class="op">=</span> <span class="st">"Control"</span>, <span class="st">"ex2"</span> <span class="op">=</span> <span class="st">"Treat"</span>
    <span class="op">)</span>
  <span class="op">)</span>
<span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"Consensus\nPeak"</span>, fill <span class="op">=</span> <span class="st">"CPM"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-profile-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="inspection-of-ranges">Inspection of Ranges<a class="anchor" aria-label="anchor" href="#inspection-of-ranges"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"grch37.cytobands"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_genes"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  gr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html" class="external-link">range</a></span><span class="op">(</span><span class="va">res_gr</span><span class="op">)</span>,
  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span>macs2 <span class="op">=</span> <span class="va">peaks</span><span class="op">)</span>,
  featcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>macs2 <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>,
  genes <span class="op">=</span> <span class="va">ex_genes</span>,
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Up <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span>,
  coverage <span class="op">=</span> <span class="va">bwfl</span>,
  linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ex1 <span class="op">=</span> <span class="st">"black"</span>, ex2 <span class="op">=</span> <span class="st">"forestgreen"</span><span class="op">)</span>,
  annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span>window <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">res_gr</span><span class="op">)</span><span class="op">)</span>,
  annotcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>window <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>,
  title.width <span class="op">=</span> <span class="fl">1.5</span>, rotation.title <span class="op">=</span> <span class="fl">90</span>,
  collapseTranscripts <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-hfgc-1.png" width="960"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/steveped" class="external-link">Stephen Pederson</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
