<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Range-based operations • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Range-based operations">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/steveped/extraChIPs/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Range-based operations</h1>
                        <h4 data-toc-skip class="author">Steve Pederson</h4>
            <address class="author_afil">
      Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical School, University of Adelaide<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/steveped/extraChIPs/blob/HEAD/vignettes/range_based_functions.Rmd" class="external-link"><code>vignettes/range_based_functions.Rmd</code></a></small>
      <div class="hidden name"><code>range_based_functions.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package is designed to enable the Gene Regulatory Analysis using Variable IP (<a href="https://github.com/steveped/GRAVI" class="external-link">GRAVI</a>) workflow, as a method for detecting differential binding in ChIP-Seq datasets. As a workflow focussed on data integration, most functions provided by the package <code>extraChIPs</code> are designed to enable comparison across datasets. This vignette looks primarily at functions which work with <code>GenomicRanges</code> objects</p>
</div>
<div class="section level2">
<h2 id="coercion">Coercion<a class="anchor" aria-label="anchor" href="#coercion"></a>
</h2>
<p>The advent of the <code>tidyverse</code> has led to <code>tibble</code> objects becoming a common alternative to <code>data.frame</code> or <code>DataFrame</code> objects. Simple functions within <code>extraChIP</code> enable coercion from <code>GRanges</code>, <code>GInteractions</code> and <code>DataFrame</code> objects to tibble objects, with list columns correctly handled. By default these coercion functions will coerce <code>GRanges</code> elements to a character vector. Similarly, <code>GRanges</code> represented as a character column can be coerced to the ranges element of a <code>GRanges</code> object.</p>
<p>First let’s coerce from a <code>tibble</code> (or S3 <code>data.frame</code>) to a <code>GRanges</code></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/steveped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">73</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1-10:+"</span>, <span class="st">"chr1:5-10:+"</span>, <span class="st">"chr1:5-6:+"</span><span class="op">)</span>,
  gene_id <span class="op">=</span> <span class="st">"gene1"</span>,
  tx_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"transcript"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,
  score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">df</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span>
<span class="co">##   range       gene_id tx_id        score</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> chr1:1-10:+ gene1   transcript1 0.442 </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> chr1:5-10:+ gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span>
<span class="co">## <span style="color: #BCBCBC;">3</span> chr1:5-6:+  gene1   transcript3 0.615</span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colToRanges-methods.html">colToRanges</a></span><span class="op">(</span><span class="va">df</span>, <span class="st">"range"</span><span class="op">)</span>
<span class="va">gr</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 3 ranges and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id       tx_id     score</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">##   [1]     chr1      1-10      + |       gene1 transcript1 0.4423369</span>
<span class="co">##   [2]     chr1      5-10      + |       gene1 transcript2 0.0831099</span>
<span class="co">##   [3]     chr1       5-6      + |       gene1 transcript3 0.6146112</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<p>Coercion back to a <code>tibble</code> will place the ranges as a character column by default. However, this can be turned off and the conventional coercion from <code>as.data.frame</code> will instead be applied, internally wrapped in <code><a href="../reference/as_tibble.html">as_tibble()</a></code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span>
<span class="co">##   range       gene_id tx_id        score</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> chr1:1-10:+ gene1   transcript1 0.442 </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> chr1:5-10:+ gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span>
<span class="co">## <span style="color: #BCBCBC;">3</span> chr1:5-6:+  gene1   transcript3 0.615</span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">gr</span>, rangeAsChar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 8</span></span>
<span class="co">##   seqnames start   end width strand gene_id tx_id        score</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> chr1         1    10    10 +      gene1   transcript1 0.442 </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> chr1         5    10     6 +      gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span>
<span class="co">## <span style="color: #BCBCBC;">3</span> chr1         5     6     2 +      gene1   transcript3 0.615</span></code></pre>
</div>
<div class="section level2">
<h2 id="simple-operations-retaining-mcols">Simple Operations Retaining <code>mcols()</code><a class="anchor" aria-label="anchor" href="#simple-operations-retaining-mcols"></a>
</h2>
<p>The standard set operations implemented in the package <code>GenomicRanges</code> will always drop the <code>mcols</code> element by default. The <code>extraChIPs</code> functions <code><a href="../reference/reduceMC-methods.html">reduceMC()</a></code>, <code><a href="../reference/setoptsMC-methods.html">setdiffMC()</a></code>, <code><a href="../reference/setoptsMC-methods.html">intersectMC()</a></code> and <code><a href="../reference/setoptsMC-methods.html">unionMC()</a></code> all produce the same output as their similarly-named functions, however, the <code>mcols()</code> elements in the query object are also returned. Where required, columns are coerced into <code>CompressedList</code> columns. This can particularly useful when needed to propagate the information contained in the initial ranges through to subsequent analytic steps</p>
<div class="section level3">
<h3 id="reducemc">
<code>reduceMC()</code><a class="anchor" aria-label="anchor" href="#reducemc"></a>
</h3>
<p>The classical approach to defining TSS regions for a set of transcripts would be to use the function <code>resize</code>()`, setting the width as 1.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span><span class="va">gr</span>, width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">tss</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 3 ranges and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id       tx_id     score</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span>
<span class="co">##   [1]     chr1         1      + |       gene1 transcript1 0.4423369</span>
<span class="co">##   [2]     chr1         5      + |       gene1 transcript2 0.0831099</span>
<span class="co">##   [3]     chr1         5      + |       gene1 transcript3 0.6146112</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<p>As we can see, two transcripts start at position 5, so we may choose to reduce this, which would lose the <code>mcols</code> element. The alternative <code><a href="../reference/reduceMC-methods.html">reduceMC()</a></code> will retain all <code>mcols</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span>
<span class="co">##       seqnames    ranges strand</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">##   [1]     chr1         1      +</span>
<span class="co">##   [2]     chr1         5      +</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/reduceMC-methods.html">reduceMC</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id                   tx_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;CharacterList&gt;</span>
<span class="co">##   [1]     chr1         1      + |       gene1             transcript1</span>
<span class="co">##   [2]     chr1         5      + |       gene1 transcript2,transcript3</span>
<span class="co">##                     score</span>
<span class="co">##             &lt;NumericList&gt;</span>
<span class="co">##   [1]            0.442337</span>
<span class="co">##   [2] 0.0831099,0.6146112</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<p>By default, this function will attempt to coerce <code>mcols</code> to a new <code>mcol</code> of the appropriate type, however, when multiple values are inevitable such as for the <code>tx_id</code> column above, these will be coerced to a <code>CompressedList</code>. The simplification of the multiple values seen in the <code>gene_id</code> can also be turned off if desired should repeated values be important for downstream analysis.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/reduceMC-methods.html">reduceMC</a></span><span class="op">(</span><span class="va">tss</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |         gene_id                   tx_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;         &lt;CharacterList&gt;</span>
<span class="co">##   [1]     chr1         1      + |           gene1             transcript1</span>
<span class="co">##   [2]     chr1         5      + |     gene1,gene1 transcript2,transcript3</span>
<span class="co">##                     score</span>
<span class="co">##             &lt;NumericList&gt;</span>
<span class="co">##   [1]            0.442337</span>
<span class="co">##   [2] 0.0831099,0.6146112</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<p>This allows for simple integration with <code>tidyverse</code> nesting strategies.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/reduceMC-methods.html">reduceMC</a></span><span class="op">(</span><span class="va">tss</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span>
<span class="co">##   range    gene_id tx_id        score</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> chr1:1:+ gene1   transcript1 0.442 </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> chr1:5:+ gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span>
<span class="co">## <span style="color: #BCBCBC;">3</span> chr1:5:+ gene1   transcript3 0.615</span></code></pre>
</div>
<div class="section level3">
<h3 id="set-operations">Set Operations<a class="anchor" aria-label="anchor" href="#set-operations"></a>
</h3>
<p>Whilst <code>reduce/reduceMC</code> is applied to a single <code>GRanges</code> object, the set operation functions <code>intersect</code>, <code>setdiff</code> and <code>union</code> are valuable approaches for comparing ranges. Using the <code>*MC()</code> functions will retain <code>mcols</code> elements from the <em>query range</em>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1:9-12:*"</span><span class="op">)</span>
<span class="va">peaks</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">&lt;-</span> <span class="st">"peak1"</span>
<span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/setops-methods.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 1 range and 0 metadata columns:</span>
<span class="co">##       seqnames    ranges strand</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">##   [1]     chr1      9-10      *</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/setoptsMC-methods.html">intersectMC</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 1 range and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id                   tx_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;CharacterList&gt;</span>
<span class="co">##   [1]     chr1      9-10      * |       gene1 transcript1,transcript2</span>
<span class="co">##                     score</span>
<span class="co">##             &lt;NumericList&gt;</span>
<span class="co">##   [1] 0.4423369,0.0831099</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/setoptsMC-methods.html">setdiffMC</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 1 range and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">##   [1]     chr1       1-8      * |       gene1</span>
<span class="co">##                                     tx_id                         score</span>
<span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span>
<span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/setoptsMC-methods.html">unionMC</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 1 range and 3 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">##   [1]     chr1      1-12      * |       gene1</span>
<span class="co">##                                     tx_id                         score</span>
<span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span>
<span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<p>There is a performance overhead to preparation of mcols as <code>CompressedList</code> objects and all <code>mcols</code> in the query object will be returned. If only wishing to retain a subset of <code>mcols</code>, these should be selected prior to passing to these functions.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span>
<span class="va">gr</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">select</a></span><span class="op">(</span><span class="va">tx_id</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="../reference/setoptsMC-methods.html">intersectMC</a></span><span class="op">(</span><span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 1 range and 1 metadata column:</span>
<span class="co">##       seqnames    ranges strand |                   tx_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |         &lt;CharacterList&gt;</span>
<span class="co">##   [1]     chr1      9-10      * | transcript1,transcript2</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
</div>
<div class="section level3">
<h3 id="range-paritioning">Range Paritioning<a class="anchor" aria-label="anchor" href="#range-paritioning"></a>
</h3>
<p>This enables the identification of specific ranges from the query range (<code>x</code>) which overlap ranges from the subject range (<code>y</code>) Under this approach, <code>mcols</code> from both <code>query</code> and <code>subject</code> ranges will be returned to enable the clear ranges which are common and distinct within the two sets of ranges.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/partitionRanges-methods.html">partitionRanges</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 3 ranges and 4 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">##   [1]     chr1       1-8      + |       gene1</span>
<span class="co">##   [2]     chr1       5-8      + |       gene1</span>
<span class="co">##   [3]     chr1      9-10      + |       gene1</span>
<span class="co">##                                     tx_id                         score</span>
<span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span>
<span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   [2] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   [3]             transcript1,transcript2           0.4423369,0.0831099</span>
<span class="co">##           peak_id</span>
<span class="co">##       &lt;character&gt;</span>
<span class="co">##   [1]        &lt;NA&gt;</span>
<span class="co">##   [2]        &lt;NA&gt;</span>
<span class="co">##   [3]       peak1</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/partitionRanges-methods.html">partitionRanges</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 3 ranges and 4 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">##   [1]     chr1       1-8      + |       gene1</span>
<span class="co">##   [2]     chr1       5-8      + |       gene1</span>
<span class="co">##   [3]     chr1      9-10      + |       gene1</span>
<span class="co">##                                     tx_id                         score</span>
<span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span>
<span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   [2] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   [3]             transcript1,transcript2           0.4423369,0.0831099</span>
<span class="co">##           peak_id</span>
<span class="co">##       &lt;character&gt;</span>
<span class="co">##   [1]        &lt;NA&gt;</span>
<span class="co">##   [2]        &lt;NA&gt;</span>
<span class="co">##   [3]       peak1</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/partitionRanges-methods.html">partitionRanges</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">gr</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 2 ranges and 4 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     peak_id     gene_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span>
<span class="co">##   [1]     chr1      9-10      * |       peak1       gene1</span>
<span class="co">##   [2]     chr1     11-12      * |       peak1        &lt;NA&gt;</span>
<span class="co">##                                     tx_id                         score</span>
<span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span>
<span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span>
<span class="co">##   [2]                                                                  </span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
<p>Whilst this shares some similarity with <code><a href="../reference/setoptsMC-methods.html">intersectMC()</a></code> the additional capabilities provide greater flexibility.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/partitionRanges-methods.html">partitionRanges</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">peak_id</span> <span class="op">==</span> <span class="st">"peak1"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 1 range and 4 metadata columns:</span>
<span class="co">##       seqnames    ranges strand |     gene_id                   tx_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;CharacterList&gt;</span>
<span class="co">##   [1]     chr1      9-10      + |       gene1 transcript1,transcript2</span>
<span class="co">##                     score     peak_id</span>
<span class="co">##             &lt;NumericList&gt; &lt;character&gt;</span>
<span class="co">##   [1] 0.4423369,0.0831099       peak1</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/steveped" class="external-link">Stephen Pederson</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
