<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Differential Signal Analysis • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Differential Signal Analysis">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_signal.html">Differential Signal Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/smped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Differential Signal Analysis</h1>
                        <h4 data-toc-skip class="author">Stevie
Pederson</h4>
            <address class="author_afil">
      Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide,
AustraliaDame Roma Mitchell Cancer Researc Laboratories, University of
AdelaideJohn Curtin School of Medical Research, Australian National
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/smped/extraChIPs/blob/HEAD/vignettes/differential_signal.Rmd" class="external-link"><code>vignettes/differential_signal.Rmd</code></a></small>
      <div class="hidden name"><code>differential_signal.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <a href="https://github.com/smped/GRAVI" class="external-link">GRAVI</a> workflow, for
which this package is designed, uses sliding windows for differential
signal analysis in a manner similar to the package <code>csaw</code>,
but also incorporating <code>macs2</code> peaks. The workflow itself
extends to integrating multiple ChIP targets and external data sources,
and as such, this package introduces a handful of functions to simplify
and enable these analyses. Whilst many existing approaches refer to this
type of analysis as Differential Binding analysis, we prefer the term
<em>Differential Signal Analysis</em> as this more accurately captures
the range of ChIP targets which are likely to be investigated.</p>
<p>The majority of examples below use heavily reduced datasets to
provide general guidance on using the functions. Some results may appear
trivial as a result, but will hopefully prove far more useful in a true
experimental context. All data, along with this vignette are available
<a href="https://github.com/smped/extraChIPs_vignette" class="external-link">here</a>. Please
place all contents of the data directory in a directory named data in
your own working directory.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>In order to use the package <code>extraChIPs</code> and follow this
vignette, we recommend using the package <code>BiocManager</code> hosted
on CRAN. Once this is installed, the additional packages required for
this vignette (<code>tidyverse</code>, <code>Rsamtools</code>,
<code>csaw</code>, <code>BiocParallel</code> and
<code>rtracklayer</code>) can also be installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="va">pkg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"tidyverse"</span>, <span class="st">"Rsamtools"</span>, <span class="st">"csaw"</span>, <span class="st">"BiocParallel"</span>, <span class="st">"rtracklayer"</span>, <span class="st">"edgeR"</span>, </span>
<span>  <span class="st">"patchwork"</span>, <span class="st">"extraChIPs"</span>, <span class="st">"plyranges"</span>, <span class="st">"scales"</span>, <span class="st">"ggside"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">pkg</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Once these packages are installed, we can load them easily</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csaw</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jtlandis/ggside" class="external-link">ggside</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h3>
<p>All data for this vignette is expected to be in a sub-directory of
the working directory named “data”, and all paths will be predicated on
this. Please ensure you have all data in this location, obtained from <a href="https://github.com/smped/extraChIPs_vignette" class="external-link">here</a>.</p>
<p>The data itself is ChIP-Seq data targeting the histone mark H3K27ac,
and is taken from the cell-line MDA-MB-453 under Vehicle and
DHT-stimulated conditions. Using CRCh37 as the reference genome, a
subset of regions found on chromosome 10 are included in this dataset
for simplicity.</p>
<p><code>Seqinfo</code> objects are the foundation of working with
GRanges, so let’s define a suitable object for consistency throughout th
analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hg19</span> <span class="op">&lt;-</span> <span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/getChromInfoFromUCSC.html" class="external-link">getChromInfoFromUCSC</a></span><span class="op">(</span><span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="va">grch37</span> <span class="op">&lt;-</span> <span class="va">hg19</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chrom</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>genome <span class="op">=</span> <span class="st">"GRCh37"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="va">chrom</span>, seqlengths <span class="op">=</span> <span class="va">size</span>, isCircular <span class="op">=</span> <span class="va">circular</span>, <span class="va">genome</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="st">"Seqinfo"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="working-with-peaks">Working With Peaks<a class="anchor" aria-label="anchor" href="#working-with-peaks"></a>
</h2>
<p>The provided dataset includes six files produced by
<code>macs2 callpeak</code> (Zhang et al. 2008) in the
<code>narrowPeak</code> format, and these are able to be easily parsed
using <code>extraChIPs</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peakFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"data"</span>, pattern <span class="op">=</span> <span class="st">"narrowPeak"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importPeaks.html">importPeaks</a></span><span class="op">(</span><span class="va">peakFiles</span>, seqinfo <span class="op">=</span> <span class="va">grch37</span><span class="op">)</span></span></code></pre></div>
<p>This will import the peaks from all files as a single
<code>GRangesList</code> object, adding the file-name to each element by
default. We can easily modify these names if we so wish.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove_all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span>, <span class="st">"_peaks.narrowPeak"</span><span class="op">)</span></span></code></pre></div>
<p>Once loaded, we can easily check how similar our replicates are.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOverlaps-methods.html">plotOverlaps</a></span><span class="op">(</span><span class="va">peaks</span>, min_size <span class="op">=</span> <span class="fl">10</span>, .sort_sets <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-overlaps-1.png" alt=""><p class="caption"><em>UpSet plot showing overlapping peaks across all
replicates</em></p>
</div>
<p>Optionally, specifying a column and a suitable function will produce
an additional panel summarising that value. In the following, we’ll show
the maximum score obtained, highlighting that for peaks identified in
only one or two replicates, the overall signal intensity is generally
lower.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOverlaps-methods.html">plotOverlaps</a></span><span class="op">(</span><span class="va">peaks</span>, min_size <span class="op">=</span> <span class="fl">10</span>, .sort_sets <span class="op">=</span> <span class="cn">FALSE</span>, var <span class="op">=</span> <span class="st">"score"</span>, f <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-overlaps-score-1.png" alt=""><p class="caption"><em>UpSet plot showing overlapping peaks across all
replicates, with the maximum score across all replicates shown in the
upper panel.</em></p>
</div>
<p>A common task at this point may be to define consensus peaks within
each treatment group, by retaining only the peaks found in 2 of the 3
replicates. The default approach is to take the union of all ranges,
with the returned object containing logical values for each sample, as
well as the number of samples where an overlapping peak was found.</p>
<p>If we wish to retain any of the original columns, such as the
<code>macs2 callpeak</code> score, we can simply pass the column names
to <code><a href="../reference/makeConsensus.html">makeConsensus()</a></code></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_veh</span> <span class="op">&lt;-</span> <span class="va">peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">"Veh"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span></span>
<span><span class="va">consensus_dht</span> <span class="op">&lt;-</span> <span class="va">peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">"DHT"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">2</span> <span class="op">/</span> <span class="fl">3</span>, var <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, we could find the centre of the peaks as part of this
process, by averaging across the estimated peak centres for each sample.
Whilst this is very common for <em>transcription factor</em> peaks, this
may be less informative for other types of ChIP targets, such as the
histone marks we have.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_veh</span> <span class="op">&lt;-</span> <span class="va">peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">"Veh"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">endoapply</span><span class="op">(</span><span class="va">mutate</span>, centre <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">peak</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">consensus_dht</span> <span class="op">&lt;-</span> <span class="va">peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">"DHT"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">endoapply</span><span class="op">(</span><span class="va">mutate</span>, centre <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">peak</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can also inspect these using <code><a href="../reference/plotOverlaps-methods.html">plotOverlaps()</a></code> provided
we use a <code>GRangesList</code> for the input.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span></span>
<span>  Veh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">consensus_veh</span><span class="op">)</span>, DHT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">consensus_dht</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotOverlaps-methods.html">plotOverlaps</a></span><span class="op">(</span>set_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey70"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/venn-consensus-overlap-1.png" alt=""><p class="caption"><em>Overlap between consensus peaks identified in a
treatment-specific manner</em></p>
</div>
<p>We could go one step further and define the set of peaks found in
either treatment. Given we’re being inclusive here, we can leave p = 0
so any peak found in either treatment is included.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span></span>
<span>  Veh <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">consensus_veh</span>, <span class="va">centre</span><span class="op">)</span>, DHT <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">consensus_dht</span>, <span class="va">centre</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-signal-analysis">Differential Signal Analysis<a class="anchor" aria-label="anchor" href="#differential-signal-analysis"></a>
</h2>
<div class="section level3">
<h3 id="sliding-windows">Sliding Windows<a class="anchor" aria-label="anchor" href="#sliding-windows"></a>
</h3>
<p>The standard approach of tools such as DiffBind (Ross-Innes et
al. 2012) is to take a set of peaks, re-centre them, then set all
regions to be the same width. From there, data is passed to
<code>edgeR</code> (Chen, Lun, and Smyth 2016) or <code>DESeq2</code>
(Love, Huber, and Anders 2014) for analysis. Whilst this approach can be
replicated using the defined peaks above, the suggested approach for
<code>extraChIPs</code> is to us sliding windows, as per
<code>csaw</code>. The resultant <em>variable width regions</em> can be
particularly advantageous for ChIP targets such as H3K27ac where regions
marked by histone-acetylation can vary greatly in size.</p>
<p>The starting point for differential signal analyses using
<code>extraChIPs</code> is to define a set of sliding windows across the
genome, then count reads from a set of bam files, defined as a
<code>BamFileList.</code> Commonly one or more IP input/control samples
is also produced during a ChIP-Seq experiment, and these should be
included at this stage of the analysis. The example files provided here
contain a small subset of reads from chromosome 10 across two
experimental conditions and one input sample, and we will define them
all as a <code>BamFileList</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"data"</span>, pattern <span class="op">=</span> <span class="st">"bam$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove_all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span>, <span class="st">".bam"</span><span class="op">)</span></span></code></pre></div>
<p><strong>NB:</strong> It should also be noted that counting all reads
across a <code>BamFileList</code> using sliding windows, <strong>will
require a significant amount of RAM</strong> and will be beyond the
capacity of most laptops as of the time of writing. When working with
complete datasets, this step is best performed on an HPC or a similar
interactive server.</p>
<p>The approach taken below is to first define a set of sliding windows
across the genome, using the capabilities of <code>csaw</code>. After
counting reads across all windows, a set of pre-defined regions is then
used guide the function <code><a href="../reference/dualFilter.html">dualFilter()</a></code> which will discard
low-signal windows, retaining only those a) above a minimum signal level
and b) with signal notably above that of any input samples. These
regions can be obtained from any external resource, or can even be taken
from <code>macs2</code>-defined peaks from the same samples.</p>
<p>First we can define our windows and count the alignments using the
existing capabilities and functions provided in the <code>csaw</code>
package (Lun and Smyth 2016). In the following, we’ll use a sliding
window of 120bp and a step size of 40bp, meaning each nucleotide is
covered by 3 windows. In addition, we’ll exclude blacklisted and
greylisted regions as provided in the dataset. These can be obtained
easily by using the <code>GreyListChIP</code> package, which is beyond
the scope of this vignette.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">greylist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span><span class="st">"data/chr10_greylist_subset.bed"</span>, seqinfo <span class="op">=</span> <span class="va">grch37</span><span class="op">)</span></span>
<span><span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span><span class="st">"data/chr10_blacklist_subset.bed"</span>, seqinfo <span class="op">=</span> <span class="va">grch37</span><span class="op">)</span></span>
<span><span class="va">rp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/readParam.html" class="external-link">readParam</a></span><span class="op">(</span></span>
<span>  pe <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  dedup <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  restrict <span class="op">=</span> <span class="st">"chr10"</span>,</span>
<span>  discard <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">greylist</span>, <span class="va">blacklist</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">wincounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/windowCounts.html" class="external-link">windowCounts</a></span><span class="op">(</span></span>
<span>  bam.files <span class="op">=</span> <span class="va">bfl</span>,</span>
<span>  spacing <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">120</span>,</span>
<span>  ext <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span>,</span>
<span>  param <span class="op">=</span> <span class="va">rp</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This produces a <code>RangesSummarizedExperiment</code> with windows
included which passed the minimum threshold of 7 total reads. We can
check which windows passed this threshold using
<code>rowRanges()</code></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 262151 ranges and 0 metadata columns:</span></span>
<span><span class="co">##            seqnames            ranges strand</span></span>
<span><span class="co">##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]    chr10 42820161-42820280      *</span></span>
<span><span class="co">##        [2]    chr10 42820201-42820320      *</span></span>
<span><span class="co">##        [3]    chr10 42820241-42820360      *</span></span>
<span><span class="co">##        [4]    chr10 42820281-42820400      *</span></span>
<span><span class="co">##        [5]    chr10 42821001-42821120      *</span></span>
<span><span class="co">##        ...      ...               ...    ...</span></span>
<span><span class="co">##   [262147]    chr10 99998721-99998840      *</span></span>
<span><span class="co">##   [262148]    chr10 99998761-99998880      *</span></span>
<span><span class="co">##   [262149]    chr10 99998801-99998920      *</span></span>
<span><span class="co">##   [262150]    chr10 99998841-99998960      *</span></span>
<span><span class="co">##   [262151]    chr10 99998881-99999000      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<p>We can also add some key information to the <code>colData</code>
element of this object, which will also be propagated to all downstream
objects.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wincounts</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span></span>
<span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span>, <span class="st">"(Veh|DHT)"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct.html" class="external-link">fct</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Veh"</span>, <span class="st">"DHT"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 7 rows and 6 columns</span></span>
<span><span class="co">##            bam.files    totals       ext      rlen      sample    treat</span></span>
<span><span class="co">##          &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;factor&gt;</span></span>
<span><span class="co">## DHT_1 data/DHT_1.bam    252973       200        74       DHT_1      DHT</span></span>
<span><span class="co">## DHT_2 data/DHT_2.bam    295028       200        74       DHT_2      DHT</span></span>
<span><span class="co">## DHT_3 data/DHT_3.bam    295105       200        74       DHT_3      DHT</span></span>
<span><span class="co">## Input data/Input.bam     82276       200        72       Input      NA </span></span>
<span><span class="co">## Veh_1 data/Veh_1.bam    250970       200        74       Veh_1      Veh</span></span>
<span><span class="co">## Veh_2 data/Veh_2.bam    280898       200        74       Veh_2      Veh</span></span>
<span><span class="co">## Veh_3 data/Veh_3.bam    294356       200        74       Veh_3      Veh</span></span></code></pre>
<p>A density plot can be simply drawn of these counts, with the vast
majority of windows receiving very low counts, due to the nature of
transcription factor binding, where long stretches are unbound. The
windows with higher counts tend to be associated with the samples
targeting a transcription factor (TF), as seen in the two treatment
group samples.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">wincounts</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, trans <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-densities-1.png" alt=""><p class="caption"><em>Read Densities for all returned windows across
all samples</em></p>
</div>
</div>
<div class="section level3">
<h3 id="filtering-of-sliding-windows">Filtering of Sliding Windows<a class="anchor" aria-label="anchor" href="#filtering-of-sliding-windows"></a>
</h3>
<p>After counting all reads in the sliding genomic windows, the next
step is to discard windows for which counts are unlikely to represent
true signal from our ChIP target. The strategy employed in
<code>extraChIPs</code> uses a set of consensus peaks to automatically
set thresholds based on 1) counts strongly above the counts from the
input sample, and 2) the windows with the overall highest signal.
Thresholds are determined such that a proportion
(e.g. <code>q = 0.5</code>) of the windows which overlap one of the
supplied consensus peaks will be returned. Higher values for
<code>q</code> will return more windows, however many of these will tend
to only marginally overlap a peak in one of the tail regions, and these
will most likely be covered by neighbouring windows. Experience has
shown that values such as <code>q = 0.5</code> tend to return a
considerable proportion of windows containing true signal from the ChIP
target.</p>
<p>The we can pass these to the function <code><a href="../reference/dualFilter.html">dualFilter()</a></code> which
utilises the strategy described above. On large datasets, this can be
quite time-consuming, as can the initial counting step. Multiple
alternative filtering strategies are also provided by the package
<code>csaw</code> and these can be accessed using
<code><a href="https://rdrr.io/pkg/csaw/man/defunct.html" class="external-link">?csaw::filterWindows</a></code></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dualFilter.html">dualFilter</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  bg <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span><span class="op">)</span><span class="op">]</span>, </span>
<span>  ref <span class="op">=</span> <span class="va">all_consensus</span>,</span>
<span>  q <span class="op">=</span> <span class="fl">0.6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Thus we have reduced our initial set of 262,151 sliding windows to
the 16,141 windows most likely to contain true signal from our ChIP
target. The returned object will by default contain <code>counts</code>
and <code>logCPM</code> assays, with the complete library sizes used for
the calculation of <code>logCPM</code> values. Similarly, <em>the input
sample is no longer included</em> in the data object, although
additional columns can easily be added to the returned object using any
number of strategies.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 262151      7</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 16141     6</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): counts logCPM</span></span></code></pre>
<p>We can once again check our signal distributions, this time on the
logCPM values.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plotcpm-1.png" alt=""><p class="caption"><em>Densities for logCPM values across all samples
after discarding windows less likely to contain H3K27ac signal</em></p>
</div>
<p>The <code>rowData</code> element of the returned object will contain
a logical column indicating where each specific retained window
overlapped one of the supplied consensus peaks.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 16141 ranges and 1 metadata column:</span></span>
<span><span class="co">##           seqnames            ranges strand | overlaps_ref</span></span>
<span><span class="co">##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;logical&gt;</span></span>
<span><span class="co">##       [1]    chr10 43047561-43047680      * |         TRUE</span></span>
<span><span class="co">##       [2]    chr10 43047601-43047720      * |         TRUE</span></span>
<span><span class="co">##       [3]    chr10 43047641-43047760      * |         TRUE</span></span>
<span><span class="co">##       [4]    chr10 43047681-43047800      * |         TRUE</span></span>
<span><span class="co">##       [5]    chr10 43047721-43047840      * |         TRUE</span></span>
<span><span class="co">##       ...      ...               ...    ... .          ...</span></span>
<span><span class="co">##   [16137]    chr10 99894961-99895080      * |         TRUE</span></span>
<span><span class="co">##   [16138]    chr10 99895001-99895120      * |         TRUE</span></span>
<span><span class="co">##   [16139]    chr10 99895041-99895160      * |         TRUE</span></span>
<span><span class="co">##   [16140]    chr10 99895081-99895200      * |         TRUE</span></span>
<span><span class="co">##   [16141]    chr10 99895121-99895240      * |         TRUE</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">overlaps_ref</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.9997522</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="initial-visualisation">Initial Visualisation<a class="anchor" aria-label="anchor" href="#initial-visualisation"></a>
</h3>
<p>Inspecting your data is a common first step, and a common QC step is
Relative Log-Expression (RLE) (Gandolfo and Speed 2018). In the
following, we first inspect the RLE across the entire dataset, followed
by RLE grouping <em>within treatments</em>. This can be particularly
useful when distributions vary significantly between treatment groups,
such as may occur with a cytoplasmic to nuclear shift by a given ChIP
target. Here, however, there is minimal difference between the two
approaches as H3K27ac signal tends to be broadly consistent between
these treatment groups.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAssayRle-methods.html">plotAssayRle</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, fill <span class="op">=</span> <span class="st">"treat"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, colour <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"RLE: Across All Samples"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAssayRle-methods.html">plotAssayRle</a></span><span class="op">(</span></span>
<span>  <span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, fill <span class="op">=</span> <span class="st">"treat"</span>, rle_group <span class="op">=</span> <span class="st">"treat"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, colour <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"RLE: Within Treatment Groups"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">+</span> <span class="va">b</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-assay-rle-1.png" alt=""><p class="caption"><em>RLE plots across all samples (A) and with values
calculated within treatment groups (B).</em></p>
</div>
<p>We can also check the samples using a PCA plot, again colouring the
points by treatment group and adding labels, which will repel by default
if the points are shown.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssayPCA-methods.html">plotAssayPCA</a></span><span class="op">(</span><span class="va">filtcounts</span>, <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, label <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-pca-1.png" alt=""><p class="caption"><em>PCA plot based on the logCPM assay</em></p>
</div>
</div>
<div class="section level3">
<h3 id="statistical-testing">Statistical Testing<a class="anchor" aria-label="anchor" href="#statistical-testing"></a>
</h3>
<p>Multiple methods are enabled in the package <code>extraChIPs</code>
via the function <code><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff()</a></code>, with the possibility of
incorporating any additional normalisation strategies from external
packages. The two basic strategies are 1) Quasi-Likelihood Fits (Lund et
al. 2012) and 2) <code>limma-trend</code> (Law et al. 2014). This first
(method = “qlf”) uses counts with any of the provided normalisation
strategies from <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code>, and setting the
normalisation method to “none” is the equivalent of library-size
normalisation, which replicates the default normalisation strategy from
DiffBind (Ross-Innes et al. 2012). If choosing to normalise within
treatment groups, a factor can be provided via the groups argument,
essentially adding this as an option for all methods provided in
<code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code>. The second method (method = “lt”)
is specifically for logCPM values and these can be provided as output by
<code><a href="../reference/dualFilter.html">dualFilter()</a></code> or may be normalised using any number of
additional methods. In addition to the above methods, a range-based
<span class="math inline">\(H_0\)</span> (McCarthy and Smyth 2009) can
be specified by providing a value to the <code>fc</code> or
<code>lfc</code> arguments.</p>
<p>Here, we’ll fit our data using Quasi-Likelihood Fits, library-size
normalisation and setting a change in signal beyond the range of <span class="math inline">\(\pm\)</span> 20% as being of interest. By default,
the returned object, will contain the results from model fitting in the
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> element as these are result associated with each
row element in the <code>SummarizedExperiment</code> object. If the
object is a <code>RangedSummarizedExperiment</code> object, setting
<code>asRanges = TRUE</code> will simply return the set of GRanges along
with the testing results.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">treat</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff</a></span><span class="op">(</span><span class="va">filtcounts</span>, design <span class="op">=</span> <span class="va">X</span>, fc <span class="op">=</span> <span class="fl">1.2</span>, asRanges <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-windows">Merging Windows<a class="anchor" aria-label="anchor" href="#merging-windows"></a>
</h3>
<p>After an analysis has been performed, common values contained in the
output may be estimated signal (<code>logCPM</code>), estimated change
(<code>logFC</code>) with both raw and adjusted p-values. Given the
dependency of neighbouring windows, any adjusted p-values will not be
appropriate and a merging of overlapping and/or neighbouring windows
should be performed. Multiple <code>csaw</code> methods are wrapped
using <code><a href="../reference/mergeByCol-methods.html">mergeByCol()</a></code>, <code><a href="../reference/mergeBySig-methods.html">mergeBySig()</a></code> with minor
changes to the returned object, such as the inclusion of the
representative range in the column <code>keyval_range</code>.</p>
<p>For this vignette, we’ll merge using the asymptotically exact
harmonic mean p-value, which can also be used for merging dependent
p-values (Wilson 2019). When merging windows using the harmonic mean
p-values, instead of values from a representative window, weighted
averages for the expression and logFC estimates are returned using the
weights <span class="math inline">\(w_i = \frac{1}{p_i}\)</span>. A
representative window, corresponding to the original window with the
lowest p-value is returned.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeByHMP-methods.html">mergeByHMP</a></span><span class="op">(</span><span class="va">fit_gr</span>, inc_cols <span class="op">=</span> <span class="st">"overlaps_ref"</span>, merge_within <span class="op">=</span> <span class="fl">120</span><span class="op">)</span></span>
<span><span class="va">results_gr</span><span class="op">$</span><span class="va">status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>  <span class="va">results_gr</span><span class="op">$</span><span class="va">hmp_fdr</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">~</span> <span class="st">"Unchanged"</span>,</span>
<span>  <span class="va">results_gr</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Increased"</span>,</span>
<span>  <span class="va">results_gr</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Decreased"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">hmp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 5 ranges and 10 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand | n_windows      n_up    n_down</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]    chr10 79266481-79268400      * |        38         3         0</span></span>
<span><span class="co">##   [2]    chr10 43689161-43690240      * |        25         4         0</span></span>
<span><span class="co">##   [3]    chr10 58717481-58717840      * |         7         3         0</span></span>
<span><span class="co">##   [4]    chr10 67671561-67671800      * |         4         3         0</span></span>
<span><span class="co">##   [5]    chr10 67670801-67671160      * |         7         1         0</span></span>
<span><span class="co">##       overlaps_ref            keyval_range    logCPM     logFC         hmp</span></span>
<span><span class="co">##          &lt;logical&gt;               &lt;GRanges&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]         TRUE chr10:79266721-79266840   7.60592   2.33531 1.35064e-13</span></span>
<span><span class="co">##   [2]         TRUE chr10:43689401-43689520   7.43019   2.28678 4.79764e-13</span></span>
<span><span class="co">##   [3]         TRUE chr10:58717521-58717640   6.75830   2.52756 4.78855e-11</span></span>
<span><span class="co">##   [4]         TRUE chr10:67671601-67671720   6.64579   2.54747 1.10884e-10</span></span>
<span><span class="co">##   [5]         TRUE chr10:67670801-67670920   6.56879   2.62167 3.73516e-10</span></span>
<span><span class="co">##           hmp_fdr      status</span></span>
<span><span class="co">##         &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1] 7.50956e-11   Increased</span></span>
<span><span class="co">##   [2] 1.33375e-10   Increased</span></span>
<span><span class="co">##   [3] 8.87478e-09   Increased</span></span>
<span><span class="co">##   [4] 1.54129e-08   Increased</span></span>
<span><span class="co">##   [5] 4.15350e-08   Increased</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<p>In the above, we returned 50 ranges which we might consider using the
significance threshold <span class="math inline">\(\alpha\)</span> =
0.05. A particularly beneficial feature of this approach is that the
final ranges will be of highly variable width, with this select region
of chromosome 10 producing merged windows ranging from 120 to 17360bp,
as may be expected for H3K27ac signal.</p>
<p>We can also quickly check out results using a heatmap across the
retained windows which correspond to our results for differential
signal, although this will not provide any specific genomic context.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtcounts</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">hmp</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">hmp</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/plotAssayHeatmap-methods.html">plotAssayHeatmap</a></span><span class="op">(</span>assay <span class="op">=</span> <span class="st">"logCPM"</span>, ysideline <span class="op">=</span> <span class="cn">TRUE</span>, yside_col <span class="op">=</span> <span class="st">"treat"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggside/man/ggside-scales-continuous.html" class="external-link">scale_ysidex_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>, minor_breaks <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-assay-heatmap-1.png" alt=""><p class="caption"><em>Heatmap showing retained sliding windows whih
correspond to the most highly-ranked region for differential
signal.</em></p>
</div>
</div>
<div class="section level3">
<h3 id="mapping-of-windows-to-genes">Mapping of Windows To Genes<a class="anchor" aria-label="anchor" href="#mapping-of-windows-to-genes"></a>
</h3>
<p>Once the changes in signal for our given ChIP target have been
determined, a common next step is to assess which genes are likely to be
impacted. Whilst no definitive, single methodology exists for this
process, the function <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> offers an intuitive
approach, taking into account any previously defined regulatory
features. These regulatory features may be defined by simple proximity
to TSS regions, by histone marks, downloaded from external repositories
or any other possibility. Whilst these features can improve the
precision of mapping, even without these this function can still enable
a useful assignment of target gene to binding event.</p>
<p>The process undertaken inside <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> is a
sequential checking of each range’s association with regulatory features
and the most likely target as a result. These steps are:</p>
<ol style="list-style-type: decimal">
<li><strong>Check for any HiC interactions</strong></li>
</ol>
<ul>
<li>All genes which directly overlap an interaction anchor are
considered part of the regulatory network for that interaction, and as
such, all genes associated with both anchors are assigned to a peak
which overlaps a HiC Interaction</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><strong>Check for any overlaps with a promoter</strong></li>
</ol>
<ul>
<li>All genes regulated by that promoter are assigned as regulatory
targets. By default, this is by direct promoter/gene overlap
(<code>prom2gene = 0</code>)</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li><strong>Check for any overlaps with an enhancer</strong></li>
</ol>
<ul>
<li>Peaks which overlap an enhancer are assigned to <em>all</em> genes
within the distance specified by <code>enh2gene</code> (default =
100kb)</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li><strong>Check for genes with no previous mappings</strong></li>
</ol>
<ul>
<li>Peaks <em>with no previous mappings</em> are assigned to all
directly overlapping genes, or the nearest gene within a specified
distance (default <code>gr2gene</code> = 100kb)</li>
</ul>
<p>As a result, if no promoters, enhancers or long-range interactions
are supplied, all genes will be mapped to peaks using step 4.</p>
<p>A set of annotated regions has been provided for our subset of
chromosome 10, as has the set of genes within this region. These can be
loaded as follows.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html" class="external-link">read_rds</a></span><span class="op">(</span><span class="st">"data/chr10_region_subset.rds"</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="st">"data/chr10_gene_subset.gtf"</span><span class="op">)</span></span></code></pre></div>
<p>For our mapping steps, we’ll simply use the <code>promoters</code>
element as these are the regions directly overlapping a TSS for all
transcripts within this region of the genome.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapByFeature.html">mapByFeature</a></span><span class="op">(</span><span class="va">results_gr</span>, genes <span class="op">=</span> <span class="va">genes</span>, prom <span class="op">=</span> <span class="va">regions</span><span class="op">$</span><span class="va">promoters</span><span class="op">)</span></span></code></pre></div>
<p>Now we have our regions showing changed signal along with the likely
regulatory targets. Our top-ranked region is as follows, and this
appears to be associated with the gene <em>KCNMA1</em>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">hmp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 1 range and 12 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand | n_windows      n_up    n_down</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]    chr10 79266481-79268400      * |        38         3         0</span></span>
<span><span class="co">##       overlaps_ref            keyval_range    logCPM     logFC         hmp</span></span>
<span><span class="co">##          &lt;logical&gt;               &lt;GRanges&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]         TRUE chr10:79266721-79266840   7.60592   2.33531 1.35064e-13</span></span>
<span><span class="co">##           hmp_fdr      status         gene_id       gene_name</span></span>
<span><span class="co">##         &lt;numeric&gt; &lt;character&gt; &lt;CharacterList&gt; &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1] 7.50956e-11   Increased ENSG00000156113          KCNMA1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<div class="section level4">
<h4 id="mapping-of-windows-to-regions">Mapping of Windows to Regions<a class="anchor" aria-label="anchor" href="#mapping-of-windows-to-regions"></a>
</h4>
<p>If we also have a set of annotated regions, we can easily map our
regions to the region that it has the greatest proportion of overlap.
First, we’ll arrange our regions to be a single <code>GRanges</code>
object with the column “region”. From these we cn simply add the column
<code>bestRegion</code></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions_gr</span> <span class="op">&lt;-</span> <span class="va">regions</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">select</span>, <span class="va">region</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">region_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">regions</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">region</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">results_gr</span><span class="op">$</span><span class="va">bestRegion</span> <span class="op">&lt;-</span> <span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/bestOverlap-methods.html">bestOverlap</a></span><span class="op">(</span><span class="va">regions_gr</span>, var <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct.html" class="external-link">fct</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="va">region_levels</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualisation-of-results">Visualisation of Results<a class="anchor" aria-label="anchor" href="#visualisation-of-results"></a>
</h2>
<div class="section level3">
<h3 id="association-with-annotated-features">Association with Annotated Features<a class="anchor" aria-label="anchor" href="#association-with-annotated-features"></a>
</h3>
<p>The association of windows or peaks with defined features, such as
histone marks or regulatory elements can be important for describing the
binding characteristics of any given transcription factor. We have
already defined the association of the merged windows with annotated
regions, and we can easily visualise these using
<code><a href="../reference/plotPie-methods.html">plotPie()</a></code>.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotPie-methods.html">plotPie</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"bestRegion"</span><span class="op">)</span></span></code></pre></div>
<p>However, given the default plots can often be slightly
unsatisfactory, <code><a href="../reference/plotPie-methods.html">plotPie()</a></code> is heavily customisable, in
particular taking advantage of the <code>glue</code> syntax to customise
labels.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotPie-methods.html">plotPie</a></span><span class="op">(</span></span>
<span>    fill <span class="op">=</span> <span class="st">"bestRegion"</span>, min_p <span class="op">=</span> <span class="fl">0.05</span>, total_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    cat_alpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    cat_glue <span class="op">=</span> <span class="st">"{str_wrap(bestRegion, 10)}\n{n}\n{percent(p, 0.1)}"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-pie-1.png" alt=""><p class="caption"><em>Pie chart showing the best overlapping region for
each merged windows. The best overlap is determined simply by the region
with the lergest amount of overlap.</em></p>
</div>
<p>We can also scale by additional columns. Here, we’ll find the
proportion of each window which overlaps each type of region, providing
a complete summary of the proportion of our ranges which overlaps each
feature.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/propOverlap-methods.html">propOverlap</a></span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">results_gr</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"region"</span>, values_to <span class="op">=</span> <span class="st">"p"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    region <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/fct.html" class="external-link">fct</a></span><span class="op">(</span><span class="va">region_levels</span><span class="op">[</span><span class="va">region</span><span class="op">]</span>, levels <span class="op">=</span> <span class="va">region_levels</span><span class="op">)</span>,</span>
<span>    w <span class="op">=</span> <span class="va">p</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">range</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1e3</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotPie-methods.html">plotPie</a></span><span class="op">(</span></span>
<span>    scale_by <span class="op">=</span> <span class="st">"w"</span>, fill <span class="op">=</span> <span class="st">"region"</span>, min_p <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    total_glue <span class="op">=</span> <span class="st">"{round(N, 1)}kb"</span>, total_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    cat_glue <span class="op">=</span> <span class="st">"{str_wrap(region, 10)}\n{percent(p, 0.1)}"</span>,</span>
<span>    cat_alpha <span class="op">=</span> <span class="fl">0.5</span>, cat_size <span class="op">=</span> <span class="fl">4</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-pie-width-1.png" alt=""><p class="caption"><em>Pie chart with regions scaled by width to show
the proportions of the total regions which overlap each type of
annotated region.</em></p>
</div>
<p>These results can be extended further using
<code><a href="../reference/plotSplitDonut-methods.html">plotSplitDonut()</a></code> to show more complex results.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotSplitDonut-methods.html">plotSplitDonut</a></span><span class="op">(</span>inner <span class="op">=</span> <span class="st">"status"</span>, outer <span class="op">=</span> <span class="st">"bestRegion"</span><span class="op">)</span></span></code></pre></div>
<p>Again, this plot is heavily customisable, and is able to utilise
separate palettes for the inner and outer rings if preferred. Specific
slices can also be <em>exploded</em> for emphasis.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">region_levels</span><span class="op">)</span>, <span class="st">"Viridis"</span>, rev <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">status</span> <span class="op">!=</span> <span class="st">"Unchanged"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotSplitDonut-methods.html">plotSplitDonut</a></span><span class="op">(</span></span>
<span>    inner <span class="op">=</span> <span class="st">"status"</span>, outer <span class="op">=</span> <span class="st">"bestRegion"</span>, min_p <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>    inner_palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#3333E6"</span>, <span class="st">"#E6334D"</span><span class="op">)</span>,</span>
<span>    outer_palette <span class="op">=</span> <span class="va">region_col</span>,</span>
<span>    inner_glue <span class="op">=</span> <span class="st">"H3K27ac\n{status}\n{n}"</span>, inner_label_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    outer_glue <span class="op">=</span> <span class="st">"{str_wrap(bestRegion, 10)}\n{n}"</span>, outer_label <span class="op">=</span> <span class="st">"text"</span>,</span>
<span>    explode_outer <span class="op">=</span> <span class="st">"Promoter"</span>, explode_r <span class="op">=</span> <span class="fl">0.2</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-split-donut-1.png" alt=""><p class="caption"><em>Summary of regions with changed signal and the
annotated region with the largest overlap</em></p>
</div>
</div>
<div class="section level3">
<h3 id="profile-heatmaps">Profile Heatmaps<a class="anchor" aria-label="anchor" href="#profile-heatmaps"></a>
</h3>
<p>A very common approach to visualising the results of altered TF
binding is to plot <em>profile heatmaps</em> centred around the window
(or peak), and extending out a given number of of bases. The data
required for this is referred to in <code>extraChIPs</code> as profile
data, and given that extracting this from a set of
<code>BigWigFile</code>s can be time consuming, this step is performed
prior to the actual plotting, so that ranges can be added or excluded as
desired.</p>
<p>First we need to define a <code>BigWigFileList</code> as these are
conventionally very large files which shouldn’t be retained in memory,
but are just accessed to import the key regions for a particular
process. Typically, these can be generated during peak calling but they
can also be created directly from <code>bam</code> files if needed. Once
again, our reduced dataset makes this trivial, but the process may
require significant computation resources if being performed on a
complete dataset. In the following, we’ll create the BigWig files an
immediately create our <code>BigWigFileList</code>.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sbp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">ScanBamParam</a></span><span class="op">(</span>which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr10:40000000-100000000"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Sum the coverage across all Vehicle samples then create a merged BigWig</span></span>
<span><span class="va">cov_veh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"Veh"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html" class="external-link">coverage</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, param <span class="op">=</span> <span class="va">sbp</span><span class="op">)</span><span class="op">$</span><span class="va">chr10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">RleList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cov_veh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">RleList</a></span><span class="op">(</span>chr10 <span class="op">=</span> <span class="va">cov_veh</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">export</span><span class="op">(</span><span class="va">cov_veh</span>, <span class="st">"data/Veh_merged.bw"</span><span class="op">)</span></span>
<span><span class="co">## Repeat for DHT</span></span>
<span><span class="va">cov_dht</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"DHT"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html" class="external-link">coverage</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, param <span class="op">=</span> <span class="va">sbp</span><span class="op">)</span><span class="op">$</span><span class="va">chr10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">RleList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cov_dht</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/AtomicList-class.html" class="external-link">RleList</a></span><span class="op">(</span>chr10 <span class="op">=</span> <span class="va">cov_dht</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">export</span><span class="op">(</span><span class="va">cov_dht</span>, <span class="st">"data/DHT_merged.bw"</span><span class="op">)</span></span>
<span><span class="va">bwfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFileList</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Veh <span class="op">=</span> <span class="st">"data/Veh_merged.bw"</span>, DHT <span class="op">=</span> <span class="st">"data/DHT_merged.bw"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now we have our <code>BigWigFileList</code> we can define the ranges
to plot along with the profile data. In order to find the region within
each range with the maximal signal, we’ll compare back to the original
results before windows were merged, then return the original window with
the maximal signal (i.e. logCPM). Note that we already had the windows
in “keyval_range” which corresponded to the ranges with the most
statistically significant change. Realistically we can choose any
strategy for centring ranges that we wish.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr_increase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">status</span> <span class="op">==</span> <span class="st">"Increased"</span><span class="op">)</span></span>
<span><span class="va">max_peaks</span> <span class="op">&lt;-</span> <span class="va">gr_increase</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html" class="external-link">join_overlap_left</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">fit_gr</span>, window <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">fit_gr</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="va">logCPM</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/distinctMC.html">distinctMC</a></span><span class="op">(</span>.keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/colToRanges-methods.html">colToRanges</a></span><span class="op">(</span><span class="st">"window"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Now we’ve found the original windows with the strongest signal, we
can define the profile data using the surrounding regions. Our widest
region with significant change in signal is 5200bp, so let’s set our
profiles to be created stretching 10kb either side of the peak
centre.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getProfileData-methods.html">getProfileData</a></span><span class="op">(</span><span class="va">bwfl</span>, <span class="va">max_peaks</span>, upstream <span class="op">=</span> <span class="fl">1e4</span>, log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">pd</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 2:</span></span>
<span><span class="co">## $Veh</span></span>
<span><span class="co">## GRanges object with 33 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames            ranges strand |</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span></span>
<span><span class="co">##    [1]    chr10 79246860-79266859      * |</span></span>
<span><span class="co">##    [2]    chr10 88335820-88355819      * |</span></span>
<span><span class="co">##    [3]    chr10 79297340-79317339      * |</span></span>
<span><span class="co">##    [4]    chr10 76771460-76791459      * |</span></span>
<span><span class="co">##    [5]    chr10 72997140-73017139      * |</span></span>
<span><span class="co">##    ...      ...               ...    ... .</span></span>
<span><span class="co">##   [29]    chr10 67661660-67681659      * |</span></span>
<span><span class="co">##   [30]    chr10 63687060-63707059      * |</span></span>
<span><span class="co">##   [31]    chr10 79255940-79275939      * |</span></span>
<span><span class="co">##   [32]    chr10 44966940-44986939      * |</span></span>
<span><span class="co">##   [33]    chr10 79254900-79274899      * |</span></span>
<span><span class="co">##                                                 profile_data</span></span>
<span><span class="co">##                                         &lt;SplitDataFrameList&gt;</span></span>
<span><span class="co">##    [1]       1.1206:u1:-9900,0.7100:u2:-9700,1.3750:u3:-9500</span></span>
<span><span class="co">##    [2] 0.879397:u1:-9900,0.665000:u2:-9700,0.075000:u3:-9500</span></span>
<span><span class="co">##    [3]    1.23618:u1:-9900,0.99000:u2:-9700,0.37000:u3:-9500</span></span>
<span><span class="co">##    [4] 0.266332:u1:-9900,0.100000:u2:-9700,0.030000:u3:-9500</span></span>
<span><span class="co">##    [5]          0.375:u1:-9900,0.000:u2:-9700,0.365:u3:-9500</span></span>
<span><span class="co">##    ...                                                   ...</span></span>
<span><span class="co">##   [29]          0.000:u1:-9900,0.290:u2:-9700,0.075:u3:-9500</span></span>
<span><span class="co">##   [30] 0.296482:u1:-9900,0.135000:u2:-9700,0.235000:u3:-9500</span></span>
<span><span class="co">##   [31] 19.89500:u1:-9900,11.28500:u2:-9700, 5.71642:u3:-9500</span></span>
<span><span class="co">##   [32] 0.738693:u1:-9900,1.260000:u2:-9700,2.790000:u3:-9500</span></span>
<span><span class="co">##   [33]        8.590:u1:-9900,13.750:u2:-9700, 9.565:u3:-9500</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $DHT</span></span>
<span><span class="co">## GRanges object with 33 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames            ranges strand |</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span></span>
<span><span class="co">##    [1]    chr10 79246860-79266859      * |</span></span>
<span><span class="co">##    [2]    chr10 88335820-88355819      * |</span></span>
<span><span class="co">##    [3]    chr10 79297340-79317339      * |</span></span>
<span><span class="co">##    [4]    chr10 76771460-76791459      * |</span></span>
<span><span class="co">##    [5]    chr10 72997140-73017139      * |</span></span>
<span><span class="co">##    ...      ...               ...    ... .</span></span>
<span><span class="co">##   [29]    chr10 67661660-67681659      * |</span></span>
<span><span class="co">##   [30]    chr10 63687060-63707059      * |</span></span>
<span><span class="co">##   [31]    chr10 79255940-79275939      * |</span></span>
<span><span class="co">##   [32]    chr10 44966940-44986939      * |</span></span>
<span><span class="co">##   [33]    chr10 79254900-79274899      * |</span></span>
<span><span class="co">##                                                 profile_data</span></span>
<span><span class="co">##                                         &lt;SplitDataFrameList&gt;</span></span>
<span><span class="co">##    [1]    5.36181:u1:-9900,2.61500:u2:-9700,2.20000:u3:-9500</span></span>
<span><span class="co">##    [2]             0.00:u1:-9900,0.95:u2:-9700,0.50:u3:-9500</span></span>
<span><span class="co">##    [3] 0.708543:u1:-9900,1.500000:u2:-9700,0.740000:u3:-9500</span></span>
<span><span class="co">##    [4]             0.00:u1:-9900,0.28:u2:-9700,0.09:u3:-9500</span></span>
<span><span class="co">##    [5]             0.00:u1:-9900,1.05:u2:-9700,0.04:u3:-9500</span></span>
<span><span class="co">##    ...                                                   ...</span></span>
<span><span class="co">##   [29]             0.37:u1:-9900,0.00:u2:-9700,0.03:u3:-9500</span></span>
<span><span class="co">##   [30]             0.00:u1:-9900,0.74:u2:-9700,0.37:u3:-9500</span></span>
<span><span class="co">##   [31]       28.210:u1:-9900,16.725:u2:-9700, 7.270:u3:-9500</span></span>
<span><span class="co">##   [32]    2.88945:u1:-9900,0.56000:u2:-9700,1.73500:u3:-9500</span></span>
<span><span class="co">##   [33]       10.950:u1:-9900,15.230:u2:-9700,16.265:u3:-9500</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>This produces a <code>GRangesList</code> with a <code>GRanges</code>
element for every file in the <code>BigWigFileList</code>, which has the
profile data stored in the final column. Each element of these columns
is a <code>DataFrame</code> with the region broken into a defined number
of bins, and an average coverage value calculated. We can then simply
plot this data by specifying this column in the function
<code><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap()</a></code>, which produces a <code>ggplot2</code>
object able to be customised in the conventional manner. Here, we’ll add
a colour scale and <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw()</a></code></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap</a></span><span class="op">(</span><span class="va">pd</span>, <span class="st">"profile_data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Counts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/profile-heatmap-1.png" alt=""><p class="caption"><em>Profile Heatmap of the regions on chromosome 10
showing evidence for increased H3K27ac signal. Regions are centred at
the point of maximal signal.</em></p>
</div>
<p>Note that H3K27ac often contains regions with significantly lower
signal where the transcriptional machinery is instead occupying the DNA
and as such, H3K27ac peaks are often centred using peaks from an
additional transcription factor. Here, we’ve also just plotted raw
counts. <code>macs2</code> is able to produce BigWigFiles containing
logCPM values, or enrichment over input and these are often used in
preference to raw counts.</p>
</div>
<div class="section level3">
<h3 id="inspection-of-ranges">Inspection of Ranges<a class="anchor" aria-label="anchor" href="#inspection-of-ranges"></a>
</h3>
<p>Another important step in the analysis of ChIP-Seq data is to look at
the binding patterns using coverage, and inspect these in reference to
genes and any other feature of interest. The function
<code><a href="../reference/plotHFGC.html">plotHFGC()</a></code> provides a simple, standardised layout using the
visualisation tools from <code>Gviz</code>. If supplied, tracks will be
drawn in the order 1) HiC; 2) Features; 3) Genes, and 4) Coverage. All
tracks are optional, and if none are provided a simply cytogenetic plot
will be produced. Whilst a simple and intuitive function to use, it also
provides a great deal of flexibility for advanced customisation. All
plots require a <code>GRanges</code> object to define the plotting
region along with a set of cytogenetic bands. These are provided in the
package for GRCh37 and GRCh38.</p>
<p>Let’s start by plotting the first promoter showing changed signal in
<code>results_gr</code> using the minimal data possible. This is a
<code>GRanges</code> object and some cytogenetic bands.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"grch37.cytobands"</span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">hmp_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">gene_name</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="st">"PRXL2A"</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">logical</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">logCPM</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">logCPM</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span><span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-empty-hfgc-1.png" alt=""><p class="caption"><em>Basic output from plotHFGC without any of the
optional tracks</em></p>
</div>
<div class="section level4">
<h4 id="including-coverage">Including Coverage<a class="anchor" aria-label="anchor" href="#including-coverage"></a>
</h4>
<p>In order to show our changed signal in context we can show the
coverage using a <code>BigWigFileList</code> as for the Profile Heatmap.
If providing a <code>BigWigFileList</code>, separate tracks will drawn
for each element of the object. In more complex scenarios where multiple
targets are required a list of <code>BigWigFileList</code>s can be
provided and each element will then be drawn as a track with overlapping
coverage within each BigWigFileList. Colours need to be provided in the
identical structure to match the provided object. Here, we’ve provided a
single <code>BigWigFileList</code> so we can provide a simple list
matching the names within the <code>BigWigFileList.</code></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span></span>
<span>  <span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, </span>
<span>  coverage <span class="op">=</span> <span class="va">bwfl</span>, linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Veh"</span> <span class="op">=</span> <span class="st">"grey30"</span>, <span class="st">"DHT"</span> <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>  zoom <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-hfgc-coverge-1.png" alt=""><p class="caption"><em>Coverage for the region of interest, shown with a
blue highlight. Note how the counts are lower in the highlighted region
for the merged Veh coverage track.</em></p>
</div>
</div>
<div class="section level4">
<h4 id="displaying-genes">Displaying Genes<a class="anchor" aria-label="anchor" href="#displaying-genes"></a>
</h4>
<p>Next we might like to add gene models to provide the regulatory
context. These are supplied here in the layout required by the defaults
of the <code>GeneRegionTrack()</code> function, with all exons and
transcripts annotated.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_models</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html" class="external-link">read_rds</a></span><span class="op">(</span><span class="st">"data/chr10_trans_subset.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span></span>
<span>  <span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, </span>
<span>  coverage <span class="op">=</span> <span class="va">bwfl</span>, linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Veh"</span> <span class="op">=</span> <span class="st">"grey30"</span>, <span class="st">"DHT"</span> <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">gene_models</span>, genecol <span class="op">=</span> <span class="st">"wheat"</span>,</span>
<span>  zoom <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/add-genes-1.png" alt=""><p class="caption"><em>Coverage for our region showing the relationship
of signal to annotated genes</em></p>
</div>
</div>
<div class="section level4">
<h4 id="adding-features">Adding Features<a class="anchor" aria-label="anchor" href="#adding-features"></a>
</h4>
<p>Another useful track to add might be some key features such as
promoters and other annotated regions. Features must
<strong>always</strong> be a <code>GRangesList</code>, with each element
defining a different type of feature, as we already have in our
<code>regions_gr</code> object.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_grl</span> <span class="op">&lt;-</span> <span class="fu">splitAsList</span><span class="op">(</span><span class="va">regions_gr</span>, <span class="va">regions_gr</span><span class="op">$</span><span class="va">region</span><span class="op">)</span><span class="op">[</span><span class="va">region_levels</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">region_col</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">region_levels</span></span>
<span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span></span>
<span>  <span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, </span>
<span>  features <span class="op">=</span> <span class="va">region_grl</span>, featcol <span class="op">=</span> <span class="va">region_col</span>, featstack <span class="op">=</span> <span class="st">"dense"</span>,</span>
<span>  coverage <span class="op">=</span> <span class="va">bwfl</span>, linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Veh"</span> <span class="op">=</span> <span class="st">"grey30"</span>, <span class="st">"DHT"</span> <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">gene_models</span>, gene_col <span class="op">=</span> <span class="st">"wheat"</span>,</span>
<span>  zoom <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-hfgc-1.png" alt=""><p class="caption"><em>The same figure as previously, but with annotated
regions added as features. Any type of feature can be added
here.</em></p>
</div>
</div>
<div class="section level4">
<h4 id="adding-hic-interactions">Adding HiC Interactions<a class="anchor" aria-label="anchor" href="#adding-hic-interactions"></a>
</h4>
<p>If long-range interactions are available, these can also be provided
as a GenomicInteractions object, completing all available options for
the HFGC components.</p>
</div>
<div class="section level4">
<h4 id="adding-annotations-to-coverage">Adding Annotations To Coverage<a class="anchor" aria-label="anchor" href="#adding-annotations-to-coverage"></a>
</h4>
<p>An indication of which regions are associated with increased or
decreased ChIP signal can also be a useful annotation to add to plots
such as the above. Although we technically performed no statistical
testing, let’s consider a window with logFC below -1 to be showing
decreased signal.</p>
<p>Similar to the features track, where the names of
<code>GRangesList</code> elements denote the different feature types,
able to then assigned a colour, coverage annotation tracks follow these
same rules. For each coverage track being annotated, a
<code>GRangesList</code> object can denote the ranges which can be
assigned different colours.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cov_annot</span> <span class="op">&lt;-</span> <span class="fu">splitAsList</span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">results_gr</span><span class="op">$</span><span class="va">status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">endoapply</span><span class="op">(</span><span class="va">granges</span><span class="op">)</span></span></code></pre></div>
<p>In the above, we have Unchanged and Decreased signal denoted as
annotations. In keeping with the approach of having a matching list
element for every coverage track, we would need to pass this as a list
which matched the coverage track</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span></span>
<span>  <span class="va">gr</span>, </span>
<span>  features <span class="op">=</span> <span class="va">region_grl</span>, featcol <span class="op">=</span> <span class="va">region_col</span>, featstack <span class="op">=</span> <span class="st">"dense"</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">gene_models</span>, genecol <span class="op">=</span> <span class="st">"wheat"</span>,</span>
<span>  coverage <span class="op">=</span> <span class="va">bwfl</span>, linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Veh"</span> <span class="op">=</span> <span class="st">"grey30"</span>, <span class="st">"DHT"</span> <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">cov_annot</span>, </span>
<span>  annotcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Decreased <span class="op">=</span> <span class="st">"#3333E6"</span>, Increased <span class="op">=</span> <span class="st">"#E6334D"</span><span class="op">)</span>,</span>
<span>  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_files/figure-gfm/plot-annot-1.png" alt=""><p class="caption"><em>The addition of an annotation track for the
coverage tracks shows which regions were retained during the analysis,
as well as those which were considered as showing changed or unchanged
signal.</em></p>
</div>
<p>Plots are able to be tweaked considerably further via multiple
parameters, however these basic approaches cover the core functionality
of <code>plotHFCG()</code> for enabling simple &amp; reproducible
plotting across regions for multiple sites within a larger
experiment.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-edgeR2016" class="csl-entry">
Chen, Yunshun, Aaron A T Lun, and Gordon K Smyth. 2016. “From Reads to
Genes to Pathways: Differential Expression Analysis of RNA-Seq
Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”
<em>F1000Research</em> 5: 1438. <a href="https://doi.org/10.12688/f1000research.8987.2" class="external-link uri">https://doi.org/10.12688/f1000research.8987.2</a>.
</div>
<div id="ref-Gandolfo2018-oc" class="csl-entry">
Gandolfo, Luke C, and Terence P Speed. 2018. “RLE Plots: Visualizing
Unwanted Variation in High Dimensional Data.” <em>PLoS One</em> 13 (2):
e0191629.
</div>
<div id="ref-Law2014-xq" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom:
Precision Weights Unlock Linear Model Analysis Tools for <span class="nocase">RNA-seq</span> Read Counts.” <em>Genome Biol.</em> 15
(2): R29.
</div>
<div id="ref-DESeq22014" class="csl-entry">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. “Moderated
Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”
<em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link uri">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-csaw2016" class="csl-entry">
Lun, Aaron T L, and Gordon K Smyth. 2016. “Csaw: A Bioconductor Package
for Differential Binding Analysis of ChIP-Seq Data Using Sliding
Windows.” <em>Nucleic Acids Res.</em> 44 (5): e45.
</div>
<div id="ref-Lund2012-xo" class="csl-entry">
Lund, Steven P, Dan Nettleton, Davis J McCarthy, and Gordon K Smyth.
2012. “Detecting Differential Expression in <span class="nocase">RNA-sequence</span> Data Using Quasi-Likelihood with
Shrunken Dispersion Estimates.” <em>Stat. Appl. Genet. Mol. Biol.</em>
11 (5).
</div>
<div id="ref-McCarthy2009-qf" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. “Testing Significance
Relative to a Fold-Change Threshold Is a TREAT.” <em>Bioinformatics</em>
25 (6): 765–71.
</div>
<div id="ref-DiffBind2012" class="csl-entry">
Ross-Innes, Caryn S., Rory Stark, Andrew E. Teschendorff, Kelly A.
Holmes, H. Raza Ali, Mark J. Dunning, Gordon D. Brown, et al. 2012.
“Differential Oestrogen Receptor Binding Is Associated with Clinical
Outcome in Breast Cancer.” <em>Nature</em> 481: –4. <a href="http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html" class="external-link uri">http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html</a>.
</div>
<div id="ref-Wilson2019-ln" class="csl-entry">
Wilson, Daniel J. 2019. “The Harmonic Mean <em>p</em>-Value for
Combining Dependent Tests.” <em>Proc. Natl. Acad. Sci. U. S. A.</em> 116
(4): 1195–1200.
</div>
<div id="ref-Zhang2008-ms" class="csl-entry">
Zhang, Yong, Tao Liu, Clifford A Meyer, Jérôme Eeckhoute, David S
Johnson, Bradley E Bernstein, Chad Nusbaum, et al. 2008. “Model-Based
Analysis of ChIP-Seq (MACS).” <em>Genome Biol.</em> 9 (9): R137.
</div>
</div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.6 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggside_0.2.2                scales_1.2.1               </span></span>
<span><span class="co">##  [3] plyranges_1.18.0            extraChIPs_1.3.10          </span></span>
<span><span class="co">##  [5] patchwork_1.1.2             edgeR_3.40.2               </span></span>
<span><span class="co">##  [7] limma_3.54.1                rtracklayer_1.58.0         </span></span>
<span><span class="co">##  [9] BiocParallel_1.32.5         csaw_1.32.0                </span></span>
<span><span class="co">## [11] SummarizedExperiment_1.28.0 Biobase_2.58.0             </span></span>
<span><span class="co">## [13] MatrixGenerics_1.10.0       matrixStats_0.63.0         </span></span>
<span><span class="co">## [15] Rsamtools_2.14.0            Biostrings_2.66.0          </span></span>
<span><span class="co">## [17] XVector_0.38.0              GenomicRanges_1.50.2       </span></span>
<span><span class="co">## [19] GenomeInfoDb_1.34.9         IRanges_2.32.0             </span></span>
<span><span class="co">## [21] S4Vectors_0.36.1            BiocGenerics_0.44.0        </span></span>
<span><span class="co">## [23] lubridate_1.9.2             forcats_1.0.0              </span></span>
<span><span class="co">## [25] stringr_1.5.0               dplyr_1.1.0                </span></span>
<span><span class="co">## [27] purrr_1.0.1                 readr_2.1.4                </span></span>
<span><span class="co">## [29] tidyr_1.3.0                 tibble_3.1.8               </span></span>
<span><span class="co">## [31] ggplot2_3.4.1               tidyverse_2.0.0            </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] backports_1.4.1            circlize_0.4.15           </span></span>
<span><span class="co">##   [3] Hmisc_4.8-0                BiocFileCache_2.6.1       </span></span>
<span><span class="co">##   [5] igraph_1.4.1               lazyeval_0.2.2            </span></span>
<span><span class="co">##   [7] splines_4.2.3              digest_0.6.31             </span></span>
<span><span class="co">##   [9] ensembldb_2.22.0           foreach_1.5.2             </span></span>
<span><span class="co">##  [11] htmltools_0.5.4            fansi_1.0.4               </span></span>
<span><span class="co">##  [13] magrittr_2.0.3             checkmate_2.1.0           </span></span>
<span><span class="co">##  [15] EnrichedHeatmap_1.27.2     memoise_2.0.1             </span></span>
<span><span class="co">##  [17] BSgenome_1.66.3            cluster_2.1.4             </span></span>
<span><span class="co">##  [19] doParallel_1.0.17          InteractionSet_1.26.1     </span></span>
<span><span class="co">##  [21] tzdb_0.3.0                 ComplexHeatmap_2.14.0     </span></span>
<span><span class="co">##  [23] timechange_0.2.0           prettyunits_1.1.1         </span></span>
<span><span class="co">##  [25] jpeg_0.1-10                colorspace_2.1-0          </span></span>
<span><span class="co">##  [27] ggrepel_0.9.3              blob_1.2.3                </span></span>
<span><span class="co">##  [29] rappdirs_0.3.3             xfun_0.37                 </span></span>
<span><span class="co">##  [31] crayon_1.5.2               RCurl_1.98-1.10           </span></span>
<span><span class="co">##  [33] VariantAnnotation_1.44.1   survival_3.5-5            </span></span>
<span><span class="co">##  [35] iterators_1.0.14           glue_1.6.2                </span></span>
<span><span class="co">##  [37] polyclip_1.10-4            gtable_0.3.1              </span></span>
<span><span class="co">##  [39] zlibbioc_1.44.0            GetoptLong_1.0.5          </span></span>
<span><span class="co">##  [41] DelayedArray_0.24.0        shape_1.4.6               </span></span>
<span><span class="co">##  [43] futile.options_1.0.1       DBI_1.1.3                 </span></span>
<span><span class="co">##  [45] Rcpp_1.0.10                viridisLite_0.4.1         </span></span>
<span><span class="co">##  [47] progress_1.2.2             htmlTable_2.4.1           </span></span>
<span><span class="co">##  [49] clue_0.3-64                foreign_0.8-84            </span></span>
<span><span class="co">##  [51] bit_4.0.5                  Formula_1.2-5             </span></span>
<span><span class="co">##  [53] metapod_1.6.0              htmlwidgets_1.6.1         </span></span>
<span><span class="co">##  [55] httr_1.4.5                 RColorBrewer_1.1-3        </span></span>
<span><span class="co">##  [57] ellipsis_0.3.2             farver_2.1.1              </span></span>
<span><span class="co">##  [59] pkgconfig_2.0.3            XML_3.99-0.13             </span></span>
<span><span class="co">##  [61] Gviz_1.42.1                nnet_7.3-18               </span></span>
<span><span class="co">##  [63] dbplyr_2.3.1               deldir_1.0-6              </span></span>
<span><span class="co">##  [65] locfit_1.5-9.7             utf8_1.2.3                </span></span>
<span><span class="co">##  [67] labeling_0.4.2             tidyselect_1.2.0          </span></span>
<span><span class="co">##  [69] rlang_1.0.6                AnnotationDbi_1.60.0      </span></span>
<span><span class="co">##  [71] munsell_0.5.0              tools_4.2.3               </span></span>
<span><span class="co">##  [73] cachem_1.0.7               cli_3.6.0                 </span></span>
<span><span class="co">##  [75] generics_0.1.3             RSQLite_2.3.0             </span></span>
<span><span class="co">##  [77] broom_1.0.3                evaluate_0.20             </span></span>
<span><span class="co">##  [79] fastmap_1.1.1              yaml_2.3.7                </span></span>
<span><span class="co">##  [81] knitr_1.42                 bit64_4.0.5               </span></span>
<span><span class="co">##  [83] AnnotationFilter_1.22.0    KEGGREST_1.38.0           </span></span>
<span><span class="co">##  [85] formatR_1.14               xml2_1.3.3                </span></span>
<span><span class="co">##  [87] biomaRt_2.54.0             compiler_4.2.3            </span></span>
<span><span class="co">##  [89] rstudioapi_0.14            filelock_1.0.2            </span></span>
<span><span class="co">##  [91] curl_5.0.0                 png_0.1-8                 </span></span>
<span><span class="co">##  [93] tweenr_2.0.2               stringi_1.7.12            </span></span>
<span><span class="co">##  [95] highr_0.10                 futile.logger_1.4.3       </span></span>
<span><span class="co">##  [97] GenomicFeatures_1.50.4     lattice_0.20-45           </span></span>
<span><span class="co">##  [99] ProtGenerics_1.30.0        Matrix_1.5-1              </span></span>
<span><span class="co">## [101] vctrs_0.5.2                pillar_1.8.1              </span></span>
<span><span class="co">## [103] GenomicInteractions_1.32.0 lifecycle_1.0.3           </span></span>
<span><span class="co">## [105] BiocManager_1.30.20        GlobalOptions_0.1.2       </span></span>
<span><span class="co">## [107] ComplexUpset_1.3.3         data.table_1.14.8         </span></span>
<span><span class="co">## [109] bitops_1.0-7               R6_2.5.1                  </span></span>
<span><span class="co">## [111] BiocIO_1.8.0               latticeExtra_0.6-30       </span></span>
<span><span class="co">## [113] gridExtra_2.3              codetools_0.2-19          </span></span>
<span><span class="co">## [115] lambda.r_1.2.4             dichromat_2.0-0.1         </span></span>
<span><span class="co">## [117] MASS_7.3-58.2              rjson_0.2.21              </span></span>
<span><span class="co">## [119] withr_2.5.0                GenomicAlignments_1.34.0  </span></span>
<span><span class="co">## [121] GenomeInfoDbData_1.2.9     parallel_4.2.3            </span></span>
<span><span class="co">## [123] hms_1.1.2                  VennDiagram_1.7.3         </span></span>
<span><span class="co">## [125] grid_4.2.3                 rpart_4.1.19              </span></span>
<span><span class="co">## [127] rmarkdown_2.20             biovizBase_1.46.0         </span></span>
<span><span class="co">## [129] ggforce_0.4.1              base64enc_0.1-3           </span></span>
<span><span class="co">## [131] interp_1.1-3               restfulr_0.0.15</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephen Pederson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
