<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Differential Binding Analysis • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Differential Binding Analysis">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_binding.html">Differential Binding</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/steveped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Differential Binding Analysis</h1>
                        <h4 data-toc-skip class="author">Steve
Pederson</h4>
            <address class="author_afil">
      Dame Roma Mitchell Cancer Research Laboratories, Adelaide Medical
School, University of
Adelaide<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/steveped/extraChIPs/blob/HEAD/vignettes/differential_binding.Rmd" class="external-link"><code>vignettes/differential_binding.Rmd</code></a></small>
      <div class="hidden name"><code>differential_binding.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <a href="https://github.com/steveped/GRAVI" class="external-link">GRAVI</a> workflow,
for which this package is designed, uses sliding windows for
differential binding analysis in a manner similar to the package
<code>csaw</code>, but also incorporating <code>macs2</code> peaks. The
workflow itself extends to integrating multiple ChIP targets and
external data sources, and as such, this package introduces a handful of
functions to enable these analyses.</p>
<p>The majority of examples below use extremely simplified datasets to
provide general guidance on using the functions. Some results may appear
trivial as a result, but will hopefully prove far more useful in a true
experimental context.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>In order to use the package <code>extraChIPs</code> and follow this
vignette, we recommend using the package <code>BiocManager</code> hosted
on CRAN. Once this is installed, the additional packages required for
this vignette (<code>tidyverse</code>, <code>Rsamtools</code>,
<code>csaw</code>, <code>BiocParallel</code> and
<code>rtracklayer</code>) can also be installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="va">pkg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tidyverse"</span>, <span class="st">"Rsamtools"</span>, <span class="st">"csaw"</span>, <span class="st">"BiocParallel"</span>, <span class="st">"rtracklayer"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">pkg</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"extraChIPs"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-binding-and-chip-seq-analysis">Differential Binding and ChIP-Seq Analysis<a class="anchor" aria-label="anchor" href="#differential-binding-and-chip-seq-analysis"></a>
</h2>
<div class="section level3">
<h3 id="sliding-windows">Sliding Windows<a class="anchor" aria-label="anchor" href="#sliding-windows"></a>
</h3>
<p>The starting point for differential binding analysis using sliding
windows is to define windows, then count reads within each window using
the bam files. Commonly one or IP input/control samples is also produced
during a ChIP-Seq experiment. The example files provided here contain a
small subset of reads from chromosome 10 across two experimental and one
input sample.</p>
<p>The approach taken below is to define a set of sliding windows, using
the capabilities of <code>csaw</code>, but to then use <code>macs</code>
peaks to define regions of most likely signal. First we can define our
windows and count the alignments using existing tools. In the following,
we’ll use a sliding window of 180bp and a step size of 60bp, meaning
each nucleotide is covered by 3 windows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csaw</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
<span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata"</span>, <span class="st">"bam"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1.bam"</span>, <span class="st">"ex2.bam"</span>, <span class="st">"input.bam"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"extraChIPs"</span>
<span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1"</span>, <span class="st">"ex2"</span>, <span class="st">"input"</span><span class="op">)</span>
<span class="va">rp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/readParam.html" class="external-link">readParam</a></span><span class="op">(</span>
  pe <span class="op">=</span> <span class="st">"none"</span>,
  dedup <span class="op">=</span> <span class="cn">TRUE</span>,
  restrict <span class="op">=</span> <span class="st">"chr10"</span>
<span class="op">)</span>
<span class="va">wincounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/windowCounts.html" class="external-link">windowCounts</a></span><span class="op">(</span>
  bam.files <span class="op">=</span> <span class="va">bfl</span>,
  spacing <span class="op">=</span> <span class="fl">60</span>,
  width <span class="op">=</span> <span class="fl">180</span>,
  ext <span class="op">=</span> <span class="fl">200</span>,
  filter <span class="op">=</span> <span class="fl">1</span>,
  param <span class="op">=</span> <span class="va">rp</span>
<span class="op">)</span></code></pre></div>
<p>This produces a <code>RangesSummarizedExperiment</code> with windows
included which passed the minimum threshold of 1 total read. As we’ve
only counted reads within a very small window, the complete library
sizes will be highly inaccurate. The true library sizes can be added
here noting that <em>this step is not normally</em> <em>required</em>,
but given these values are essential for accurate CPM values, they will
be added here.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wincounts</span><span class="op">$</span><span class="va">totals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">964076L</span>, <span class="fl">989543L</span>, <span class="fl">1172179L</span><span class="op">)</span></code></pre></div>
<p>We can also add some key information to the <code>colData</code>
element of this object, which will also be propagated to all downstream
objects.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">wincounts</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span>
<span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ctrl"</span>, <span class="st">"treat"</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 3 rows and 6 columns</span>
<span class="co">##                    bam.files    totals       ext      rlen      sample    treat</span>
<span class="co">##                  &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt; &lt;factor&gt;</span>
<span class="co">## ex1   /__w/_temp/Library/e..    964076       200        73         ex1    ctrl </span>
<span class="co">## ex2   /__w/_temp/Library/e..    989543       200        74         ex2    treat</span>
<span class="co">## input /__w/_temp/Library/e..   1172179       200        74       input    NA</span></code></pre>
<p>A density plot can be simply drawn of these counts, with the vast
majority of windows receiving very low counts, due to the nature of
transcription factor binding, where long stretches are unbound. The
windows with higher counts tend to be associated with the samples
targeting a transcription factor (TF), as seen in the two treatment
group samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/steveped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">wincounts</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, trans <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-densities-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="filtering-of-sliding-windows">Filtering of Sliding Windows<a class="anchor" aria-label="anchor" href="#filtering-of-sliding-windows"></a>
</h3>
<p>After counting all reads in the sliding windows, the next step is to
discard windows for which counts are unlikely to represent TF binding.
The package <code>extraChIPs</code> uses a set of consensus peaks to
automatically set a threshold based on 1) counts strongly above the
counts from the input sample, and 2) the windows with the overall
highest signal. Thresholds are determined such that <code>q = 0.5</code>
of the retained windows overlap on of the supplied consensus peaks.
Higher values for <code>q</code> will return more windows, however many
of these will tend to only marginally overlap a peak. Experience has
shown that values such as <code>q = 0.5</code> tend to return a
considerable proportion of windows containing true TF binding
signal.</p>
<p>First we can load the peaks, supplied here as a simple
<code>bed</code> file.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"peaks.bed.gz"</span>, package <span class="op">=</span> <span class="st">"extraChIPs"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></code></pre></div>
<p>The we can pass these to the function <code><a href="../reference/dualFilter.html">dualFilter()</a></code> which
utilises the strategy described above. On large datasets, this can be
quite time-consuming, as can the initial counting step. Due to the small
example dataset, a more inclusive threshold for <code>q</code> will be
used here.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filtcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dualFilter.html">dualFilter</a></span><span class="op">(</span>
  x <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span><span class="op">)</span><span class="op">]</span>,
  bg <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">$</span><span class="va">treat</span><span class="op">)</span><span class="op">]</span>, 
  ref <span class="op">=</span> <span class="va">peaks</span>,
  q <span class="op">=</span> <span class="fl">0.8</span> <span class="co"># Better to use q = 0.5 on real data</span>
<span class="op">)</span></code></pre></div>
<p>The returned object will by default contain <code>counts</code> and
<code>logCPM</code> assays, with the complete library sizes used for the
calculation of <code>logCPM</code> values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plotcpm-1.png" width="960"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotAssayPCA-methods.html">plotAssayPCA</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, label <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plotcpm-2.png" width="960"></p>
<p>Whilst the initial set of counts contained 1007 windows, these have
now been reduced to 108 windows. Similarly, the input sample is no
longer included in the data object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 1007    3</span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 108   2</span></code></pre>
<p>The <code>rowData</code> element of the returned object will contain
a logical column indicating where each specific retained window
overlapped one of the supplied consensus peaks.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 108 ranges and 1 metadata column:</span>
<span class="co">##         seqnames              ranges strand | overlaps_ref</span>
<span class="co">##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;logical&gt;</span>
<span class="co">##     [1]    chr10 103865521-103865700      * |         TRUE</span>
<span class="co">##     [2]    chr10 103865581-103865760      * |         TRUE</span>
<span class="co">##     [3]    chr10 103865641-103865820      * |         TRUE</span>
<span class="co">##     [4]    chr10 103865701-103865880      * |         TRUE</span>
<span class="co">##     [5]    chr10 103874161-103874340      * |        FALSE</span>
<span class="co">##     ...      ...                 ...    ... .          ...</span>
<span class="co">##   [104]    chr10 103911901-103912080      * |        FALSE</span>
<span class="co">##   [105]    chr10 103911961-103912140      * |        FALSE</span>
<span class="co">##   [106]    chr10 103912021-103912200      * |        FALSE</span>
<span class="co">##   [107]    chr10 103912081-103912260      * |        FALSE</span>
<span class="co">##   [108]    chr10 103912261-103912440      * |        FALSE</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">overlaps_ref</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 0.7314815</span></code></pre>
</div>
<div class="section level3">
<h3 id="using-voom">Using Voom<a class="anchor" aria-label="anchor" href="#using-voom"></a>
</h3>
<p>Multiple approaches are available for analysis of differential
binding, and given the small example dataset, only a brief example of
conventional results will be used. <code>extraChIPs</code> does provide
a simple coercion function to covert <code>logCPM</code> to a
<code>voom</code> object, which requires the relationship between
library sizes and <code>logCPM</code> values to be intact. Whist this
will not be discussed further here should this be a viable approach for
an analysis, the following code may prove helpful.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voomWeightsFromCPM.html">voomWeightsFromCPM</a></span><span class="op">(</span>
  cpm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>, <span class="st">"logCPM"</span><span class="op">)</span>, 
  lib.size <span class="op">=</span> <span class="va">filtcounts</span><span class="op">$</span><span class="va">totals</span>, 
  isLogCPM <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-windows">Merging Windows<a class="anchor" aria-label="anchor" href="#merging-windows"></a>
</h3>
<p>After an analysis has been performed, common values contained in the
output may be estimated signal (<code>logCPM</code>), estimated change
(<code>logFC</code>) with both raw and adjusted p-values. Given the
dependency of neighbouring windows, any adjusted p-values will not be
appropriate and a merging of overlapping windows will be performed.</p>
<p>For our example dataset we’ll add these manually, however this is
just for demonstration purposes for the eventual merging of windows.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">logCPM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>,<span class="st">"logCPM"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;-</span> <span class="fu">rowDiffs</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>,<span class="st">"logCPM"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">PValue</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Chisquare.html" class="external-link">pchisq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">$</span><span class="va">logFC</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Now we have some example values, we can merge any overlapping windows
using <code><a href="../reference/mergeByCol-methods.html">mergeByCol()</a></code>. During this process, overlapping ranges
are merged into a single range with representative values taken from one
of the initial sliding windows. The recommended approach for retaining
statistical independence between windows is to choose the window with
the largest signal as representative of the entire merged window.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeByCol-methods.html">mergeByCol</a></span><span class="op">(</span><span class="va">filtcounts</span>, col <span class="op">=</span> <span class="st">"logCPM"</span>, pval <span class="op">=</span> <span class="st">"PValue"</span><span class="op">)</span>
<span class="va">res_gr</span><span class="op">$</span><span class="va">overlaps_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">overlapsAny</a></span><span class="op">(</span><span class="va">res_gr</span>, <span class="va">peaks</span><span class="op">)</span></code></pre></div>
<p>A <code>GRanges</code> object is returned with representative values
for each merged window. The <code>mcol</code> <code>keyval_range</code>
provides the original range from which the representative values were
taken. A column with adjusted p-values will also be added if
<code>p_adj_method</code> is not set to “none”.</p>
</div>
<div class="section level3">
<h3 id="mapping-of-windows-to-genes">Mapping of Windows To Genes<a class="anchor" aria-label="anchor" href="#mapping-of-windows-to-genes"></a>
</h3>
<p>Once the binding characteristics of a transcription factor have been
characterised, a common next step is to assess which genes are likely to
be under regulatory influence. Whilst no definitive, single methodology
exists for this process, the function <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> offers
an intuitive approach, taking into account any defined regulatory
features. These regulatory features may be defined by simple proximity
to TSS regions, by histone marks, downloaded from external repositories
or any other possibility. Whilst these features can improve the
precision of mapping, even without these this function can still enable
a useful assignment of target gene to binding event.</p>
<p>The process undertaken inside <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> is a
sequential checking of each range’s association with regulatory features
and the most likely target as a result. These steps are:</p>
<ol style="list-style-type: decimal">
<li><strong>Check for any HiC interactions</strong></li>
</ol>
<ul>
<li>All genes which directly overlap an interaction anchor are
considered part of the regulatory network for that interaction, and as
such, all genes associated with both anchors are assigned to a peak
which overlaps a HiC Interaction</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><strong>Check for any overlaps with a promoter</strong></li>
</ol>
<ul>
<li>All genes regulated by that promoter are assigned as regulatory
targets. By default, this is by direct promoter/gene overlap
(<code>prom2gene = 0</code>)</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li><strong>Check for any overlaps with an enhancer</strong></li>
</ol>
<ul>
<li>Peaks which overlap an enhancer are assigned to <em>all</em> genes
within the distance specified by <code>enh2gene</code> (default =
100kb)</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li><strong>Check for genes with no previous mappings</strong></li>
</ol>
<ul>
<li>Peaks <em>with no previous mappings</em> are assigned to all
directly overlapping genes, or the nearest gene within a specified
distance (default <code>gr2gene</code> = 100kb)</li>
</ul>
<p>As a result, if no promoters, enhancers or interactions are supplied,
all genes will be mapped to peaks using step 4</p>
<p>The two essential data objects to perform simple gene assignment are
1) a set of ranges representing binding events of interest, such as
<code>res_gr</code> above, and 2) a set of ranges defining genes, as
contained in the example dataset <code>ex_genes</code>. This contains
the two <code>mcols</code> <em>gene</em> and <em>symbol</em>, and we can
ask for both in the returned object.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_genes"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_prom"</span><span class="op">)</span>
<span class="fu"><a href="../reference/mapByFeature.html">mapByFeature</a></span><span class="op">(</span>
  <span class="va">res_gr</span>, genes <span class="op">=</span> <span class="va">ex_genes</span>, prom <span class="op">=</span> <span class="va">ex_prom</span>, cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"symbol"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 9 ranges and 11 metadata columns:</span>
<span class="co">##       seqnames              ranges strand | n_windows      n_up    n_down</span>
<span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">##   [1]    chr10 103865521-103865880      * |         4         0         3</span>
<span class="co">##   [2]    chr10 103874161-103874940      * |         8         1         6</span>
<span class="co">##   [3]    chr10 103876921-103878000      * |        16         0         6</span>
<span class="co">##   [4]    chr10 103878601-103878840      * |         2         1         1</span>
<span class="co">##   [5]    chr10 103878961-103881060      * |        33         0        17</span>
<span class="co">##   [6]    chr10 103881181-103881540      * |         4         1         2</span>
<span class="co">##   [7]    chr10 103892161-103893420      * |        19         5        12</span>
<span class="co">##   [8]    chr10 103899841-103900560      * |        10         2         5</span>
<span class="co">##   [9]    chr10 103911481-103912440      * |        12         3         7</span>
<span class="co">##                    keyval_range    logCPM      logFC    PValue PValue_fdr</span>
<span class="co">##                       &lt;GRanges&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">##   [1] chr10:103865521-103865700   3.46979 -0.5610608  0.574756   0.975993</span>
<span class="co">##   [2] chr10:103874641-103874820   4.24295 -0.1115489  0.911181   0.975993</span>
<span class="co">##   [3] chr10:103877281-103877460   5.61526 -1.4816426  0.138435   0.975993</span>
<span class="co">##   [4] chr10:103878661-103878840   3.35589 -0.0300928  0.975993   0.975993</span>
<span class="co">##   [5] chr10:103880281-103880460   7.72879 -0.6250501  0.531938   0.975993</span>
<span class="co">##   [6] chr10:103881241-103881420   3.73440 -0.0318288  0.974609   0.975993</span>
<span class="co">##   [7] chr10:103892581-103892760   7.40460 -0.0633132  0.949517   0.975993</span>
<span class="co">##   [8] chr10:103900141-103900320   4.64587  0.1423445  0.886808   0.975993</span>
<span class="co">##   [9] chr10:103911901-103912080   4.81325 -0.1924133  0.847419   0.975993</span>
<span class="co">##       overlaps_ref            gene          symbol</span>
<span class="co">##          &lt;logical&gt; &lt;CharacterList&gt; &lt;CharacterList&gt;</span>
<span class="co">##   [1]         TRUE ENSG00000198728            LDB1</span>
<span class="co">##   [2]        FALSE ENSG00000198728            LDB1</span>
<span class="co">##   [3]         TRUE ENSG00000198728            LDB1</span>
<span class="co">##   [4]        FALSE ENSG00000198728            LDB1</span>
<span class="co">##   [5]         TRUE ENSG00000198728            LDB1</span>
<span class="co">##   [6]         TRUE ENSG00000198728            LDB1</span>
<span class="co">##   [7]         TRUE ENSG00000148840           PPRC1</span>
<span class="co">##   [8]         TRUE ENSG00000148840           PPRC1</span>
<span class="co">##   [9]        FALSE ENSG00000166197           NOLC1</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></code></pre>
<p>For this dataset, we have an example HiC interaction, which we can
now pass to the mapping process. (This time we’ll save the object)</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_hic"</span><span class="op">)</span>
<span class="va">res_gr_mapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapByFeature.html">mapByFeature</a></span><span class="op">(</span>
  <span class="va">res_gr</span>, 
  genes <span class="op">=</span> <span class="va">ex_genes</span>, 
  prom <span class="op">=</span> <span class="va">ex_prom</span>,
  gi <span class="op">=</span> <span class="va">ex_hic</span>, 
  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"symbol"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">res_gr_mapped</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 9 ranges and 11 metadata columns:</span>
<span class="co">##       seqnames              ranges strand | n_windows      n_up    n_down</span>
<span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">##   [1]    chr10 103865521-103865880      * |         4         0         3</span>
<span class="co">##   [2]    chr10 103874161-103874940      * |         8         1         6</span>
<span class="co">##   [3]    chr10 103876921-103878000      * |        16         0         6</span>
<span class="co">##   [4]    chr10 103878601-103878840      * |         2         1         1</span>
<span class="co">##   [5]    chr10 103878961-103881060      * |        33         0        17</span>
<span class="co">##   [6]    chr10 103881181-103881540      * |         4         1         2</span>
<span class="co">##   [7]    chr10 103892161-103893420      * |        19         5        12</span>
<span class="co">##   [8]    chr10 103899841-103900560      * |        10         2         5</span>
<span class="co">##   [9]    chr10 103911481-103912440      * |        12         3         7</span>
<span class="co">##                    keyval_range    logCPM      logFC    PValue PValue_fdr</span>
<span class="co">##                       &lt;GRanges&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">##   [1] chr10:103865521-103865700   3.46979 -0.5610608  0.574756   0.975993</span>
<span class="co">##   [2] chr10:103874641-103874820   4.24295 -0.1115489  0.911181   0.975993</span>
<span class="co">##   [3] chr10:103877281-103877460   5.61526 -1.4816426  0.138435   0.975993</span>
<span class="co">##   [4] chr10:103878661-103878840   3.35589 -0.0300928  0.975993   0.975993</span>
<span class="co">##   [5] chr10:103880281-103880460   7.72879 -0.6250501  0.531938   0.975993</span>
<span class="co">##   [6] chr10:103881241-103881420   3.73440 -0.0318288  0.974609   0.975993</span>
<span class="co">##   [7] chr10:103892581-103892760   7.40460 -0.0633132  0.949517   0.975993</span>
<span class="co">##   [8] chr10:103900141-103900320   4.64587  0.1423445  0.886808   0.975993</span>
<span class="co">##   [9] chr10:103911901-103912080   4.81325 -0.1924133  0.847419   0.975993</span>
<span class="co">##       overlaps_ref                            gene          symbol</span>
<span class="co">##          &lt;logical&gt;                 &lt;CharacterList&gt; &lt;CharacterList&gt;</span>
<span class="co">##   [1]         TRUE                 ENSG00000198728            LDB1</span>
<span class="co">##   [2]        FALSE                 ENSG00000198728            LDB1</span>
<span class="co">##   [3]         TRUE                 ENSG00000198728            LDB1</span>
<span class="co">##   [4]        FALSE                 ENSG00000198728            LDB1</span>
<span class="co">##   [5]         TRUE ENSG00000198728,ENSG00000148840      LDB1,PPRC1</span>
<span class="co">##   [6]         TRUE ENSG00000198728,ENSG00000148840      LDB1,PPRC1</span>
<span class="co">##   [7]         TRUE ENSG00000148840,ENSG00000198728      PPRC1,LDB1</span>
<span class="co">##   [8]         TRUE                 ENSG00000148840           PPRC1</span>
<span class="co">##   [9]        FALSE                 ENSG00000166197           NOLC1</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></code></pre>
<p>The 5<sup>th</sup> to 7<sup>th</sup> windows are now mapped to both
<em>LDB1</em> and <em>PPRC1</em>, whereas previously these windows were
only mapped to <em>LDB1</em>.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualisation-of-results">Visualisation of Results<a class="anchor" aria-label="anchor" href="#visualisation-of-results"></a>
</h2>
<div class="section level3">
<h3 id="association-with-features">Association with Features<a class="anchor" aria-label="anchor" href="#association-with-features"></a>
</h3>
<p>The association of windows or peaks with defined features, such as
histone marks or regulatory elements can be important for describing the
binding characteristics of any given transcription factor. We have
already defined the association of the merged windows with consensus
peaks identified by <code>macs2</code>. We can easily visualise these
using <code><a href="../reference/plotPie-methods.html">plotPie()</a></code></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="../reference/plotPie-methods.html">plotPie</a></span><span class="op">(</span><span class="st">"overlaps_ref"</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-pie-1.png" width="960"></p>
<p>These distribution charts can be drawn across three separate
categories. Let’s include promoters.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_gr</span><span class="op">$</span><span class="va">Feature</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bestOverlap-methods.html">bestOverlap</a></span><span class="op">(</span>
  <span class="va">res_gr</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span>Promoter <span class="op">=</span> <span class="va">ex_prom</span><span class="op">)</span>, missing <span class="op">=</span> <span class="st">"None"</span>
<span class="op">)</span>
<span class="va">res_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="../reference/plotPie-methods.html">plotPie</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Feature"</span>, fill <span class="op">=</span> <span class="st">"overlaps_ref"</span><span class="op">)</span> </code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-dual-pie-1.png" width="960"></p>
<p>In a real world context where we’re dealing with thousands of ranges
and multiple features, this represents a quick an easy way to asses
binding characteristics. As these are all <code>ggplot2</code> objects,
they can be easily customised using <code>theme</code> and
<code>scale_fill_*</code> capabilities.</p>
</div>
<div class="section level3">
<h3 id="profile-heatmaps">Profile Heatmaps<a class="anchor" aria-label="anchor" href="#profile-heatmaps"></a>
</h3>
<p>A very common approach to visualising the results of altered TF
binding is to plot <em>profile heatmaps</em> centred around the window
(or peak), and extending out a given number of of bases. The data
required for this is referred to in <code>extraChIPs</code> as profile
data, and given that extracting this from a set of
<code>BigWigFile</code>s can be time consuming, this step is performed
prior to the actual plotting, so that ranges can be added or excluded as
desired.</p>
<p>First we need to define a <code>BigWigFileList</code> as these are
conventionally very large files which shouldn’t be retained in memory,
but are just accessed to import the key regions for a particular
process.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bwfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata"</span>, <span class="st">"bigwig"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1.bw"</span>, <span class="st">"ex2.bw"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"extraChIPs"</span>
<span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFileList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ex1"</span>, <span class="st">"ex2"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now we have our <code>BigWigFileList</code> we can define the profile
data</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getProfileData-methods.html">getProfileData</a></span><span class="op">(</span><span class="va">bwfl</span>, <span class="va">res_gr</span><span class="op">)</span>
<span class="va">pd</span></code></pre></div>
<pre><code><span class="co">## GRangesList object of length 2:</span>
<span class="co">## $ex1</span>
<span class="co">## GRanges object with 9 ranges and 11 metadata columns:</span>
<span class="co">##       seqnames              ranges strand | n_windows      n_up    n_down</span>
<span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">##   [1]    chr10 103863200-103868199      * |         4         0         3</span>
<span class="co">##   [2]    chr10 103872050-103877049      * |         8         1         6</span>
<span class="co">##   [3]    chr10 103874960-103879959      * |        16         0         6</span>
<span class="co">##   [4]    chr10 103876220-103881219      * |         2         1         1</span>
<span class="co">##   [5]    chr10 103877510-103882509      * |        33         0        17</span>
<span class="co">##   [6]    chr10 103878860-103883859      * |         4         1         2</span>
<span class="co">##   [7]    chr10 103890290-103895289      * |        19         5        12</span>
<span class="co">##   [8]    chr10 103897700-103902699      * |        10         2         5</span>
<span class="co">##   [9]    chr10 103909460-103914459      * |        12         3         7</span>
<span class="co">##                    keyval_range    logCPM      logFC    PValue PValue_fdr</span>
<span class="co">##                       &lt;GRanges&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">##   [1] chr10:103865521-103865700   3.46979 -0.5610608  0.574756   0.975993</span>
<span class="co">##   [2] chr10:103874641-103874820   4.24295 -0.1115489  0.911181   0.975993</span>
<span class="co">##   [3] chr10:103877281-103877460   5.61526 -1.4816426  0.138435   0.975993</span>
<span class="co">##   [4] chr10:103878661-103878840   3.35589 -0.0300928  0.975993   0.975993</span>
<span class="co">##   [5] chr10:103880281-103880460   7.72879 -0.6250501  0.531938   0.975993</span>
<span class="co">##   [6] chr10:103881241-103881420   3.73440 -0.0318288  0.974609   0.975993</span>
<span class="co">##   [7] chr10:103892581-103892760   7.40460 -0.0633132  0.949517   0.975993</span>
<span class="co">##   [8] chr10:103900141-103900320   4.64587  0.1423445  0.886808   0.975993</span>
<span class="co">##   [9] chr10:103911901-103912080   4.81325 -0.1924133  0.847419   0.975993</span>
<span class="co">##       overlaps_ref     Feature</span>
<span class="co">##          &lt;logical&gt; &lt;character&gt;</span>
<span class="co">##   [1]         TRUE        None</span>
<span class="co">##   [2]        FALSE    Promoter</span>
<span class="co">##   [3]         TRUE        None</span>
<span class="co">##   [4]        FALSE        None</span>
<span class="co">##   [5]         TRUE    Promoter</span>
<span class="co">##   [6]         TRUE    Promoter</span>
<span class="co">##   [7]         TRUE    Promoter</span>
<span class="co">##   [8]         TRUE    Promoter</span>
<span class="co">##   [9]        FALSE    Promoter</span>
<span class="co">##                                                profile_data</span>
<span class="co">##                                        &lt;SplitDataFrameList&gt;</span>
<span class="co">##   [1]    1.00260:u1:-2475,1.00260:u2:-2425,1.10286:u3:-2375</span>
<span class="co">##   [2]    2.65996:u1:-2475,3.64946:u2:-2425,3.68957:u3:-2375</span>
<span class="co">##   [3]    8.64240:u1:-2475,6.81767:u2:-2425,5.35387:u3:-2375</span>
<span class="co">##   [4]    7.17861:u1:-2475,7.65985:u2:-2425,6.29632:u3:-2375</span>
<span class="co">##   [5]    26.9719:u1:-2475,20.3929:u2:-2425,13.8760:u3:-2375</span>
<span class="co">##   [6]    4.21092:u1:-2475,4.39138:u2:-2425,3.56925:u3:-2375</span>
<span class="co">##   [7]    4.01040:u1:-2475,3.36874:u2:-2425,3.44894:u3:-2375</span>
<span class="co">##   [8]    3.47841:u1:-2475,2.04530:u2:-2425,2.00520:u3:-2375</span>
<span class="co">##   [9] 0.306918:u1:-2475,1.283328:u2:-2425,2.005200:u3:-2375</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome</span>
<span class="co">## </span>
<span class="co">## $ex2</span>
<span class="co">## GRanges object with 9 ranges and 11 metadata columns:</span>
<span class="co">##       seqnames              ranges strand | n_windows      n_up    n_down</span>
<span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">##   [1]    chr10 103863200-103868199      * |         4         0         3</span>
<span class="co">##   [2]    chr10 103872050-103877049      * |         8         1         6</span>
<span class="co">##   [3]    chr10 103874960-103879959      * |        16         0         6</span>
<span class="co">##   [4]    chr10 103876220-103881219      * |         2         1         1</span>
<span class="co">##   [5]    chr10 103877510-103882509      * |        33         0        17</span>
<span class="co">##   [6]    chr10 103878860-103883859      * |         4         1         2</span>
<span class="co">##   [7]    chr10 103890290-103895289      * |        19         5        12</span>
<span class="co">##   [8]    chr10 103897700-103902699      * |        10         2         5</span>
<span class="co">##   [9]    chr10 103909460-103914459      * |        12         3         7</span>
<span class="co">##                    keyval_range    logCPM      logFC    PValue PValue_fdr</span>
<span class="co">##                       &lt;GRanges&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">##   [1] chr10:103865521-103865700   3.46979 -0.5610608  0.574756   0.975993</span>
<span class="co">##   [2] chr10:103874641-103874820   4.24295 -0.1115489  0.911181   0.975993</span>
<span class="co">##   [3] chr10:103877281-103877460   5.61526 -1.4816426  0.138435   0.975993</span>
<span class="co">##   [4] chr10:103878661-103878840   3.35589 -0.0300928  0.975993   0.975993</span>
<span class="co">##   [5] chr10:103880281-103880460   7.72879 -0.6250501  0.531938   0.975993</span>
<span class="co">##   [6] chr10:103881241-103881420   3.73440 -0.0318288  0.974609   0.975993</span>
<span class="co">##   [7] chr10:103892581-103892760   7.40460 -0.0633132  0.949517   0.975993</span>
<span class="co">##   [8] chr10:103900141-103900320   4.64587  0.1423445  0.886808   0.975993</span>
<span class="co">##   [9] chr10:103911901-103912080   4.81325 -0.1924133  0.847419   0.975993</span>
<span class="co">##       overlaps_ref     Feature</span>
<span class="co">##          &lt;logical&gt; &lt;character&gt;</span>
<span class="co">##   [1]         TRUE        None</span>
<span class="co">##   [2]        FALSE    Promoter</span>
<span class="co">##   [3]         TRUE        None</span>
<span class="co">##   [4]        FALSE        None</span>
<span class="co">##   [5]         TRUE    Promoter</span>
<span class="co">##   [6]         TRUE    Promoter</span>
<span class="co">##   [7]         TRUE    Promoter</span>
<span class="co">##   [8]         TRUE    Promoter</span>
<span class="co">##   [9]        FALSE    Promoter</span>
<span class="co">##                                                profile_data</span>
<span class="co">##                                        &lt;SplitDataFrameList&gt;</span>
<span class="co">##   [1]    1.02886:u1:-2475,1.02886:u2:-2425,1.02886:u3:-2375</span>
<span class="co">##   [2]    3.80047:u1:-2475,5.32947:u2:-2425,7.20199:u3:-2375</span>
<span class="co">##   [3]    6.79045:u1:-2475,4.13600:u2:-2425,3.88907:u3:-2375</span>
<span class="co">##   [4]    5.53524:u1:-2475,5.80274:u2:-2425,5.57640:u3:-2375</span>
<span class="co">##   [5]    9.58893:u1:-2475,8.53950:u2:-2425,6.35833:u3:-2375</span>
<span class="co">##   [6]    6.21428:u1:-2475,5.04139:u2:-2425,5.32947:u3:-2375</span>
<span class="co">##   [7]    2.05771:u1:-2475,1.56386:u2:-2425,2.05771:u3:-2375</span>
<span class="co">##   [8] 0.629914:u1:-2475,0.000000:u2:-2425,0.000000:u3:-2375</span>
<span class="co">##   [9]    4.11542:u1:-2475,3.25119:u2:-2425,3.08657:u3:-2375</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></code></pre>
<p>This produces a <code>GRangesList</code> with a <code>GRanges</code>
element for every file in the <code>BigWigFileList</code>, which has the
profile data stored in the final column. Each element of these columns
is a <code>DataFrame</code> with the region broken into a defined number
of bins, and an average coverage value calculated. We can then simply
plot this data by specifying this column in the function
<code><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap()</a></code>, which produces a <code>ggplot2</code>
object able to be customised in the conventional manner. Here, we’ll add
a colour scale and <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw()</a></code></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap</a></span><span class="op">(</span><span class="va">pd</span>, <span class="st">"profile_data"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"CPM"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/profile-heatmap-1.png" width="960"></p>
<p>In our initial merging of sliding windows we chose our representative
values to be from the sliding window with the highest signal. This may
not be at the centre of the final merged window, but having retained
this in the <code>keyval_range</code> column, we can use this range for
generation of the profile data, ensuring we have our profile heatmaps
centred at the point of the highest signal.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getProfileData-methods.html">getProfileData</a></span><span class="op">(</span><span class="va">bwfl</span>, <span class="fu"><a href="../reference/colToRanges-methods.html">colToRanges</a></span><span class="op">(</span><span class="va">res_gr</span>, <span class="st">"keyval_range"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap</a></span><span class="op">(</span><span class="va">pd</span>, <span class="st">"profile_data"</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"CPM"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/centred-heatmap-1.png" width="960"></p>
<p>As we’re using <code>ggplot2</code> we can also separate peaks by any
of the categorical columns in our initial ranges, such as the
<code>overlaps_ref</code> column. This will not only create facets along
the y-axis, but the traces for each panel are drawn separately for each
facet, and these can be simply assigned colours or linetype using
standard <code>ggplot2</code> syntax.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap</a></span><span class="op">(</span>
  <span class="va">pd</span>, <span class="st">"profile_data"</span>, facetY <span class="op">=</span> <span class="st">"overlaps_ref"</span>, linetype <span class="op">=</span> <span class="st">"overlaps_ref"</span>
<span class="op">)</span>  <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"CPM"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/facet-heatmap-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="inspection-of-ranges">Inspection of Ranges<a class="anchor" aria-label="anchor" href="#inspection-of-ranges"></a>
</h3>
<p>Another important step in the analysis of ChIP-Seq data is to look at
the binding patterns using coverage, and inspect these in reference to
genes and any other feature of interest. The function
<code><a href="../reference/plotHFGC.html">plotHFGC()</a></code> provides a simple, standardised layout using the
visualisation tools from <code>Gviz</code>. If supplied, tracks will be
drawn in the order 1) HiC; 2) Features; 3) Genes, and 4) Coverage.
Whilst a simple and intuitive function to use, it also provides a great
deal of flexibility for advanced customisation. All plots require a
<code>GRanges</code> object to define the plotting region, with all
other tracks being optional.</p>
<div class="section level4">
<h4 id="displaying-genes">Displaying Genes<a class="anchor" aria-label="anchor" href="#displaying-genes"></a>
</h4>
<p>Let’s start by plotting the entire region contained in
<code>res_gr</code> using the minimal data possible, a
<code>GRanges</code> object and some cytogenetic bands.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"grch37.cytobands"</span><span class="op">)</span>
<span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html" class="external-link">range</a></span><span class="op">(</span><span class="va">res_gr</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span><span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-empty-hfgc-1.png" width="960"></p>
<p>This is clearly of minimal interest, so let’s add some transcript
models. These are supplied here in the layout required by the defaults
of the <code>GeneRegionTrack()</code> function, with all exons and
transcripts annotated.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_trans"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span><span class="va">gr</span>, genes <span class="op">=</span> <span class="va">ex_trans</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span><span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/add-genes-1.png" width="960"></p>
<p>As these are collapsed into <em>meta-transcripts</em> by default,
let’s 1) add colour, 2) expand transcripts, and 3) zoom out a little.
The initial range is highlighted by default, but this can also be turned
off if preferred.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  genes <span class="op">=</span> <span class="va">ex_trans</span>, genecol <span class="op">=</span> <span class="st">"wheat"</span>,
  collapseTranscripts <span class="op">=</span> <span class="cn">FALSE</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-trans-1.png" width="960"></p>
<p>The object <code>ex_trans</code> contains the column
<code>status</code>, and we might like to use this to display these
genes on separate tracks. In this case, we would pass a
<code>GRangesList</code> to the <code>genes</code> argument, and each
element within that list will be drawn as a separate track.</p>
<ul>
<li>Colours should be provided as a <em>named</em> list with on element
for each element of the genes <code>GRangesList</code>, or as a single
colour</li>
<li>
<code>collapseTranscripts</code> can also be provided as a matching
(`<em>named</em>) list with each element being applied to the respective
track, or as a single value</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">status_trans</span> <span class="op">&lt;-</span> <span class="fu">splitAsList</span><span class="op">(</span><span class="va">ex_trans</span>, <span class="va">ex_trans</span><span class="op">$</span><span class="va">status</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  genes <span class="op">=</span> <span class="va">status_trans</span>, 
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="st">"forestgreen"</span>, Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Undetected <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span>,
  collapseTranscripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="cn">FALSE</span>, Unchanged <span class="op">=</span> <span class="cn">FALSE</span>, Undetected <span class="op">=</span> <span class="st">"meta"</span><span class="op">)</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/split-trans-1.png" width="960"></p>
<p>This idea of providing a matching <em>named</em> list is applied
across the genes, features and coverage tracks in the sections
below.</p>
</div>
<div class="section level4">
<h4 id="adding-features">Adding Features<a class="anchor" aria-label="anchor" href="#adding-features"></a>
</h4>
<p>Another useful track to add might be some key features such as
promoters. Unlike the genes track, features must <strong>always</strong>
be a <code>GRangesList</code>, with each element defining a different
type of feature. Given that we only have promoters, we’ll still need to
set this up as a <code>GRangesList</code></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ex_prom"</span><span class="op">)</span>
<span class="va">feat_grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span>Promoters <span class="op">=</span> <span class="va">ex_prom</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  features <span class="op">=</span> <span class="va">feat_grl</span>, featcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Promoters <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,
  genes <span class="op">=</span> <span class="va">status_trans</span>, 
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="st">"forestgreen"</span>, Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Undetected <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span>,
  collapseTranscripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="cn">FALSE</span>, Unchanged <span class="op">=</span> <span class="cn">FALSE</span>, Undetected <span class="op">=</span> <span class="st">"meta"</span><span class="op">)</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-hfgc-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="adding-hic-interactions">Adding HiC Interactions<a class="anchor" aria-label="anchor" href="#adding-hic-interactions"></a>
</h4>
<p>Adding the HiC Interactions becomes very simple. All that we need is
a GInteractions object.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  hic <span class="op">=</span> <span class="va">ex_hic</span>,
  features <span class="op">=</span> <span class="va">feat_grl</span>, featcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Promoters <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,
  genes <span class="op">=</span> <span class="va">status_trans</span>, 
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="st">"forestgreen"</span>, Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Undetected <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span>,
  collapseTranscripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="cn">FALSE</span>, Unchanged <span class="op">=</span> <span class="cn">FALSE</span>, Undetected <span class="op">=</span> <span class="st">"meta"</span><span class="op">)</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-with-hic-1.png" width="960"></p>
<p>If interactions extend beyond the plot range (<code>gr</code>), the
plotting range will be automatically extended to incorporate all
interactions. Given these can extend to a very long distance, only
interactions within 10Mb are included by default. This can be modified
using the <code>max</code> argument.</p>
</div>
<div class="section level4">
<h4 id="adding-peakscoverage">Adding Peaks/Coverage<a class="anchor" aria-label="anchor" href="#adding-peakscoverage"></a>
</h4>
<p>The simplest approach to adding coverage is to simply provide a
single <code>BigWigFileList</code>. In this scenario, each individual
file will be drawn on a separate track. Colours for lines are passed as
a simple vector/list with names matching the names of the
<code>BigWigFileList</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  hic <span class="op">=</span> <span class="va">ex_hic</span>,
  features <span class="op">=</span> <span class="va">feat_grl</span>, featcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Promoters <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,
  genes <span class="op">=</span> <span class="va">status_trans</span>, 
  coverage <span class="op">=</span> <span class="va">bwfl</span>, linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ex1 <span class="op">=</span> <span class="st">"#4B0055"</span>, ex2 <span class="op">=</span> <span class="st">"#007094"</span><span class="op">)</span>,
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="st">"forestgreen"</span>, Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Undetected <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span>,
  collapseTranscripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="cn">FALSE</span>, Unchanged <span class="op">=</span> <span class="cn">FALSE</span>, Undetected <span class="op">=</span> <span class="st">"meta"</span><span class="op">)</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-with-coverage-1.png" width="960"></p>
<p>Alternatively, by providing a list of <code>BigWigFileList</code>
objects, each list element will be drawn as a single overlaid track. In
this way, unlimited coverage tracks can effectively be drawn.</p>
<p>If choosing this option, colours must again be passed as a matching,
<em>named</em> list.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cov_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>TF1 <span class="op">=</span> <span class="va">bwfl</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  hic <span class="op">=</span> <span class="va">ex_hic</span>,
  features <span class="op">=</span> <span class="va">feat_grl</span>, featcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Promoters <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,
  genes <span class="op">=</span> <span class="va">status_trans</span>, 
  coverage <span class="op">=</span> <span class="va">cov_list</span>, 
  linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>TF1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ex1 <span class="op">=</span> <span class="st">"#4B0055"</span>, ex2 <span class="op">=</span> <span class="st">"#007094"</span><span class="op">)</span><span class="op">)</span>,
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="st">"forestgreen"</span>, Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Undetected <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span>,
  collapseTranscripts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="cn">FALSE</span>, Unchanged <span class="op">=</span> <span class="cn">FALSE</span>, Undetected <span class="op">=</span> <span class="st">"meta"</span><span class="op">)</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-with-tracks-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="adding-annotations-to-coverage">Adding Annotations To Coverage<a class="anchor" aria-label="anchor" href="#adding-annotations-to-coverage"></a>
</h4>
<p>An indication of which regions are associated with increased or
decreased binding can also be a useful annotation to add to plots such
as the above. Although we technically performed no statistical testing,
let’s consider a window with logFC below -1 to be showing decreased
binding.</p>
<p>Similar to the features track, where the names of
<code>GRangesList</code> elements denote the different feature types,
able to then assigned a colour, coverage annotation tracks follow these
same rules. For each coverage track being annotated, a
<code>GRangesList</code> object can denote the ranges which can be
assigned different colours.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cov_annot</span> <span class="op">&lt;-</span> <span class="fu">splitAsList</span><span class="op">(</span><span class="va">res_gr</span>, <span class="va">res_gr</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Unchanged"</span>, <span class="st">"Decreased"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu">endoapply</span><span class="op">(</span><span class="va">granges</span><span class="op">)</span></code></pre></div>
<p>In the above, we have Unchanged and Decreased binding denoted as
annotations. In keeping with the approach of having a matching list
element for every coverage track, we would need to pass this as a list
which matched the coverage track</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span>
  <span class="va">gr</span>, 
  hic <span class="op">=</span> <span class="va">ex_hic</span>,
  features <span class="op">=</span> <span class="va">feat_grl</span>, featcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Promoters <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,
  genes <span class="op">=</span> <span class="va">status_trans</span>, 
  coverage <span class="op">=</span> <span class="va">cov_list</span>, 
  annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>TF1 <span class="op">=</span> <span class="va">cov_annot</span><span class="op">)</span>, 
  annotcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Decreased <span class="op">=</span> <span class="st">"#3333CC"</span><span class="op">)</span>,
  linecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>TF1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ex1 <span class="op">=</span> <span class="st">"#4B0055"</span>, ex2 <span class="op">=</span> <span class="st">"#007094"</span><span class="op">)</span><span class="op">)</span>,
  genecol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Up <span class="op">=</span> <span class="st">"forestgreen"</span>, Unchanged <span class="op">=</span> <span class="st">"grey"</span>, Undetected <span class="op">=</span> <span class="st">"grey80"</span><span class="op">)</span>,
  collapseTranscripts <span class="op">=</span> <span class="st">"meta"</span>,
  cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, zoom <span class="op">=</span> <span class="fl">1.2</span>
<span class="op">)</span></code></pre></div>
<p><img src="differential_binding_files/figure-html/plot-annot-1.png" width="960"></p>
<p>Plots are able to be tweaked considerably further via multiple
parameters, however these basic approaches cover the elemental
functionality of <code>plotHFCG()</code> for enabling simple &amp;
reproducible plotting across regions for multiple sites within a larger
experiment.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R Under development (unstable) (2022-03-14 r81896)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] extraChIPs_0.99.6           rtracklayer_1.55.4         </span>
<span class="co">##  [3] BiocParallel_1.29.20        csaw_1.29.1                </span>
<span class="co">##  [5] SummarizedExperiment_1.25.3 Biobase_2.55.2             </span>
<span class="co">##  [7] MatrixGenerics_1.7.0        matrixStats_0.61.0         </span>
<span class="co">##  [9] Rsamtools_2.11.0            Biostrings_2.63.3          </span>
<span class="co">## [11] XVector_0.35.0              GenomicRanges_1.47.6       </span>
<span class="co">## [13] GenomeInfoDb_1.31.7         IRanges_2.29.1             </span>
<span class="co">## [15] S4Vectors_0.33.17           BiocGenerics_0.41.2        </span>
<span class="co">## [17] forcats_0.5.1               stringr_1.4.0              </span>
<span class="co">## [19] dplyr_1.0.8                 purrr_0.3.4                </span>
<span class="co">## [21] readr_2.1.2                 tidyr_1.2.0                </span>
<span class="co">## [23] tibble_3.1.6                ggplot2_3.3.5              </span>
<span class="co">## [25] tidyverse_1.3.1             BiocStyle_2.23.1           </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] utf8_1.2.2                 tidyselect_1.1.2          </span>
<span class="co">##   [3] RSQLite_2.2.12             AnnotationDbi_1.57.1      </span>
<span class="co">##   [5] htmlwidgets_1.5.4          grid_4.2.0                </span>
<span class="co">##   [7] scatterpie_0.1.7           munsell_0.5.0             </span>
<span class="co">##   [9] codetools_0.2-18           ragg_1.2.2                </span>
<span class="co">##  [11] withr_2.5.0                colorspace_2.0-3          </span>
<span class="co">##  [13] filelock_1.0.2             highr_0.9                 </span>
<span class="co">##  [15] knitr_1.38                 rstudioapi_0.13           </span>
<span class="co">##  [17] ggside_0.2.0               labeling_0.4.2            </span>
<span class="co">##  [19] GenomeInfoDbData_1.2.8     polyclip_1.10-0           </span>
<span class="co">##  [21] farver_2.1.0               bit64_4.0.5               </span>
<span class="co">##  [23] rprojroot_2.0.3            vctrs_0.4.1               </span>
<span class="co">##  [25] generics_0.1.2             xfun_0.30                 </span>
<span class="co">##  [27] biovizBase_1.43.1          BiocFileCache_2.3.4       </span>
<span class="co">##  [29] R6_2.5.1                   doParallel_1.0.17         </span>
<span class="co">##  [31] clue_0.3-60                locfit_1.5-9.5            </span>
<span class="co">##  [33] AnnotationFilter_1.19.0    bitops_1.0-7              </span>
<span class="co">##  [35] cachem_1.0.6               DelayedArray_0.21.2       </span>
<span class="co">##  [37] assertthat_0.2.1           BiocIO_1.5.0              </span>
<span class="co">##  [39] scales_1.2.0               nnet_7.3-17               </span>
<span class="co">##  [41] gtable_0.3.0               ensembldb_2.19.10         </span>
<span class="co">##  [43] rlang_1.0.2                systemfonts_1.0.4         </span>
<span class="co">##  [45] GlobalOptions_0.1.2        splines_4.2.0             </span>
<span class="co">##  [47] lazyeval_0.2.2             dichromat_2.0-0           </span>
<span class="co">##  [49] broom_0.8.0                checkmate_2.0.0           </span>
<span class="co">##  [51] BiocManager_1.30.16        yaml_2.3.5                </span>
<span class="co">##  [53] modelr_0.1.8               GenomicFeatures_1.47.13   </span>
<span class="co">##  [55] backports_1.4.1            Hmisc_4.6-0               </span>
<span class="co">##  [57] EnrichedHeatmap_1.25.2     tools_4.2.0               </span>
<span class="co">##  [59] bookdown_0.26              ellipsis_0.3.2            </span>
<span class="co">##  [61] jquerylib_0.1.4            RColorBrewer_1.1-3        </span>
<span class="co">##  [63] Rcpp_1.0.8.3               base64enc_0.1-3           </span>
<span class="co">##  [65] progress_1.2.2             zlibbioc_1.41.0           </span>
<span class="co">##  [67] RCurl_1.98-1.6             prettyunits_1.1.1         </span>
<span class="co">##  [69] rpart_4.1.16               GetoptLong_1.0.5          </span>
<span class="co">##  [71] haven_2.5.0                ggrepel_0.9.1             </span>
<span class="co">##  [73] cluster_2.1.3              fs_1.5.2                  </span>
<span class="co">##  [75] magrittr_2.0.3             data.table_1.14.2         </span>
<span class="co">##  [77] circlize_0.4.14            reprex_2.0.1              </span>
<span class="co">##  [79] ProtGenerics_1.27.2        hms_1.1.1                 </span>
<span class="co">##  [81] evaluate_0.15              XML_3.99-0.9              </span>
<span class="co">##  [83] jpeg_0.1-9                 readxl_1.4.0              </span>
<span class="co">##  [85] gridExtra_2.3              shape_1.4.6               </span>
<span class="co">##  [87] compiler_4.2.0             biomaRt_2.51.4            </span>
<span class="co">##  [89] crayon_1.5.1               htmltools_0.5.2           </span>
<span class="co">##  [91] ggfun_0.0.6                tzdb_0.3.0                </span>
<span class="co">##  [93] Formula_1.2-4              lubridate_1.8.0           </span>
<span class="co">##  [95] DBI_1.1.2                  tweenr_1.0.2              </span>
<span class="co">##  [97] dbplyr_2.1.1               ComplexHeatmap_2.11.1     </span>
<span class="co">##  [99] MASS_7.3-56                GenomicInteractions_1.29.5</span>
<span class="co">## [101] rappdirs_0.3.3             Matrix_1.4-1              </span>
<span class="co">## [103] cli_3.2.0                  parallel_4.2.0            </span>
<span class="co">## [105] Gviz_1.39.6                metapod_1.3.0             </span>
<span class="co">## [107] igraph_1.3.0               pkgconfig_2.0.3           </span>
<span class="co">## [109] pkgdown_2.0.2.9000         GenomicAlignments_1.31.2  </span>
<span class="co">## [111] foreign_0.8-82             xml2_1.3.3                </span>
<span class="co">## [113] InteractionSet_1.23.1      foreach_1.5.2             </span>
<span class="co">## [115] bslib_0.3.1                rvest_1.0.2               </span>
<span class="co">## [117] VariantAnnotation_1.41.3   digest_0.6.29             </span>
<span class="co">## [119] rmarkdown_2.13             cellranger_1.1.0          </span>
<span class="co">## [121] htmlTable_2.4.0            edgeR_3.37.1              </span>
<span class="co">## [123] restfulr_0.0.13            curl_4.3.2                </span>
<span class="co">## [125] rjson_0.2.21               lifecycle_1.0.1           </span>
<span class="co">## [127] jsonlite_1.8.0             viridisLite_0.4.0         </span>
<span class="co">## [129] desc_1.4.1                 limma_3.51.7              </span>
<span class="co">## [131] BSgenome_1.63.6            fansi_1.0.3               </span>
<span class="co">## [133] pillar_1.7.0               lattice_0.20-45           </span>
<span class="co">## [135] KEGGREST_1.35.0            fastmap_1.1.0             </span>
<span class="co">## [137] httr_1.4.2                 survival_3.3-1            </span>
<span class="co">## [139] glue_1.6.2                 png_0.1-7                 </span>
<span class="co">## [141] iterators_1.0.14           bit_4.0.4                 </span>
<span class="co">## [143] ggforce_0.3.3              stringi_1.7.6             </span>
<span class="co">## [145] sass_0.4.1                 blob_1.2.3                </span>
<span class="co">## [147] textshaping_0.3.6          latticeExtra_0.6-29       </span>
<span class="co">## [149] memoise_2.0.1</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/steveped" class="external-link">Stephen
Pederson</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
