<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Differential Signal Using Fixed-Width Windows • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Differential Signal Using Fixed-Width Windows">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_signal_sliding.html">Differential Signal Using Sliding Windows</a>
    </li>
    <li>
      <a href="../articles/differential_signal_fixed.html">Differential Signal Using Fixed-Width Windows</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/smped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Differential Signal Using
Fixed-Width Windows</h1>
                        <h4 data-toc-skip class="author">Stevie
Pederson</h4>
            <address class="author_afil">
      Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide,
AustraliaDame Roma Mitchell Cancer Researc Laboratories, University of
AdelaideJohn Curtin School of Medical Research, Australian National
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/smped/extraChIPs/blob/HEAD/vignettes/differential_signal_fixed.Rmd" class="external-link"><code>vignettes/differential_signal_fixed.Rmd</code></a></small>
      <div class="hidden name"><code>differential_signal_fixed.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>message <span class="op">=</span> <span class="cn">FALSE</span>, crop <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <a href="https://github.com/smped/GRAVI" class="external-link">GRAVI</a> workflow, for
which this package is designed, uses sliding windows for differential
signal analysis. However, the use of fixed-width windows, as is common
under DiffBind-style (Ross-Innes et al. 2012) approaches is also
possible with <code>extraChIPs</code>. This vignette focusses on using
conventional peak calls and fixed-width approaches to replicate and
extend these approaches.</p>
<p>The majority of examples below use heavily reduced datasets to
provide general guidance on using the functions. Some results may appear
trivial as a result, but will hopefully prove far more useful in a true
experimental context. All data, along with this vignette are available
<a href="https://github.com/smped/extraChIPs_vignette" class="external-link">here</a>. Please
place all contents of the data directory in a directory named data in
your own working directory.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>In order to use the package <code>extraChIPs</code> and follow this
vignette, we recommend using the package <code>BiocManager</code> hosted
on CRAN. Once this is installed, the additional packages required for
this vignette (<code>tidyverse</code>, <code>Rsamtools</code>,
<code>csaw</code>, <code>BiocParallel</code> and
<code>rtracklayer</code>) can also be installed.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="va">pkg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"tidyverse"</span>, <span class="st">"Rsamtools"</span>, <span class="st">"csaw"</span>, <span class="st">"BiocParallel"</span>, <span class="st">"rtracklayer"</span>, <span class="st">"edgeR"</span>, </span>
<span>  <span class="st">"patchwork"</span>, <span class="st">"extraChIPs"</span>, <span class="st">"plyranges"</span>, <span class="st">"scales"</span>, <span class="st">"here"</span>, <span class="st">"quantro"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">pkg</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Once these packages are installed, we can load them easily</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csaw</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidyverse/glue" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">quantro</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h3>
<p>All data for this vignette is expected to be in a sub-directory of
the working directory named “data”, and all paths will be predicated on
this. Please ensure you have all data in this location, obtained from <a href="https://github.com/smped/extraChIPs_vignette" class="external-link">here</a>.</p>
<p>The data itself is ChIP-Seq data targeting the Estrogen Receptor
(ER), and is taken from the cell-line ZR-75-1 cell-line using data from
the BioProject , Pre-processing was performed using the <a href="https://github.com/smped/prepareChIPs" class="external-link"><code>prepareChIPs</code></a>
workflow, written in snakemake (Mölder et al. 2021) and all code is
available at <a href="https://github.com/smped/PRJNA509779" class="external-link uri">https://github.com/smped/PRJNA509779</a>. ER binding was
assessed under Vehicle (E2) and DHT-stimulated (E2DHT) conditions. Using
GRCh37 as the reference genome, a subset of regions found on chromosome
10 are included in this dataset for simplicity.</p>
<p>First we’ll define our sample data then define our two treatment
groups. Defining a consistent colour palette for all plots is also a
good habit to develop.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treat_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"E2"</span>, <span class="st">"E2DHT"</span><span class="op">)</span></span>
<span><span class="va">treat_colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steelblue"</span>, <span class="st">"red3"</span><span class="op">)</span>, <span class="va">treat_levels</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  accession <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SRR831518"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  target <span class="op">=</span> <span class="st">"ER"</span>,</span>
<span>  treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">treat_levels</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, levels <span class="op">=</span> <span class="va">treat_levels</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">##   accession  target treatment</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> SRR8315180 ER     E2       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> SRR8315181 ER     E2       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> SRR8315182 ER     E2       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> SRR8315183 ER     E2DHT    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> SRR8315184 ER     E2DHT    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> SRR8315185 ER     E2DHT</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">accessions</span> <span class="op">&lt;-</span> <span class="va">samples</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">.</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pull</span>, <span class="st">"accession"</span><span class="op">)</span></span></code></pre></div>
<p>We’ll eventually be loading counts for differential signal analysis
from a set of BamFiles, so first we’ll create a <code>BamFileList</code>
with all of these files.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"ER"</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{samples$accession}.bam"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove_all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE TRUE TRUE TRUE TRUE TRUE</span></span></code></pre>
<p>This also enables creation of a <code>Seqinfo</code> object based on
the actual reference genome to which the reads were aligned during data
preparation. <code>Seqinfo</code> objects are the foundation of working
with GRanges, so defining an object at the start of a workflow is good
practice. As is common practice for ChIP-Seq analyses, we’ll restrict
our focus to the standard chromosomes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span></span>
<span><span class="va">sq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">isCircular</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GRCh37"</span></span></code></pre></div>
<p>Another key preparatory step for working with peaks is to define a
set of regions as either blacklisted or grey-listed regions. The former
are known problematic regions based on each genome, with data freely
available from <a href="https://github.com/Boyle-Lab/Blacklist/tree/master/lists" class="external-link uri">https://github.com/Boyle-Lab/Blacklist/tree/master/lists</a>,
whilst grey-listed regions are defined from potentially problematic
regions as detected within the input sample. For our samples code for
this is included in the previously provided repository (<a href="https://github.com/smped/PRJNA509779" class="external-link uri">https://github.com/smped/PRJNA509779</a>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">greylist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data/chr10_greylist.bed"</span><span class="op">)</span>, seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span></span>
<span><span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data/chr10_blacklist.bed"</span><span class="op">)</span>, seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span></span>
<span><span class="va">omit_ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">greylist</span>, <span class="va">blacklist</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="working-with-peaks">Working With Peaks<a class="anchor" aria-label="anchor" href="#working-with-peaks"></a>
</h2>
<p>The provided dataset includes six files produced by
<code>macs2 callpeak</code> (Zhang et al. 2008) in the
<code>narrowPeak</code> format, and these are able to be easily parsed
using <code>extraChIPs</code>. We’ll immediately pass our black &amp;
grey-listed regions to our parsing function so we can exclude these
regions right from the start. By passing the above seqinfo object to
this function, we’re also telling <code><a href="../reference/importPeaks.html">importPeaks()</a></code> to ignore
any peaks not on the included chromosomes.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"ER"</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{samples$accession}_peaks.narrowPeak"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/importPeaks.html">importPeaks</a></span><span class="op">(</span>seqinfo <span class="op">=</span> <span class="va">sq</span>, blacklist <span class="op">=</span> <span class="va">omit_ranges</span><span class="op">)</span></span></code></pre></div>
<p>This will import the peaks from all files as a single
<code>GRangesList</code> object, adding the file-name to each element by
default. We can easily modify these names if we so wish.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove_all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span>, <span class="st">"_peaks.narrowPeak"</span><span class="op">)</span></span></code></pre></div>
<p>Once loaded, we can easily check how similar our replicates are using
<code><a href="../reference/plotOverlaps-methods.html">plotOverlaps()</a></code>. When three or more sets of peaks are
contained in the <code>GRangesList</code>, an UpSet plot will be drawn
by default.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOverlaps-methods.html">plotOverlaps</a></span><span class="op">(</span></span>
<span>  <span class="va">peaks</span>, min_size <span class="op">=</span> <span class="fl">10</span>, .sort_sets <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  set_col <span class="op">=</span> <span class="va">treat_colours</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Optionally, specifying a column and a suitable function will produce
an additional panel summarising that value. In the following, we’ll show
the maximum score obtained, highlighting that for peaks identified in
only one or two replicates, the overall signal intensity is generally
lower, even in the sample with the strongest signal.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOverlaps-methods.html">plotOverlaps</a></span><span class="op">(</span></span>
<span>  <span class="va">peaks</span>, min_size <span class="op">=</span> <span class="fl">10</span>, .sort_sets <span class="op">=</span> <span class="cn">FALSE</span>, var <span class="op">=</span> <span class="st">"score"</span>, f <span class="op">=</span> <span class="st">"max"</span>,</span>
<span>   set_col <span class="op">=</span> <span class="va">treat_colours</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_fixed_files/figure-html/plot-overlaps-score-1.png" alt="*UpSet plot showing overlapping peaks across all replicates, with the maximum score across all replicates shown in the upper panel.*" width="700"><p class="caption">
<em>UpSet plot showing overlapping peaks across all replicates, with the
maximum score across all replicates shown in the upper panel.</em>
</p>
</div>
<p>A common task at this point may be to define consensus peaks within
each treatment group, by retaining only the peaks found in 2 of the 3
replicates <code>(p = 2/3)</code>. The default approach is to take the
union of all ranges, with the returned object containing logical values
for each sample, as well as the number of samples where an overlapping
peak was found.</p>
<p>If we wish to retain any of the original columns, such as the
<code>macs2 callpeak</code> score, we can simply pass the column names
to <code><a href="../reference/makeConsensus.html">makeConsensus()</a></code></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_e2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">[</span><span class="va">accessions</span><span class="op">$</span><span class="va">E2</span><span class="op">]</span>, p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span></span>
<span><span class="va">consensus_e2dht</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">[</span><span class="va">accessions</span><span class="op">$</span><span class="va">E2DHT</span><span class="op">]</span>, p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"score"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, we could find the centre of the peaks as part of this
process, by averaging across the estimated peak centres for each sample.
This is a very common step for ChIP-Seq data where the target is a
<em>transcription factor</em>, and also forms a key step in the DiffBind
workflow.</p>
<p>In the following code chunk, we first find the centre for each sample
using the information provided by <code>macs2</code>, before retaining
this column when calling <code><a href="../reference/makeConsensus.html">makeConsensus()</a></code>. This will return
each of the individual centre-position estimates as a list for each
merged range, and using <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply()</a></code> we then take the mean
position as our estimate for the combined peak centre.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_e2</span> <span class="op">&lt;-</span> <span class="va">peaks</span><span class="op">[</span><span class="va">accessions</span><span class="op">$</span><span class="va">E2</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">endoapply</span><span class="op">(</span><span class="va">mutate</span>, centre <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">peak</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">consensus_e2</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 164 ranges and 5 metadata columns:</span></span>
<span><span class="co">##         seqnames            ranges strand |    centre SRR8315180 SRR8315181</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;  &lt;logical&gt;  &lt;logical&gt;</span></span>
<span><span class="co">##     [1]    chr10 43048195-43048529      * |  43048362       TRUE       TRUE</span></span>
<span><span class="co">##     [2]    chr10 43521739-43522260      * |  43522020       TRUE       TRUE</span></span>
<span><span class="co">##     [3]    chr10 43540042-43540390      * |  43540272       TRUE      FALSE</span></span>
<span><span class="co">##     [4]    chr10 43606238-43606573      * |  43606416       TRUE       TRUE</span></span>
<span><span class="co">##     [5]    chr10 43851214-43851989      * |  43851719      FALSE       TRUE</span></span>
<span><span class="co">##     ...      ...               ...    ... .       ...        ...        ...</span></span>
<span><span class="co">##   [160]    chr10 99096784-99097428      * |  99097254       TRUE       TRUE</span></span>
<span><span class="co">##   [161]    chr10 99168353-99168649      * |  99168502       TRUE       TRUE</span></span>
<span><span class="co">##   [162]    chr10 99207868-99208156      * |  99207998      FALSE       TRUE</span></span>
<span><span class="co">##   [163]    chr10 99331363-99331730      * |  99331595       TRUE       TRUE</span></span>
<span><span class="co">##   [164]    chr10 99621632-99621961      * |  99621818      FALSE       TRUE</span></span>
<span><span class="co">##         SRR8315182         n</span></span>
<span><span class="co">##          &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]       TRUE         3</span></span>
<span><span class="co">##     [2]       TRUE         3</span></span>
<span><span class="co">##     [3]       TRUE         2</span></span>
<span><span class="co">##     [4]       TRUE         3</span></span>
<span><span class="co">##     [5]       TRUE         2</span></span>
<span><span class="co">##     ...        ...       ...</span></span>
<span><span class="co">##   [160]       TRUE         3</span></span>
<span><span class="co">##   [161]       TRUE         3</span></span>
<span><span class="co">##   [162]       TRUE         2</span></span>
<span><span class="co">##   [163]       TRUE         3</span></span>
<span><span class="co">##   [164]       TRUE         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_e2dht</span> <span class="op">&lt;-</span> <span class="va">peaks</span><span class="op">[</span><span class="va">accessions</span><span class="op">$</span><span class="va">E2DHT</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">endoapply</span><span class="op">(</span><span class="va">mutate</span>, centre <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">peak</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can also inspect these using <code><a href="../reference/plotOverlaps-methods.html">plotOverlaps()</a></code> provided
we use a <code>GRangesList</code> for the input. Now that we only have
two elements (one for each treatment) a VennDiagram will be generated
instead of an UpSet plot.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span>E2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">consensus_e2</span><span class="op">)</span>, E2DHT <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">consensus_e2dht</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotOverlaps-methods.html">plotOverlaps</a></span><span class="op">(</span>set_col <span class="op">=</span> <span class="va">treat_colours</span><span class="op">[</span><span class="va">treat_levels</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_fixed_files/figure-html/venn-consensus-overlap-1.png" alt="*Overlap between consensus peaks identified in a treatment-specific manner*" width="576"><p class="caption">
<em>Overlap between consensus peaks identified in a treatment-specific
manner</em>
</p>
</div>
<p>We can now go one step further and define the set of peaks found in
either treatment. Given we’re being inclusive here, we can leave p = 0
so any peak found in <em>either treatment</em> is included.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">union_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span></span>
<span>  E2 <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">consensus_e2</span>, <span class="va">centre</span><span class="op">)</span>, </span>
<span>  E2DHT <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">consensus_e2dht</span>, <span class="va">centre</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"centre"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p>Now we have a set of peaks, found in at least 2/3 of samples from
either condition, with estimates of each peak’s centre. The next step
would be to set all peaks as the same width based on the centre
position, with a common width being 500bp.</p>
<p>In the following we’ll perform multiple operations in a single call
mutate, so let’s make sure we know what’s happening.</p>
<ol style="list-style-type: decimal">
<li>
<code>glue("{seqnames}:{centre}:{strand}")</code> uses
<code>glue</code> syntax to parse the seqnames, centre position and
strand information as a character-like vector with a width of only 1,
and using the estimated centre as the Range.</li>
<li>We then coerce this to a <code>GRanges</code> object, before
resizing to the desired width.</li>
<li>We also add the original (un-centred) range as an additional column,
retaining the <code>GRanges</code> structure, but discarding anything in
the <code>mcols()</code> element, then</li>
<li>Using <code><a href="../reference/colToRanges-methods.html">colToRanges()</a></code>, we take the centred ranges and
place them as the core set of GRanges for this object.</li>
</ol>
<p>This gives a GRanges object with all original information, but with
centred peaks of a fixed width.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">w</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">centred_peaks</span> <span class="op">&lt;-</span> <span class="va">union_peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    centre <span class="op">=</span> <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{seqnames}:{centre}:{strand}"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span>width <span class="op">=</span> <span class="va">w</span><span class="op">)</span>,</span>
<span>    union_peak <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/colToRanges-methods.html">colToRanges</a></span><span class="op">(</span><span class="st">"centre"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="counting-reads">Counting Reads<a class="anchor" aria-label="anchor" href="#counting-reads"></a>
</h2>
<p>Now we have our centred, fixed-width peaks, we can count reads using
<code><a href="https://rdrr.io/pkg/csaw/man/regionCounts.html" class="external-link">csaw::regionCounts()</a></code> (Lun and Smyth 2016). We know our
fragment length is about 200bp, so we can pass this to the function for
a slightly more sophisticated approach to counting.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/regionCounts.html" class="external-link">regionCounts</a></span><span class="op">(</span><span class="va">bfl</span>, <span class="va">centred_peaks</span>, ext <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 188 6 </span></span>
<span><span class="co">## metadata(2): final.ext param</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(4): E2 E2DHT n union_peak</span></span>
<span><span class="co">## colnames(6): SRR8315180 SRR8315181 ... SRR8315184 SRR8315185</span></span>
<span><span class="co">## colData names(4): bam.files totals ext rlen</span></span></code></pre>
<p>The <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> element of the returned object as the
columns <em>bam.files</em>, <em>totals</em>, <em>ext</em> and
<em>rlen</em>, which are all informative and can be supplemented with
our <code>samples</code> data frame. In the following, we’ll 1) coerce
to a <code>tibble</code>, 2) <code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join()</a></code> the
<code>samples</code> object, 3) add the accession as the sample column,
4) set the accession back as the rownames, then 5) coerce back to the
required <code>DataFrame()</code> structure.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"accession"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">accession</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"accession"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">DataFrame</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 7 columns</span></span>
<span><span class="co">##                         bam.files    totals       ext      rlen      target</span></span>
<span><span class="co">##                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">## SRR8315180 /home/steviep/github..    317845       200        75          ER</span></span>
<span><span class="co">## SRR8315181 /home/steviep/github..    337623       200        75          ER</span></span>
<span><span class="co">## SRR8315182 /home/steviep/github..    341998       200        75          ER</span></span>
<span><span class="co">## SRR8315183 /home/steviep/github..    315872       200        75          ER</span></span>
<span><span class="co">## SRR8315184 /home/steviep/github..    352908       200        75          ER</span></span>
<span><span class="co">## SRR8315185 /home/steviep/github..    347709       200        75          ER</span></span>
<span><span class="co">##            treatment      sample</span></span>
<span><span class="co">##             &lt;factor&gt; &lt;character&gt;</span></span>
<span><span class="co">## SRR8315180     E2     SRR8315180</span></span>
<span><span class="co">## SRR8315181     E2     SRR8315181</span></span>
<span><span class="co">## SRR8315182     E2     SRR8315182</span></span>
<span><span class="co">## SRR8315183     E2DHT  SRR8315183</span></span>
<span><span class="co">## SRR8315184     E2DHT  SRR8315184</span></span>
<span><span class="co">## SRR8315185     E2DHT  SRR8315185</span></span></code></pre>
<p>For QC and visualisation, we can add an additional
<code>logCPM</code> assay to our object as well.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"logCPM"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span>, lib.size <span class="op">=</span> <span class="va">se</span><span class="op">$</span><span class="va">totals</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>First we might like to check our distribution of counts</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">se</span>, assay <span class="op">=</span> <span class="st">"counts"</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, trans <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_fixed_files/figure-html/plot-assay-densities-1.png" alt="*Count densities for all samples, using the log+1 transformation*" width="700"><p class="caption">
<em>Count densities for all samples, using the log+1 transformation</em>
</p>
</div>
<p>A PCA plot can also provide insight as to where the variability in
the data lies.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssayPCA-methods.html">plotAssayPCA</a></span><span class="op">(</span><span class="va">se</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treat"</span>, label <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_fixed_files/figure-html/plot-pca-1.png" alt="*PCA plot using logCPM values and showing that replicate variability is larger than variability between treatment groups.*" width="700"><p class="caption">
<em>PCA plot using logCPM values and showing that replicate variability
is larger than variability between treatment groups.</em>
</p>
</div>
</div>
<div class="section level2">
<h2 id="differential-signal-analysis">Differential Signal Analysis<a class="anchor" aria-label="anchor" href="#differential-signal-analysis"></a>
</h2>
<div class="section level3">
<h3 id="statistical-testing">Statistical Testing<a class="anchor" aria-label="anchor" href="#statistical-testing"></a>
</h3>
<p>In order to perform Differential Signal Analysis, we simply need to
define a model matrix, as for conventional analysis using
<code>edgeR</code> or <code>limma</code>. We can then pass this, along
with our fixed-width counts to <code><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff()</a></code>. By default
normalisation will be <em>library-size</em> normalisation, as is a
common default strategy for ChIP-Seq data. In contrast to sliding window
approaches, these results represent our final results and there is no
need for merging windows.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">treatment</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ls_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff</a></span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="va">X</span>, asRanges <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">ls_res</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3</span></span></code></pre>
<p>TMM normalisation (Robinson and Oshlack 2010) is another common
strategy, which relies on the data from all treatment groups being drawn
from the same distributions. We can formally test this using the package
<code>quantro</code> (Hicks and Irizarry 2015) , which produces p-values
for 1) H<sub>0</sub>: Group medians are drawn from the same
distribution, and 2) H<sub>0</sub>: Group-specific distributions are the
same.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">qtest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/quantro/man/quantro.html" class="external-link">quantro</a></span><span class="op">(</span>groupFactor <span class="op">=</span> <span class="va">se</span><span class="op">$</span><span class="va">treatment</span>, B <span class="op">=</span> <span class="fl">1e3</span><span class="op">)</span></span>
<span><span class="va">qtest</span></span></code></pre></div>
<pre><code><span><span class="co">## quantro: Test for global differences in distributions</span></span>
<span><span class="co">##    nGroups:  2 </span></span>
<span><span class="co">##    nTotSamples:  6 </span></span>
<span><span class="co">##    nSamplesinGroups:  3 3 </span></span>
<span><span class="co">##    anovaPval:  0.90754 </span></span>
<span><span class="co">##    quantroStat:  0.21859 </span></span>
<span><span class="co">##    quantroPvalPerm:  0.572</span></span></code></pre>
<p>Here, both p-values are &gt;0.05, so in conjunction with out visual
inspection earlier, we can confidently apply TMM normalisation. To apply
this, we simply specify the argument <code>norm = "TMM"</code> when we
call <code><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff()</a></code>. In the analysis below, we’ve also
specified a fold-change threshold <code>(fc = 1.2)</code>, below which,
changes in signal are considered to not be of interest (McCarthy and
Smyth 2009). This threshold is incorporated into the testing so there is
no requirement for <em>post-hoc</em> filtering based on a threshold.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff</a></span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="va">X</span>, norm <span class="op">=</span> <span class="st">"TMM"</span>, asRanges <span class="op">=</span> <span class="cn">TRUE</span>, fc <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">tmm_res</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7</span></span></code></pre>
<p>An MA-plot is a common way of inspecting results and in the following
we use the original ‘union_peak’ in our labelling of points. This serves
as a reminder that the fixed-width windows are in fact <em>a proxy</em>
for the entire region for which we have confidently detected ChIP
signal, and that these windows are truly the regions of interest.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>`FDR &lt; 0.05` <span class="op">=</span> <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logCPM</span>, <span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">`FDR &lt; 0.05`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">union_peak</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_fixed_files/figure-html/plot-ma-1.png" alt="*MA-plot after fitting using TMM normalisation and applying a fold-change threshold during testing. Points are labelled using the original windows obtained when merging replicates and treatment groups.*" width="700"><p class="caption">
<em>MA-plot after fitting using TMM normalisation and applying a
fold-change threshold during testing. Points are labelled using the
original windows obtained when merging replicates and treatment
groups.</em>
</p>
</div>
</div>
<div class="section level3">
<h3 id="mapping-to-genes">Mapping to Genes<a class="anchor" aria-label="anchor" href="#mapping-to-genes"></a>
</h3>
<p>Whilst knowledge of which regions are showing differential signal,
the fundamental question we are usually asking is about the downstream
regulatory consequences, such as the target gene. Before we can map
peaks to genes, we’ll need to define our genes. In the following, we’ll
use the provided Gencode gene mappings at the gene, transcript and exon
level.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gencode</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data/gencode.v43lift37.chr10.annotation.gtf.gz"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-filter-overlaps.html" class="external-link">filter_by_overlaps</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr10:42354900-100000000"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gencode</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">gencode</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sq</span></span></code></pre></div>
<p>Mapping to genes using <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> uses additional
annotations, such as whether the peak overlaps a promoter, enhancer or
long-range interaction. Here we’ll just use promoters, so let’s create a
set of promoters from our transcript-level information, ensuring we
incorporate all possible promoters within a gene, and merging any
overlapping ranges using <code><a href="../reference/reduceMC.html">reduceMC()</a></code></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">promoters</span> <span class="op">&lt;-</span> <span class="va">gencode</span><span class="op">$</span><span class="va">transcript</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">ends_with</a></span><span class="op">(</span><span class="st">"name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span>upstream <span class="op">=</span> <span class="fl">2500</span>, downstream <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/reduceMC.html">reduceMC</a></span><span class="op">(</span>simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">promoters</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 1678 ranges and 3 metadata columns:</span></span>
<span><span class="co">##          seqnames            ranges strand |</span></span>
<span><span class="co">##             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span></span>
<span><span class="co">##      [1]    chr10 42678287-42681286      + |</span></span>
<span><span class="co">##      [2]    chr10 42702938-42705937      + |</span></span>
<span><span class="co">##      [3]    chr10 42735669-42738668      + |</span></span>
<span><span class="co">##      [4]    chr10 42743933-42746932      + |</span></span>
<span><span class="co">##      [5]    chr10 42968428-42973155      + |</span></span>
<span><span class="co">##      ...      ...               ...    ... .</span></span>
<span><span class="co">##   [1674]    chr10 99635155-99638154      - |</span></span>
<span><span class="co">##   [1675]    chr10 99643500-99646805      - |</span></span>
<span><span class="co">##   [1676]    chr10 99695536-99698535      - |</span></span>
<span><span class="co">##   [1677]    chr10 99770595-99773594      - |</span></span>
<span><span class="co">##   [1678]    chr10 99789879-99793085      - |</span></span>
<span><span class="co">##                                                                     gene_id</span></span>
<span><span class="co">##                                                             &lt;CharacterList&gt;</span></span>
<span><span class="co">##      [1]                                                ENSG00000237592.2_5</span></span>
<span><span class="co">##      [2]                                                ENSG00000271650.1_7</span></span>
<span><span class="co">##      [3]                                                ENSG00000290458.1_2</span></span>
<span><span class="co">##      [4]                                                ENSG00000274167.5_8</span></span>
<span><span class="co">##      [5] ENSG00000185904.12_9,ENSG00000185904.12_9,ENSG00000185904.12_9,...</span></span>
<span><span class="co">##      ...                                                                ...</span></span>
<span><span class="co">##   [1674]                                                  ENSG00000265398.1</span></span>
<span><span class="co">##   [1675]                        ENSG00000095713.14_13,ENSG00000095713.14_13</span></span>
<span><span class="co">##   [1676]                                              ENSG00000095713.14_13</span></span>
<span><span class="co">##   [1677]                                              ENSG00000095713.14_13</span></span>
<span><span class="co">##   [1678]                        ENSG00000095713.14_13,ENSG00000095713.14_13</span></span>
<span><span class="co">##                                  gene_name</span></span>
<span><span class="co">##                            &lt;CharacterList&gt;</span></span>
<span><span class="co">##      [1]                       IGKV1OR10-1</span></span>
<span><span class="co">##      [2]                   ENSG00000271650</span></span>
<span><span class="co">##      [3]                   ENSG00000290458</span></span>
<span><span class="co">##      [4]                   ENSG00000274167</span></span>
<span><span class="co">##      [5] LINC00839,LINC00839,LINC00839,...</span></span>
<span><span class="co">##      ...                               ...</span></span>
<span><span class="co">##   [1674]                        AL139239.1</span></span>
<span><span class="co">##   [1675]                     CRTAC1,CRTAC1</span></span>
<span><span class="co">##   [1676]                            CRTAC1</span></span>
<span><span class="co">##   [1677]                            CRTAC1</span></span>
<span><span class="co">##   [1678]                     CRTAC1,CRTAC1</span></span>
<span><span class="co">##                                        transcript_name</span></span>
<span><span class="co">##                                        &lt;CharacterList&gt;</span></span>
<span><span class="co">##      [1]                               IGKV1OR10-1-201</span></span>
<span><span class="co">##      [2]                               ENST00000605702</span></span>
<span><span class="co">##      [3]                               ENST00000622823</span></span>
<span><span class="co">##      [4]                               ENST00000622650</span></span>
<span><span class="co">##      [5] LINC00839-204,LINC00839-203,LINC00839-202,...</span></span>
<span><span class="co">##      ...                                           ...</span></span>
<span><span class="co">##   [1674]                                AL139239.1-201</span></span>
<span><span class="co">##   [1675]                         CRTAC1-205,CRTAC1-206</span></span>
<span><span class="co">##   [1676]                                    CRTAC1-204</span></span>
<span><span class="co">##   [1677]                                    CRTAC1-201</span></span>
<span><span class="co">##   [1678]                         CRTAC1-203,CRTAC1-202</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<p>Now we’ll pass these to <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code>, but first, we’ll
place the original ‘union_peak’ back as the core of the GRanges object.
This will retain all the results from testing, but ensures the correct
region is mapped to genes.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_mapped_res</span> <span class="op">&lt;-</span> <span class="va">tmm_res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/colToRanges-methods.html">colToRanges</a></span><span class="op">(</span><span class="st">"union_peak"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/mapByFeature.html">mapByFeature</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="va">gencode</span><span class="op">$</span><span class="va">gene</span>, prom <span class="op">=</span> <span class="va">promoters</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    status <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>      <span class="va">FDR</span> <span class="op">&gt;=</span> <span class="fl">.05</span> <span class="op">~</span> <span class="st">"Unchanged"</span>,</span>
<span>      <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Increased"</span>,</span>
<span>      <span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Decreased"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">tmm_mapped_res</span>, <span class="va">PValue</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 188 ranges and 10 metadata columns:</span></span>
<span><span class="co">##         seqnames            ranges strand |        E2     E2DHT         n</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;logical&gt; &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]    chr10 81101906-81102928      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##     [2]    chr10 79629641-79630271      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##     [3]    chr10 89407752-89408138      * |     FALSE      TRUE         1</span></span>
<span><span class="co">##     [4]    chr10 52233596-52233998      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##     [5]    chr10 91651138-91651433      * |     FALSE      TRUE         1</span></span>
<span><span class="co">##     ...      ...               ...    ... .       ...       ...       ...</span></span>
<span><span class="co">##   [184]    chr10 89395292-89395626      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [185]    chr10 82723984-82724328      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [186]    chr10 93120411-93121224      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [187]    chr10 95755308-95755721      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [188]    chr10 79190987-79191351      * |     FALSE      TRUE         1</span></span>
<span><span class="co">##               logFC    logCPM      PValue         FDR</span></span>
<span><span class="co">##           &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]    1.880493   7.91492 2.50406e-25 4.70764e-23</span></span>
<span><span class="co">##     [2]    0.940221   8.07204 8.07816e-08 7.59347e-06</span></span>
<span><span class="co">##     [3]    1.578056   6.16574 4.18078e-07 2.61995e-05</span></span>
<span><span class="co">##     [4]    1.057814   6.67466 5.73897e-05 2.69732e-03</span></span>
<span><span class="co">##     [5]    1.546530   5.15431 1.75562e-04 6.60112e-03</span></span>
<span><span class="co">##     ...         ...       ...         ...         ...</span></span>
<span><span class="co">##   [184] -0.01346141   6.36140    0.961749    0.982656</span></span>
<span><span class="co">##   [185] -0.00855862   6.91968    0.972422    0.988154</span></span>
<span><span class="co">##   [186] -0.01173559   9.32397    0.979561    0.988154</span></span>
<span><span class="co">##   [187]  0.00721240   5.88042    0.982898    0.988154</span></span>
<span><span class="co">##   [188] -0.00157801   6.41182    0.994914    0.994914</span></span>
<span><span class="co">##                                            gene_id</span></span>
<span><span class="co">##                                    &lt;CharacterList&gt;</span></span>
<span><span class="co">##     [1]                       ENSG00000108179.14_6</span></span>
<span><span class="co">##     [2]                      ENSG00000151208.17_11</span></span>
<span><span class="co">##     [3]   ENSG00000225913.2_9,ENSG00000196566.2_10</span></span>
<span><span class="co">##     [4] ENSG00000198964.14_10,ENSG00000225303.2_10</span></span>
<span><span class="co">##     [5]                        ENSG00000280560.3_9</span></span>
<span><span class="co">##     ...                                        ...</span></span>
<span><span class="co">##   [184]   ENSG00000225913.2_9,ENSG00000196566.2_10</span></span>
<span><span class="co">##   [185]                        ENSG00000227209.1_5</span></span>
<span><span class="co">##   [186]                        ENSG00000289228.2_2</span></span>
<span><span class="co">##   [187]                      ENSG00000138193.17_12</span></span>
<span><span class="co">##   [188]                      ENSG00000156113.25_17</span></span>
<span><span class="co">##                               gene_name      status</span></span>
<span><span class="co">##                         &lt;CharacterList&gt; &lt;character&gt;</span></span>
<span><span class="co">##     [1]                            PPIF   Increased</span></span>
<span><span class="co">##     [2]                            DLG5   Increased</span></span>
<span><span class="co">##     [3] ENSG00000225913,ENSG00000196566   Increased</span></span>
<span><span class="co">##     [4]           SGMS1,ENSG00000225303   Increased</span></span>
<span><span class="co">##     [5]                       LINC01374   Increased</span></span>
<span><span class="co">##     ...                             ...         ...</span></span>
<span><span class="co">##   [184] ENSG00000225913,ENSG00000196566   Unchanged</span></span>
<span><span class="co">##   [185]                         WARS2P1   Unchanged</span></span>
<span><span class="co">##   [186]                 ENSG00000289228   Unchanged</span></span>
<span><span class="co">##   [187]                           PLCE1   Unchanged</span></span>
<span><span class="co">##   [188]                          KCNMA1   Unchanged</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="inspection-of-results">Inspection of Results<a class="anchor" aria-label="anchor" href="#inspection-of-results"></a>
</h2>
<div class="section level3">
<h3 id="profile-heatmaps">Profile Heatmaps<a class="anchor" aria-label="anchor" href="#profile-heatmaps"></a>
</h3>
<p>When analysing a transcription factor, checking the binding profile
across our treatment groups can be informative, and is often performed
using ‘Profile Heatmaps’ where coverage is smoothed within bins
surrounding our peak centre.</p>
<p>The function <code><a href="../reference/getProfileData-methods.html">getProfileData()</a></code> takes a set of ranges and
a BigWigFileList, and performs the smoothing, which is then passed to
the function <code><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap()</a></code>.</p>
<p>The following shows the three steps of 1) defining the ranges, 2)
obtaining the smoothed binding profiles, and 3) drawing the heatmap.
Note that we can facet the heatmaps by selecting the ‘status’ column to
separate any Increased or Decreased regions. By default, this will also
draw the smoothed lines in the top panel using different colours.</p>
<p>These plots can be used to show coverage-like values (SPMR or CPM) or
we can use fold-enrichment over the input sample(s), as is also produced
by <code>macs2 bdgcmp</code>. This data isn’t generally visualised using
log-transformation so we’ll set <code>log = FALSE</code> in our call to
<code><a href="../reference/getProfileData-methods.html">getProfileData()</a></code></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fe_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"ER"</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{treat_levels}_FE_chr10.bw"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFileList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">treat_levels</span><span class="op">)</span></span>
<span><span class="va">sig_ranges</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tmm_mapped_res</span>, <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">pd_fe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getProfileData-methods.html">getProfileData</a></span><span class="op">(</span><span class="va">fe_bw</span>, <span class="va">sig_ranges</span>, log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span><span class="va">pd_fe</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/plotProfileHeatmap-methods.html">plotProfileHeatmap</a></span><span class="op">(</span><span class="st">"profile_data"</span>, facetY <span class="op">=</span> <span class="st">"status"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Fold\nEnrichment"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] quantro_1.34.0              here_1.0.1                 </span></span>
<span><span class="co">##  [3] ggrepel_0.9.3               glue_1.6.2                 </span></span>
<span><span class="co">##  [5] scales_1.2.1                plyranges_1.20.0           </span></span>
<span><span class="co">##  [7] extraChIPs_1.5.6            ggside_0.2.2               </span></span>
<span><span class="co">##  [9] patchwork_1.1.2             edgeR_3.42.4               </span></span>
<span><span class="co">## [11] limma_3.56.2                rtracklayer_1.60.0         </span></span>
<span><span class="co">## [13] BiocParallel_1.34.2         csaw_1.34.0                </span></span>
<span><span class="co">## [15] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">## [17] MatrixGenerics_1.12.2       matrixStats_1.0.0          </span></span>
<span><span class="co">## [19] Rsamtools_2.16.0            Biostrings_2.68.1          </span></span>
<span><span class="co">## [21] XVector_0.40.0              GenomicRanges_1.52.0       </span></span>
<span><span class="co">## [23] GenomeInfoDb_1.36.1         IRanges_2.34.1             </span></span>
<span><span class="co">## [25] S4Vectors_0.38.1            BiocGenerics_0.46.0        </span></span>
<span><span class="co">## [27] lubridate_1.9.2             forcats_1.0.0              </span></span>
<span><span class="co">## [29] stringr_1.5.0               dplyr_1.1.2                </span></span>
<span><span class="co">## [31] purrr_1.0.1                 readr_2.1.4                </span></span>
<span><span class="co">## [33] tidyr_1.3.0                 tibble_3.2.1               </span></span>
<span><span class="co">## [35] ggplot2_3.4.2               tidyverse_2.0.0            </span></span>
<span><span class="co">## [37] BiocStyle_2.28.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.2                   ProtGenerics_1.32.0       </span></span>
<span><span class="co">##   [3] bitops_1.0-7               httr_1.4.6                </span></span>
<span><span class="co">##   [5] RColorBrewer_1.1-3         doParallel_1.0.17         </span></span>
<span><span class="co">##   [7] InteractionSet_1.28.1      doRNG_1.8.6               </span></span>
<span><span class="co">##   [9] tools_4.3.0                backports_1.4.1           </span></span>
<span><span class="co">##  [11] utf8_1.2.3                 R6_2.5.1                  </span></span>
<span><span class="co">##  [13] HDF5Array_1.28.1           mgcv_1.8-42               </span></span>
<span><span class="co">##  [15] lazyeval_0.2.2             Gviz_1.44.0               </span></span>
<span><span class="co">##  [17] rhdf5filters_1.12.1        GetoptLong_1.0.5          </span></span>
<span><span class="co">##  [19] withr_2.5.0                prettyunits_1.1.1         </span></span>
<span><span class="co">##  [21] gridExtra_2.3              base64_2.0.1              </span></span>
<span><span class="co">##  [23] VennDiagram_1.7.3          preprocessCore_1.62.1     </span></span>
<span><span class="co">##  [25] cli_3.6.1                  textshaping_0.3.6         </span></span>
<span><span class="co">##  [27] formatR_1.14               labeling_0.4.2            </span></span>
<span><span class="co">##  [29] sass_0.4.6                 genefilter_1.82.1         </span></span>
<span><span class="co">##  [31] askpass_1.1                pkgdown_2.0.7.9000        </span></span>
<span><span class="co">##  [33] systemfonts_1.0.4          foreign_0.8-84            </span></span>
<span><span class="co">##  [35] siggenes_1.74.0            illuminaio_0.42.0         </span></span>
<span><span class="co">##  [37] dichromat_2.0-0.1          scrime_1.3.5              </span></span>
<span><span class="co">##  [39] BSgenome_1.68.0            rstudioapi_0.14           </span></span>
<span><span class="co">##  [41] RSQLite_2.3.1              generics_0.1.3            </span></span>
<span><span class="co">##  [43] shape_1.4.6                BiocIO_1.10.0             </span></span>
<span><span class="co">##  [45] Matrix_1.5-4.1             interp_1.1-4              </span></span>
<span><span class="co">##  [47] futile.logger_1.4.3        fansi_1.0.4               </span></span>
<span><span class="co">##  [49] lifecycle_1.0.3            yaml_2.3.7                </span></span>
<span><span class="co">##  [51] rhdf5_2.44.0               BiocFileCache_2.8.0       </span></span>
<span><span class="co">##  [53] grid_4.3.0                 blob_1.2.4                </span></span>
<span><span class="co">##  [55] crayon_1.5.2               lattice_0.21-8            </span></span>
<span><span class="co">##  [57] ComplexUpset_1.3.3         GenomicFeatures_1.52.1    </span></span>
<span><span class="co">##  [59] annotate_1.78.0            KEGGREST_1.40.0           </span></span>
<span><span class="co">##  [61] pillar_1.9.0               knitr_1.43                </span></span>
<span><span class="co">##  [63] ComplexHeatmap_2.16.0      beanplot_1.3.1            </span></span>
<span><span class="co">##  [65] metapod_1.8.0              rjson_0.2.21              </span></span>
<span><span class="co">##  [67] codetools_0.2-19           data.table_1.14.8         </span></span>
<span><span class="co">##  [69] vctrs_0.6.3                png_0.1-8                 </span></span>
<span><span class="co">##  [71] gtable_0.3.3               cachem_1.0.8              </span></span>
<span><span class="co">##  [73] xfun_0.39                  S4Arrays_1.0.4            </span></span>
<span><span class="co">##  [75] survival_3.5-5             iterators_1.0.14          </span></span>
<span><span class="co">##  [77] GenomicInteractions_1.34.0 nlme_3.1-162              </span></span>
<span><span class="co">##  [79] bit64_4.0.5                progress_1.2.2            </span></span>
<span><span class="co">##  [81] filelock_1.0.2             rprojroot_2.0.3           </span></span>
<span><span class="co">##  [83] bslib_0.5.0                nor1mix_1.3-0             </span></span>
<span><span class="co">##  [85] rpart_4.1.19               colorspace_2.1-0          </span></span>
<span><span class="co">##  [87] DBI_1.1.3                  Hmisc_5.1-0               </span></span>
<span><span class="co">##  [89] nnet_7.3-19                tidyselect_1.2.0          </span></span>
<span><span class="co">##  [91] bit_4.0.5                  compiler_4.3.0            </span></span>
<span><span class="co">##  [93] curl_5.0.1                 htmlTable_2.4.1           </span></span>
<span><span class="co">##  [95] xml2_1.3.4                 desc_1.4.2                </span></span>
<span><span class="co">##  [97] DelayedArray_0.26.6        bookdown_0.34             </span></span>
<span><span class="co">##  [99] checkmate_2.2.0            quadprog_1.5-8            </span></span>
<span><span class="co">## [101] rappdirs_0.3.3             digest_0.6.32             </span></span>
<span><span class="co">## [103] rmarkdown_2.23             GEOquery_2.68.0           </span></span>
<span><span class="co">## [105] htmltools_0.5.5            pkgconfig_2.0.3           </span></span>
<span><span class="co">## [107] jpeg_0.1-10                base64enc_0.1-3           </span></span>
<span><span class="co">## [109] sparseMatrixStats_1.12.2   highr_0.10                </span></span>
<span><span class="co">## [111] dbplyr_2.3.2               fastmap_1.1.1             </span></span>
<span><span class="co">## [113] ensembldb_2.24.0           rlang_1.1.1               </span></span>
<span><span class="co">## [115] GlobalOptions_0.1.2        htmlwidgets_1.6.2         </span></span>
<span><span class="co">## [117] DelayedMatrixStats_1.22.1  EnrichedHeatmap_1.30.0    </span></span>
<span><span class="co">## [119] farver_2.1.1               jquerylib_0.1.4           </span></span>
<span><span class="co">## [121] jsonlite_1.8.7             mclust_6.0.0              </span></span>
<span><span class="co">## [123] VariantAnnotation_1.46.0   RCurl_1.98-1.12           </span></span>
<span><span class="co">## [125] magrittr_2.0.3             Formula_1.2-5             </span></span>
<span><span class="co">## [127] GenomeInfoDbData_1.2.10    Rhdf5lib_1.22.0           </span></span>
<span><span class="co">## [129] munsell_0.5.0              Rcpp_1.0.10               </span></span>
<span><span class="co">## [131] stringi_1.7.12             zlibbioc_1.46.0           </span></span>
<span><span class="co">## [133] MASS_7.3-60                bumphunter_1.42.0         </span></span>
<span><span class="co">## [135] plyr_1.8.8                 minfi_1.46.0              </span></span>
<span><span class="co">## [137] parallel_4.3.0             deldir_1.0-9              </span></span>
<span><span class="co">## [139] splines_4.3.0              multtest_2.56.0           </span></span>
<span><span class="co">## [141] hms_1.1.3                  circlize_0.4.15           </span></span>
<span><span class="co">## [143] locfit_1.5-9.8             igraph_1.5.0              </span></span>
<span><span class="co">## [145] rngtools_1.5.2             biomaRt_2.56.1            </span></span>
<span><span class="co">## [147] futile.options_1.0.1       XML_3.99-0.14             </span></span>
<span><span class="co">## [149] evaluate_0.21              latticeExtra_0.6-30       </span></span>
<span><span class="co">## [151] biovizBase_1.48.0          lambda.r_1.2.4            </span></span>
<span><span class="co">## [153] BiocManager_1.30.21        tzdb_0.4.0                </span></span>
<span><span class="co">## [155] foreach_1.5.2              tweenr_2.0.2              </span></span>
<span><span class="co">## [157] openssl_2.0.6              polyclip_1.10-4           </span></span>
<span><span class="co">## [159] reshape_0.8.9              clue_0.3-64               </span></span>
<span><span class="co">## [161] ggforce_0.4.1              xtable_1.8-4              </span></span>
<span><span class="co">## [163] broom_1.0.5                restfulr_0.0.15           </span></span>
<span><span class="co">## [165] AnnotationFilter_1.24.0    ragg_1.2.5                </span></span>
<span><span class="co">## [167] memoise_2.0.1              AnnotationDbi_1.62.2      </span></span>
<span><span class="co">## [169] GenomicAlignments_1.36.0   cluster_2.1.4             </span></span>
<span><span class="co">## [171] timechange_0.2.0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-gviz" class="csl-entry">
Hahne, Florian, and Robert Ivanek. 2016. “Statistical Genomics: Methods
and Protocols.” In, edited by Ewy Mathé and Sean Davis, 335–51. New
York, NY: Springer New York. <a href="https://doi.org/10.1007/978-1-4939-3578-9_16" class="external-link uri">https://doi.org/10.1007/978-1-4939-3578-9_16</a>.
</div>
<div id="ref-Hicks2015-ee" class="csl-entry">
Hicks, Stephanie C, and Rafael A Irizarry. 2015. “Quantro: A Data-Driven
Approach to Guide the Choice of an Appropriate Normalization Method.”
<em>Genome Biol.</em> 16 (1): 117.
</div>
<div id="ref-csaw2016" class="csl-entry">
Lun, Aaron T L, and Gordon K Smyth. 2016. “Csaw: A Bioconductor Package
for Differential Binding Analysis of ChIP-Seq Data Using Sliding
Windows.” <em>Nucleic Acids Res.</em> 44 (5): e45.
</div>
<div id="ref-McCarthy2009-qf" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. “Testing Significance
Relative to a Fold-Change Threshold Is a TREAT.” <em>Bioinformatics</em>
25 (6): 765–71.
</div>
<div id="ref-Molder2021-mo" class="csl-entry">
Mölder, Felix, Kim Philipp Jablonski, Brice Letcher, Michael B Hall,
Christopher H Tomkins-Tinch, Vanessa Sochat, Jan Forster, et al. 2021.
“Sustainable Data Analysis with Snakemake.” <em>F1000Res.</em> 10
(January): 33.
</div>
<div id="ref-Robinson2010-qp" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization
Method for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.” <em>Genome Biol.</em> 11 (3): R25.
</div>
<div id="ref-DiffBind2012" class="csl-entry">
Ross-Innes, Caryn S., Rory Stark, Andrew E. Teschendorff, Kelly A.
Holmes, H. Raza Ali, Mark J. Dunning, Gordon D. Brown, et al. 2012.
“Differential Oestrogen Receptor Binding Is Associated with Clinical
Outcome in Breast Cancer.” <em>Nature</em> 481: –4. <a href="http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html" class="external-link uri">http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html</a>.
</div>
<div id="ref-Zhang2008-ms" class="csl-entry">
Zhang, Yong, Tao Liu, Clifford A Meyer, Jérôme Eeckhoute, David S
Johnson, Bradley E Bernstein, Chad Nusbaum, et al. 2008. “Model-Based
Analysis of ChIP-Seq (MACS).” <em>Genome Biol.</em> 9 (9): R137.
</div>
</div>
<p><br></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephen Pederson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
