<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Differential Signal Using Sliding Windows • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Differential Signal Using Sliding Windows">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_signal_sliding.html">Differential Signal Using Sliding Windows</a>
    </li>
    <li>
      <a href="../articles/differential_signal_fixed.html">Differential Signal Using Fixed-Width Windows</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/smped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Differential Signal Using Sliding
Windows</h1>
                        <h4 data-toc-skip class="author">Stevie
Pederson</h4>
            <address class="author_afil">
      Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide,
AustraliaDame Roma Mitchell Cancer Researc Laboratories, University of
AdelaideJohn Curtin School of Medical Research, Australian National
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/smped/extraChIPs/blob/HEAD/vignettes/differential_signal_sliding.Rmd" class="external-link"><code>vignettes/differential_signal_sliding.Rmd</code></a></small>
      <div class="hidden name"><code>differential_signal_sliding.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>message <span class="op">=</span> <span class="cn">FALSE</span>, crop <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <a href="https://github.com/smped/GRAVI" class="external-link">GRAVI</a> workflow, for
which this package is designed, uses sliding windows for differential
signal analysis using strategies defined in the package
<code>csaw</code> (Lun and Smyth 2016). The GRAVI workflow itself
extends to integrating multiple ChIP targets and external data sources,
and as such, this package introduces a handful of functions to simplify
and enable these analyses. Whilst many existing approaches refer to this
type of analysis as Differential Binding analysis, we prefer the term
<em>Differential Signal Analysis</em> as this more accurately captures
the range of ChIP targets which are likely to be investigated, such as
H3K27ac or similar.</p>
<p>The majority of examples below use heavily reduced datasets to
provide general guidance on using the functions. Some results may appear
trivial as a result, but will hopefully prove far more useful in a true
experimental context. All data, along with this vignette are available
<a href="https://github.com/smped/extraChIPs_vignette" class="external-link">here</a>. In
order to replicate this vignette, please place all contents of the
<code>data</code> directory provided in this additional repository into
a directory named data in your own working directory.</p>
<p>It may also be worth noting that this vignette should run on a
conventional laptop with no particularly large performance requirements.
However, when following this workflow across an entire genome, memory
requirements may exceed those of a standard laptop, and an HPC or
high-performance workstation may be required.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>In order to use the package <code>extraChIPs</code> and follow this
vignette, we recommend using the package <code>BiocManager</code>
(hosted on CRAN) to install all packages. Once <code>BiocManager</code>
is installed, the additional packages required for this vignette can
also be installed as shown below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="va">pkg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"tidyverse"</span>, <span class="st">"Rsamtools"</span>, <span class="st">"csaw"</span>, <span class="st">"BiocParallel"</span>, <span class="st">"rtracklayer"</span>, <span class="st">"edgeR"</span>, </span>
<span>  <span class="st">"patchwork"</span>, <span class="st">"extraChIPs"</span>, <span class="st">"plyranges"</span>, <span class="st">"scales"</span>, <span class="st">"glue"</span>, <span class="st">"here"</span>, <span class="st">"quantro"</span>,</span>
<span>  <span class="st">"cqn"</span>, <span class="st">"ggrepel"</span>, <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="va">pkg</span>, update <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Once these packages are installed, we can load them easily</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csaw</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/tidyverse/glue" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">quantro</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">cqn</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h3>
<p>All data for this vignette is expected to be in a sub-directory of
the working directory named “data” (including “data/H3K27ac”), and all
paths will be predicated on this. Please ensure you have all data in
this location, obtained from <a href="https://github.com/smped/extraChIPs_vignette" class="external-link">here</a>.</p>
<p>The data itself is ChIP-Seq data targeting the histone mark H3K27ac,
and is taken from the ZR-75-1 cell-line using data from the BioProject ,
Preprocessing was performed using the <a href="https://github.com/smped/prepareChIPs" class="external-link"><code>prepareChIPs</code></a>
workflow, written in snakemake (Mölder et al. 2021) and all code is
available at <a href="https://github.com/smped/PRJNA509779" class="external-link uri">https://github.com/smped/PRJNA509779</a>. H3K27ac signal was
obtained under baseline (E2) and DHT-treated (E2DHT) conditions, with
alignments being to GRCh37. For this workflow, data has been subset to a
region found on chromosome 10 for simplicity.</p>
<p>A description of the samples can be found in the file
<code>data/PRJNA509779.tsv</code>. A pooled Input (IgG) sample was used
for the entire dataset and this accession is provided in a separate
column.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"PRJNA509779.tsv"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">target</span> <span class="op">==</span> <span class="st">"H3K27ac"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">treatment</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"E2"</span>, <span class="st">"E2DHT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">accessions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">accession</span>, <span class="va">samples</span><span class="op">$</span><span class="va">input</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">treat_colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"steelblue"</span>, <span class="st">"red3"</span>, <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">treat_colours</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span>, <span class="st">"Input"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="differential-signal-analysis">Differential Signal Analysis<a class="anchor" aria-label="anchor" href="#differential-signal-analysis"></a>
</h2>
<div class="section level3">
<h3 id="sliding-windows">Sliding Windows<a class="anchor" aria-label="anchor" href="#sliding-windows"></a>
</h3>
<p>The standard approach of tools such as DiffBind (Ross-Innes et
al. 2012) is to take a set of peaks, re-centre them, then set all
regions to be the same width. From there, data is passed to
<code>edgeR</code> (Chen, Lun, and Smyth 2016) or <code>DESeq2</code>
(Love, Huber, and Anders 2014) for analysis. This approach using
<code>extraChIPs</code> is demonstrated in a separate vignette, however,
this workflow instead focusses on a sliding window approach as per the
package <code>csaw</code> (Lun and Smyth 2016). The resultant
<em>variable width regions</em> can be particularly advantageous for
ChIP targets such as H3K27ac where regions marked by histone-acetylation
can vary greatly in size.</p>
<p>The starting point for differential signal analyses using
<code>extraChIPs</code> is to define a set of sliding windows across the
genome, then count reads from a set of bam files, defined as a
<code>BamFileList.</code> Commonly one or more IP input/control samples
is also produced during a ChIP-Seq experiment, and these should be
included at this stage of the analysis. The example files provided here
contain a small subset of reads from chromosome 10 across two
experimental conditions and one input sample, and we will define them
all as a <code>BamFileList</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"H3K27ac"</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{accessions}.bam"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove_all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>, <span class="st">".bam"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">sortSeqlevels</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">isCircular</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">genome</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GRCh37"</span></span></code></pre></div>
<p><strong>NB:</strong> It should also be noted that counting all reads
across a <code>BamFileList</code> using sliding windows, <strong>will
require a significant amount of RAM</strong> and will be beyond the
capacity of most laptops as of the time of writing. When working with
complete datasets, this particular step is best performed on an HPC or a
similar interactive server with a large amount of memory.</p>
<p>The approach taken below is to first define a set of sliding windows
across the genome, using the capabilities of <code>csaw</code>. After
counting reads across all windows, a set of <em>pre-defined regions</em>
where H3K27ac signal is confidently found, is then used guide the
function <code><a href="../reference/dualFilter.html">dualFilter()</a></code>. This function will discard
low-signal windows, retaining only those a) above a minimum signal level
and b) with signal notably above that of any input samples. These
regions can be obtained from any external resource, or can even be taken
from <code>macs2</code>-defined peaks from the same samples. Here, the
set of regions was defined as those found when merging samples from each
treatment, in both treatments, with an FDR &lt; 0.01. Importantly, our
analysis will not be restricted to these regions, but these will instead
guide <code><a href="../reference/dualFilter.html">dualFilter()</a></code> as to the best regions to retain for
analysis.</p>
<p>First we can define our windows and count the alignments using the
existing capabilities and functions provided in the <code>csaw</code>
package. In the following, we’ll use a sliding window of 120bp and a
step size of 40bp, meaning each nucleotide is covered by 3 windows. In
addition, we’ll exclude blacklisted and grey-listed regions as provided
in the dataset. These can be obtained easily by using the
<code>GreyListChIP</code> package, which is beyond the scope of this
vignette, however, code for preparing the GreyList` is available <a href="https://github.com/smped/PRJNA509779/blob/main/workflow/scripts/make_greylist.R" class="external-link">here</a>.
Fragment sizes for each sample were also estimated to be around
200nt.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">greylist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"chr10_greylist.bed"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span>seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span></span>
<span><span class="va">blacklist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"chr10_blacklist.bed"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span>seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span></span>
<span><span class="va">rp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/readParam.html" class="external-link">readParam</a></span><span class="op">(</span></span>
<span>  pe <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  dedup <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  restrict <span class="op">=</span> <span class="st">"chr10"</span>,</span>
<span>  discard <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">greylist</span>, <span class="va">blacklist</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">wincounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/csaw/man/windowCounts.html" class="external-link">windowCounts</a></span><span class="op">(</span></span>
<span>  bam.files <span class="op">=</span> <span class="va">bfl</span>,</span>
<span>  spacing <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">120</span>,</span>
<span>  ext <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">bfl</span><span class="op">)</span>,</span>
<span>  param <span class="op">=</span> <span class="va">rp</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sq</span></span></code></pre></div>
<p>This produces a <code>RangedSummarizedExperiment</code> with windows
included which passed the minimum threshold of 7 total reads. We can
check which windows passed this threshold using
<code>rowRanges()</code></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 506538 ranges and 0 metadata columns:</span></span>
<span><span class="co">##            seqnames            ranges strand</span></span>
<span><span class="co">##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]    chr10 42491801-42491920      *</span></span>
<span><span class="co">##        [2]    chr10 42491921-42492040      *</span></span>
<span><span class="co">##        [3]    chr10 42494601-42494720      *</span></span>
<span><span class="co">##        [4]    chr10 42495641-42495760      *</span></span>
<span><span class="co">##        [5]    chr10 42495681-42495800      *</span></span>
<span><span class="co">##        ...      ...               ...    ...</span></span>
<span><span class="co">##   [506534]    chr10 99999641-99999760      *</span></span>
<span><span class="co">##   [506535]    chr10 99999681-99999800      *</span></span>
<span><span class="co">##   [506536]    chr10 99999721-99999840      *</span></span>
<span><span class="co">##   [506537]    chr10 99999761-99999880      *</span></span>
<span><span class="co">##   [506538]    chr10 99999801-99999920      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<p>We can also add some key information to the <code>colData</code>
element of this object, which will also be propagated to all downstream
objects.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 7 rows and 4 columns</span></span>
<span><span class="co">##                         bam.files    totals       ext      rlen</span></span>
<span><span class="co">##                       &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">## SRR8315186 /home/steviep/github..    267342       200        75</span></span>
<span><span class="co">## SRR8315187 /home/steviep/github..    307383       200        75</span></span>
<span><span class="co">## SRR8315188 /home/steviep/github..    288965       200        75</span></span>
<span><span class="co">## SRR8315189 /home/steviep/github..    296764       200        75</span></span>
<span><span class="co">## SRR8315190 /home/steviep/github..    325211       200        75</span></span>
<span><span class="co">## SRR8315191 /home/steviep/github..    283745       200        75</span></span>
<span><span class="co">## SRR8315192 /home/steviep/github..    361835       200        75</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"accession"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">samples</span>, by <span class="op">=</span> <span class="st">"accession"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">accession</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="va">target</span>, <span class="va">treatment</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    target <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace_na.html" class="external-link">str_replace_na</a></span><span class="op">(</span><span class="va">target</span>, <span class="st">"Input"</span><span class="op">)</span>,</span>
<span>    treatment <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace_na.html" class="external-link">str_replace_na</a></span><span class="op">(</span><span class="va">treatment</span>, <span class="st">"Input"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">DataFrame</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">set_rownames</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">accession</span><span class="op">)</span></span></code></pre></div>
<p>A density plot can be simply drawn of these counts, with the vast
majority of windows receiving very low counts, due to the nature of
transcription factor binding, where long stretches are unbound. The
windows with higher counts tend to be associated with the samples
targeting a transcription factor (TF), as seen in the two treatment
group samples.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">wincounts</span>, colour <span class="op">=</span> <span class="st">"treatment"</span>, trans <span class="op">=</span> <span class="st">"log1p"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filtering-of-sliding-windows">Filtering of Sliding Windows<a class="anchor" aria-label="anchor" href="#filtering-of-sliding-windows"></a>
</h3>
<p>After counting all reads in the sliding genomic windows, the next
step is to discard windows for which counts are unlikely to represent
true signal from our ChIP target. The strategy employed in
<code>extraChIPs</code> uses a set of pre-defined regions to
automatically set thresholds based on 1) counts strongly above the
counts from the input sample, and 2) the windows with the overall
highest signal. Thresholds are determined such that a proportion
(e.g. <code>q = 0.5</code>) of the windows which overlap one of the
supplied consensus peaks will be returned. Higher values for
<code>q</code> will return more windows, however many of these will tend
to only marginally overlap a peak in one of the tail regions, and these
will most likely be covered by neighbouring windows. Experience has
shown that values such as <code>q = 0.5</code> tend to return a
considerable proportion of windows containing true signal from the ChIP
target. Higher values will tend to return more sliding windows where the
edges overlap the tails of the guide regions.</p>
<p>The we can pass these to the function <code><a href="../reference/dualFilter.html">dualFilter()</a></code> which
utilises the strategy described above. On large datasets, this can be
quite time-consuming, as can the initial counting step. Multiple
alternative filtering strategies are also provided by the package
<code>csaw</code> and these can be accessed using
<code><a href="https://rdrr.io/pkg/csaw/man/defunct.html" class="external-link">?csaw::filterWindows</a></code></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">guide_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"H3K27ac"</span>, <span class="st">"H3K27ac_chr10.bed"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html" class="external-link">import.bed</a></span><span class="op">(</span>seqinfo <span class="op">=</span> <span class="va">sq</span><span class="op">)</span></span>
<span><span class="va">filtcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dualFilter.html">dualFilter</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="va">wincounts</span><span class="op">$</span><span class="va">target</span> <span class="op">==</span> <span class="st">"H3K27ac"</span><span class="op">]</span>,</span>
<span>  bg <span class="op">=</span> <span class="va">wincounts</span><span class="op">[</span>, <span class="va">wincounts</span><span class="op">$</span><span class="va">target</span> <span class="op">==</span> <span class="st">"Input"</span><span class="op">]</span>, </span>
<span>  ref <span class="op">=</span> <span class="va">guide_regions</span>,</span>
<span>  q <span class="op">=</span> <span class="fl">0.6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Thus we have reduced our initial set of 506,538 sliding windows to
the 23,417 windows most likely to contain true signal from our ChIP
target. The returned object will by default contain <code>counts</code>
and <code>logCPM</code> assays, with the complete library sizes used for
the calculation of <code>logCPM</code> values. Similarly, <em>the input
sample is no longer included</em> in the data object, although
additional columns can easily be added to the returned object using any
number of strategies.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">wincounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 506538      7</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 23417     6</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): counts logCPM</span></span></code></pre>
<p>The <code>rowData</code> element of the returned object will contain
a logical column indicating where each specific retained window
overlapped one of the supplied consensus peaks.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 23417 ranges and 1 metadata column:</span></span>
<span><span class="co">##           seqnames            ranges strand | overlaps_ref</span></span>
<span><span class="co">##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;logical&gt;</span></span>
<span><span class="co">##       [1]    chr10 42862921-42863040      * |         TRUE</span></span>
<span><span class="co">##       [2]    chr10 42862961-42863080      * |         TRUE</span></span>
<span><span class="co">##       [3]    chr10 42863001-42863120      * |         TRUE</span></span>
<span><span class="co">##       [4]    chr10 42863041-42863160      * |         TRUE</span></span>
<span><span class="co">##       [5]    chr10 42863081-42863200      * |         TRUE</span></span>
<span><span class="co">##       ...      ...               ...    ... .          ...</span></span>
<span><span class="co">##   [23413]    chr10 99895121-99895240      * |         TRUE</span></span>
<span><span class="co">##   [23414]    chr10 99895161-99895280      * |         TRUE</span></span>
<span><span class="co">##   [23415]    chr10 99895201-99895320      * |         TRUE</span></span>
<span><span class="co">##   [23416]    chr10 99895241-99895360      * |         TRUE</span></span>
<span><span class="co">##   [23417]    chr10 99895281-99895400      * |         TRUE</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="initial-visualisation">Initial Visualisation<a class="anchor" aria-label="anchor" href="#initial-visualisation"></a>
</h3>
<p>We can once again check our signal distributions, this time on the
logCPM values. We can also check the samples using a PCA plot, again
colouring the points by treatment group and adding labels, which will
repel by default if the points are shown.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAssayDensities-methods.html">plotAssayDensities</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treatment"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAssayPCA-methods.html">plotAssayPCA</a></span><span class="op">(</span><span class="va">filtcounts</span>, <span class="st">"logCPM"</span>, colour <span class="op">=</span> <span class="st">"treatment"</span>, label <span class="op">=</span> <span class="st">"accession"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">A</span> <span class="op">+</span> <span class="va">B</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-density-pca-1.png" alt="*A) Densities for logCPM values across all samples after discarding windows less likely to contain H3K27ac signal using `dualFilter()` B) PCA plot based on the logCPM assay*" width="480"><p class="caption">
<em>A) Densities for logCPM values across all samples after discarding
windows less likely to contain H3K27ac signal using
<code><a href="../reference/dualFilter.html">dualFilter()</a></code> B) PCA plot based on the logCPM assay</em>
</p>
</div>
<p>Another common QC step is Relative Log-Expression (RLE) (Gandolfo and
Speed 2018). In the following, we first inspect the RLE across the
entire dataset, followed by RLE grouping <em>within treatments</em>.
This can be particularly useful when distributions vary significantly
between treatment groups, such as may occur with a cytoplasmic to
nuclear shift by a given ChIP target. Here, however, there is minimal
difference between the two approaches as H3K27ac signal tends to be
broadly consistent between these treatment groups.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAssayRle-methods.html">plotAssayRle</a></span><span class="op">(</span><span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, fill <span class="op">=</span> <span class="st">"treat"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, colour <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"RLE: Across All Samples"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotAssayRle-methods.html">plotAssayRle</a></span><span class="op">(</span></span>
<span>  <span class="va">filtcounts</span>, assay <span class="op">=</span> <span class="st">"logCPM"</span>, fill <span class="op">=</span> <span class="st">"treat"</span>, rle_group <span class="op">=</span> <span class="st">"treat"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, colour <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">treat_colours</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"RLE: Within Treatment Groups"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">+</span> <span class="va">b</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-assay-rle-1.png" alt="*RLE plots (A) across all samples and (B) with values calculated within treatment groups*" width="384"><p class="caption">
<em>RLE plots (A) across all samples and (B) with values calculated
within treatment groups</em>
</p>
</div>
</div>
<div class="section level3">
<h3 id="statistical-testing">Statistical Testing<a class="anchor" aria-label="anchor" href="#statistical-testing"></a>
</h3>
<p>Multiple methods are enabled in the package <code>extraChIPs</code>
via the function <code><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff()</a></code>, with the possibility of
incorporating any additional normalisation strategies from external
packages. The two basic strategies are 1) Quasi-Likelihood Fits (Lund et
al. 2012) and 2) <code>limma-trend</code> (Law et al. 2014). The first
approach (method = “qlf”) uses counts along with any of the
normalisation strategies from <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code>, and
setting the normalisation method to “none” is the equivalent of
library-size normalisation, which replicates the default normalisation
strategy from DiffBind (Ross-Innes et al. 2012). If choosing to
normalise <em>within treatment groups</em>, a factor can be provided via
the groups argument, essentially adding this as an option for all
methods provided in <code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code>. The second
method (method = “lt”) is specifically for logCPM values and these can
be provided as output by <code><a href="../reference/dualFilter.html">dualFilter()</a></code> or may be normalised
using any number of additional methods. In addition to the above
methods, a range-based <span class="math inline">\(H_0\)</span>
(McCarthy and Smyth 2009) can be specified by providing a value to the
<code>fc</code> or <code>lfc</code> arguments, which removes any need
for <em>post-hoc</em> filtering and correctly controls the FDR, unlike
<em>post-hoc</em> filtering based on a fold-change threshold.</p>
<p>Here, we’ll initially fit our data using Quasi-Likelihood Fits,
library-size normalisation and setting a change in signal beyond the
range of <span class="math inline">\(\pm\)</span> 20% as being of
interest. By default, the returned object, would be a
<code>SummarizedExperiment</code> object containing the results from
model fitting in the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> element. However, setting
<code>asRanges = TRUE</code> will simply return the set of GRanges along
with the testing results as a stand-alone object.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">treatment</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">filtcounts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff</a></span><span class="op">(</span><span class="va">filtcounts</span>, design <span class="op">=</span> <span class="va">X</span>, fc <span class="op">=</span> <span class="fl">1.2</span>, asRanges <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-windows">Merging Windows<a class="anchor" aria-label="anchor" href="#merging-windows"></a>
</h3>
<p>After an analysis has been performed, values contained in the output
will be estimated signal (<code>logCPM</code>), estimated change
(<code>logFC</code>) with both raw and adjusted p-values for all sliding
windows. Given the dependency of neighbouring windows, any adjusted
p-values will not be appropriate and a merging of overlapping and/or
neighbouring windows should be performed. Multiple <code>csaw</code>
methods are wrapped using <code><a href="../reference/mergeByCol-methods.html">mergeByCol()</a></code>,
<code><a href="../reference/mergeBySig-methods.html">mergeBySig()</a></code> with minor changes to the returned object,
such as the inclusion of the representative range in the column
<code>keyval_range</code>.</p>
<p>For this vignette, we’ll merge using the asymptotically exact
harmonic mean p-value, which can also be used for merging dependent
p-values (Wilson 2019). This approach tests H<sub>0</sub>: no <span class="math inline">\(p\)</span>-value in the set of p-values is
significant. When merging windows using the harmonic mean p-values,
instead of values from a representative window, weighted averages for
the expression and logFC estimates are returned using the weights <span class="math inline">\(w_i = \frac{1}{p_i}\)</span>. A representative
window, corresponding to the original window with the lowest p-value is
returned.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeByHMP-methods.html">mergeByHMP</a></span><span class="op">(</span><span class="va">fit_gr</span>, inc_cols <span class="op">=</span> <span class="st">"overlaps_ref"</span>, merge_within <span class="op">=</span> <span class="fl">120</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    <span class="va">status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>      <span class="va">hmp_fdr</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">~</span> <span class="st">"Unchanged"</span>,</span>
<span>      <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Increased"</span>,</span>
<span>      <span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Decreased"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">results_gr</span>, <span class="va">hmp</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 5 ranges and 10 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand | n_windows      n_up    n_down</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]    chr10 79257441-79258720      * |        30         3         0</span></span>
<span><span class="co">##   [2]    chr10 43689241-43690160      * |        21         2         0</span></span>
<span><span class="co">##   [3]    chr10 74008161-74008760      * |        13         2         0</span></span>
<span><span class="co">##   [4]    chr10 63858841-63859040      * |         3         1         0</span></span>
<span><span class="co">##   [5]    chr10 63859401-63859760      * |         7         1         0</span></span>
<span><span class="co">##       overlaps_ref            keyval_range    logCPM     logFC         hmp</span></span>
<span><span class="co">##          &lt;logical&gt;               &lt;GRanges&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]         TRUE chr10:79257761-79257880   7.02801   1.83161 1.84310e-12</span></span>
<span><span class="co">##   [2]         TRUE chr10:43689681-43689800   6.50461   1.78769 8.00605e-09</span></span>
<span><span class="co">##   [3]         TRUE chr10:74008441-74008560   6.49284   1.81722 9.04613e-09</span></span>
<span><span class="co">##   [4]         TRUE chr10:63858881-63859000   5.99283   1.75275 5.91544e-07</span></span>
<span><span class="co">##   [5]         TRUE chr10:63859601-63859720   5.99405   1.82278 9.46455e-07</span></span>
<span><span class="co">##           hmp_fdr status.......</span></span>
<span><span class="co">##         &lt;numeric&gt;   &lt;character&gt;</span></span>
<span><span class="co">##   [1] 1.32519e-09     Increased</span></span>
<span><span class="co">##   [2] 2.16805e-06     Increased</span></span>
<span><span class="co">##   [3] 2.16805e-06     Increased</span></span>
<span><span class="co">##   [4] 1.06330e-04     Increased</span></span>
<span><span class="co">##   [5] 1.36100e-04     Increased</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<p>In the above, we returned 19 ranges which we might consider using the
significance threshold <span class="math inline">\(\alpha\)</span> =
0.05. As is common, we can assess our results using an MA plot. However,
given that testing is performed using sliding windows and merging
windows using the harmonic mean will introduce a bias, we can check
using the complete set of sliding windows and the final set of merged
windows. Any bias present in our data will be visible when using the
sliding windows, whilst our final results can be inspected using the
merged windows.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">A</span> <span class="op">&lt;-</span> <span class="va">fit_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logCPM</span>, <span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>se <span class="op">=</span> <span class="cn">FALSE</span>, method <span class="op">=</span> <span class="st">"loess"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>        n <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        logCPM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">logCPM</span><span class="op">)</span> <span class="op">-</span> <span class="fl">0.15</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">logCPM</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        logFC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">-</span> <span class="fl">0.05</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        label <span class="op">=</span> <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{comma(n)}\nSliding Windows"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">fit_gr</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MA Plot: All Retained Sliding Windows"</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="va">results_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>`FDR &lt; 0.05` <span class="op">=</span> <span class="va">hmp_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logCPM</span>, <span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">`FDR &lt; 0.05`</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>se <span class="op">=</span> <span class="cn">FALSE</span>, method <span class="op">=</span> <span class="st">"loess"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">range</span>, colour <span class="op">=</span> <span class="va">`FDR &lt; 0.05`</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hmp_fdr</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">hmp_fdr</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>        n <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        logCPM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">logCPM</span><span class="op">)</span> <span class="op">-</span> <span class="fl">0.15</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">logCPM</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        logFC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">fit_gr</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span> <span class="op">-</span> <span class="fl">0.05</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">fit_gr</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        label <span class="op">=</span> <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{comma(n)}\nMerged Windows"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">fit_gr</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MA Plot: Merged Windows"</span><span class="op">)</span></span>
<span><span class="va">A</span> <span class="op">+</span> <span class="va">B</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-ma-libsize-1.png" alt="*MA Plots using A) all sliding windows (after filtering before merging) and B) the final set of merged windows. The most highly ranked merged window is also labelled in the set of merged windows. Blue lines represent a loess curve through the data.*" width="384"><p class="caption">
<em>MA Plots using A) all sliding windows (after filtering before
merging) and B) the final set of merged windows. The most highly ranked
merged window is also labelled in the set of merged windows. Blue lines
represent a loess curve through the data.</em>
</p>
</div>
<p>A particularly beneficial feature of this approach is that the final
ranges will be of highly variable width, with this select region of
chromosome 10 producing merged windows ranging from 120bp to 18.88kb, as
may be expected for H3K27ac signal.</p>
</div>
<div class="section level3">
<h3 id="alternative-normalisation-approaches">Alternative Normalisation Approaches<a class="anchor" aria-label="anchor" href="#alternative-normalisation-approaches"></a>
</h3>
<p>Using Library Size normalisation, as above, is a conservative
approach and other methods such as RLE (Love, Huber, and Anders 2014) or
TMM (Robinson and Oshlack 2010) may be appropriate. However, these
methods assume that signal distributions are drawn from the same
distributions across all samples, which unlike most RNA-Seq, is not
always able to be safely assumed for ChIP-Seq data. An example may be
when an transcription factor is primarily cytoplasmic in one treatment
group, before being translocated to the nucleus in another treatment.
The Androgen Receptor (<em>AR</em>) is a common example of this
behaviour.</p>
<p>Whilst H3K27ac signal should remain approximately constant between
these two treatments, we can easily check for differences in data
distributions as well as any GC-bias. The package <code>quantro</code>
(Hicks and Irizarry 2015) provides a useful test for any difference in
medians as well as differences in the underlying distributions.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">qtest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">filtcounts</span>, <span class="st">"counts"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/quantro/man/quantro.html" class="external-link">quantro</a></span><span class="op">(</span>groupFactor <span class="op">=</span> <span class="va">filtcounts</span><span class="op">$</span><span class="va">treatment</span>, B <span class="op">=</span> <span class="fl">1e3</span><span class="op">)</span></span>
<span><span class="va">qtest</span></span></code></pre></div>
<pre><code><span><span class="co">## quantro: Test for global differences in distributions</span></span>
<span><span class="co">##    nGroups:  2 </span></span>
<span><span class="co">##    nTotSamples:  6 </span></span>
<span><span class="co">##    nSamplesinGroups:  3 3 </span></span>
<span><span class="co">##    anovaPval:  0.49809 </span></span>
<span><span class="co">##    quantroStat:  0.19728 </span></span>
<span><span class="co">##    quantroPvalPerm:  0.68</span></span></code></pre>
<p>Here, no differences were evident across median values (ANOVA p =
0.498) or between distributions (p = 0.68) and as such, TMM/RLE
normalisation across all samples may be appropriate.</p>
<div class="section level4">
<h4 id="tmm-normalisation">TMM Normalisation<a class="anchor" aria-label="anchor" href="#tmm-normalisation"></a>
</h4>
<p>To perform an analysis using TMM-normalisation, we can simply specify
this using the argument <code>norm = "TMM"</code> when calling
<code><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff()</a></code></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitAssayDiff-methods.html">fitAssayDiff</a></span><span class="op">(</span></span>
<span>  <span class="va">filtcounts</span>, design <span class="op">=</span> <span class="va">X</span>, fc <span class="op">=</span> <span class="fl">1.2</span>, norm <span class="op">=</span> <span class="st">"TMM"</span>, asRanges <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">tmm_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeByHMP-methods.html">mergeByHMP</a></span><span class="op">(</span><span class="va">tmm_gr</span>, inc_cols <span class="op">=</span> <span class="st">"overlaps_ref"</span>, merge_within <span class="op">=</span> <span class="fl">120</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    status <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>      <span class="va">hmp_fdr</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">~</span> <span class="st">"Unchanged"</span>,</span>
<span>      <span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Decreased"</span>,</span>
<span>      <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">~</span> <span class="st">"Increased"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">tmm_results</span><span class="op">$</span><span class="va">status</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Increased Unchanged </span></span>
<span><span class="co">##        21       698</span></span></code></pre>
<p>As might be expected, the results are highly concordant, with
TMM-normalisation providing a moderate improvement in statistical power,
returning 21 windows with evidence of differential signal, instead of
the initial 19. Any of the normalisation methods taken by
<code><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">edgeR::calcNormFactors()</a></code> can be used here.</p>
<p>If a difference in distributions is found between groups, the
normalisation step can be set to only occur within treatment groups, by
passing the grouping column name to the <code>groups =</code> argument.
However, this may exaggerate differences between groups and may
introduce false positives, so should only be used when compelling
evidence exists and the underlying biology supports this.</p>
<p>Checking MA-plots shows that the very slight negative bias in
higher-signal windows when using library size normalisation, as seen
above, was lessened after TMM normalisation.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logCPM</span>, <span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>se <span class="op">=</span> <span class="cn">FALSE</span>, method <span class="op">=</span> <span class="st">"loess"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MA Plot: TMM Normalisation"</span><span class="op">)</span></span></code></pre></div>
<p>If choosing more nuanced normalisation strategies, passing any offset
matrices can be simply done using the <code>offset</code> argument,
which will over-ride any other normalisation settings.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mapping-of-windows">Mapping of Windows<a class="anchor" aria-label="anchor" href="#mapping-of-windows"></a>
</h2>
<div class="section level3">
<h3 id="mapping-to-genes">Mapping to Genes<a class="anchor" aria-label="anchor" href="#mapping-to-genes"></a>
</h3>
<p>Once the changes in signal for our given ChIP target have been
determined, a common next step is to assess which genes are likely to be
impacted. Whilst no definitive, single methodology exists for this
process, the function <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> offers an intuitive
approach, taking into account any previously defined regulatory
features. These regulatory features may be defined by simple proximity
to TSS regions, by histone marks, downloaded from external repositories
or any other possibility. Whilst these features can improve the
precision of mapping, even without these this function can still enable
a useful assignment of target gene to binding event.</p>
<p>The process undertaken inside <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> is a
sequential checking of each range’s association with regulatory features
and the most likely target as a result. These steps are:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Check for any HiC interactions</strong>
<ul>
<li>All genes which directly overlap an interaction anchor are
considered part of the regulatory network for that interaction, and as
such, all genes associated with both anchors are assigned to a peak
which overlaps a HiC Interaction</li>
</ul>
</li>
<li>
<strong>Check for any overlaps with a promoter</strong>
<ul>
<li>All genes regulated by that promoter are assigned as regulatory
targets. By default, this is by direct promoter/gene overlap
(<code>prom2gene = 0</code>)</li>
</ul>
</li>
<li>
<strong>Check for any overlaps with an enhancer</strong>
<ul>
<li>Peaks which overlap an enhancer are assigned to <em>all</em> genes
within the distance specified by <code>enh2gene</code> (default =
100kb)</li>
</ul>
</li>
<li>
<strong>Check for genes with no previous mappings</strong>
<ul>
<li>Peaks <em>with no previous mappings</em> are assigned to all
directly overlapping genes, or the <em>nearest</em> gene within a
specified distance (default <code>gr2gene</code> = 100kb)</li>
</ul>
</li>
</ol>
<p>If no promoters, enhancers or long-range interactions are supplied,
all genes will be simply mapped to ChIP-Seq regions using step 4.</p>
<p>A GTF of annotations can be used, with the following being the genes,
transcripts and exons from chr10, taken from <a href="https://www.gencodegenes.org/human/release_43lift37.html" class="external-link">Gencode
v43</a>. Whilst this example relies on this format, any additional
resource such as an <code>EnsDb</code> or <code>TxDb</code> object could
easily be used. Here, we’ll load the data then split into a
<code>GRangesList</code> with separate elements for genes, transcripts
and exons.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gencode</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data/gencode.v43lift37.chr10.annotation.gtf.gz"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/GFFFile-class.html" class="external-link">import.gff</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/ranges-filter-overlaps.html" class="external-link">filter_by_overlaps</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr10:42354900-100000000"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">gencode</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">sq</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">gencode</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sq</span></span></code></pre></div>
<p>Now we’ve loaded our gene annotations at multiple levels, we can
easily define Transcription Start Sites (TSS) using transcript
definitions and the function <code><a href="../reference/reduceMC.html">reduceMC()</a></code>. This function is
an extension of <code><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">GenomicRanges::reduce()</a></code> which retains all
supplied columns in the <code>mcols()</code> element by collapsing these
into <code>CompressedList</code> columns, and simplifying to vectors
where possible.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tss</span> <span class="op">&lt;-</span> <span class="va">gencode</span><span class="op">$</span><span class="va">transcript</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">1</span>, fix <span class="op">=</span> <span class="st">"start"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">ends_with</a></span><span class="op">(</span><span class="st">"name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/reduceMC.html">reduceMC</a></span><span class="op">(</span>min.gapwidth <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Checking the TSS for <em>PPIF</em> shows we have one site which 3
transcripts start at, along with two more unique to specific
transcripts.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">tss</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">gene_name</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="st">"PPIF"</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">logical</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |              gene_id       gene_name</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;CharacterList&gt; &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]    chr10  81107225      + | ENSG00000108179.14_6            PPIF</span></span>
<span><span class="co">##   [2]    chr10  81107256      + | ENSG00000108179.14_6            PPIF</span></span>
<span><span class="co">##   [3]    chr10  81107414      + | ENSG00000108179.14_6            PPIF</span></span>
<span><span class="co">##                  transcript_name</span></span>
<span><span class="co">##                  &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1] PPIF-205,PPIF-203,PPIF-201</span></span>
<span><span class="co">##   [2]                   PPIF-204</span></span>
<span><span class="co">##   [3]                   PPIF-202</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<p>We can also add the TSS overlap status to our set of results</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_results</span><span class="op">$</span><span class="va">tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">overlapsAny</a></span><span class="op">(</span><span class="va">tmm_results</span>, <span class="va">tss</span><span class="op">)</span></span></code></pre></div>
<p>Next we’ll want to map our regions to genes. To use the methods
available in <code><a href="../reference/mapByFeature.html">mapByFeature()</a></code> let’s define a set of
promoters. First we’ll create a promoter for each transcript, then we
can merge any overlapping promoters using <code><a href="../reference/reduceMC.html">reduceMC()</a></code>.
(Setting <code>simplify = FALSE</code> avoids an issue in this subset of
genes where one gene has multiple IDs, and helps retain a 1:1 mapping
between IDs and genes.) Importantly, this will give a set of promoters
which are unique and non-overlapping but with variable width and with
the complete set of transcripts that may be regulated by the promoter
listed. As can be seen, 4 potential promoter regions are found for
<em>CRTAC1</em></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">promoters</span> <span class="op">&lt;-</span> <span class="va">gencode</span><span class="op">$</span><span class="va">transcript</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">ends_with</a></span><span class="op">(</span><span class="st">"name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">promoters</a></span><span class="op">(</span>upstream <span class="op">=</span> <span class="fl">2500</span>, downstream <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/reduceMC.html">reduceMC</a></span><span class="op">(</span>simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">promoters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand |</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |</span></span>
<span><span class="co">##   [1]    chr10 99609056-99612055      - |</span></span>
<span><span class="co">##   [2]    chr10 99635155-99638154      - |</span></span>
<span><span class="co">##   [3]    chr10 99643500-99646805      - |</span></span>
<span><span class="co">##   [4]    chr10 99695536-99698535      - |</span></span>
<span><span class="co">##   [5]    chr10 99770595-99773594      - |</span></span>
<span><span class="co">##   [6]    chr10 99789879-99793085      - |</span></span>
<span><span class="co">##                                           gene_id       gene_name</span></span>
<span><span class="co">##                                   &lt;CharacterList&gt; &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]                        ENSG00000227356.2_10      GOLGA7B-DT</span></span>
<span><span class="co">##   [2]                           ENSG00000265398.1      AL139239.1</span></span>
<span><span class="co">##   [3] ENSG00000095713.14_13,ENSG00000095713.14_13   CRTAC1,CRTAC1</span></span>
<span><span class="co">##   [4]                       ENSG00000095713.14_13          CRTAC1</span></span>
<span><span class="co">##   [5]                       ENSG00000095713.14_13          CRTAC1</span></span>
<span><span class="co">##   [6] ENSG00000095713.14_13,ENSG00000095713.14_13   CRTAC1,CRTAC1</span></span>
<span><span class="co">##             transcript_name</span></span>
<span><span class="co">##             &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]        GOLGA7B-DT-201</span></span>
<span><span class="co">##   [2]        AL139239.1-201</span></span>
<span><span class="co">##   [3] CRTAC1-205,CRTAC1-206</span></span>
<span><span class="co">##   [4]            CRTAC1-204</span></span>
<span><span class="co">##   [5]            CRTAC1-201</span></span>
<span><span class="co">##   [6] CRTAC1-203,CRTAC1-202</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mapByFeature.html">mapByFeature</a></span><span class="op">(</span></span>
<span>  <span class="va">tmm_results</span>, genes <span class="op">=</span> <span class="va">gencode</span><span class="op">$</span><span class="va">gene</span>,</span>
<span>  prom <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">promoters</span>, <span class="va">gene_id</span>, <span class="va">gene_name</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">tmm_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hmp_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">hmp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 21 ranges and 13 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand | n_windows      n_up    n_down</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]    chr10 79257441-79258720      * |        30         3         0</span></span>
<span><span class="co">##    [2]    chr10 43689241-43690160      * |        21         2         0</span></span>
<span><span class="co">##    [3]    chr10 74008161-74008760      * |        13         2         0</span></span>
<span><span class="co">##    [4]    chr10 63858841-63859040      * |         3         1         0</span></span>
<span><span class="co">##    [5]    chr10 63859401-63859760      * |         7         1         0</span></span>
<span><span class="co">##    ...      ...               ...    ... .       ...       ...       ...</span></span>
<span><span class="co">##   [17]    chr10 79307081-79307960      * |        20         2         0</span></span>
<span><span class="co">##   [18]    chr10 74014401-74015360      * |        22         3         0</span></span>
<span><span class="co">##   [19]    chr10 90615361-90615640      * |         4         2         0</span></span>
<span><span class="co">##   [20]    chr10 63696921-63697240      * |         6         3         0</span></span>
<span><span class="co">##   [21]    chr10 74878001-74878360      * |         7         3         0</span></span>
<span><span class="co">##        overlaps_ref            keyval_range    logCPM     logFC         hmp</span></span>
<span><span class="co">##           &lt;logical&gt;               &lt;GRanges&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]         TRUE chr10:79257761-79257880   7.02793   1.85621 6.20766e-13</span></span>
<span><span class="co">##    [2]         TRUE chr10:43689681-43689800   6.51086   1.81157 4.45187e-09</span></span>
<span><span class="co">##    [3]         TRUE chr10:74008441-74008560   6.49275   1.84270 5.07996e-09</span></span>
<span><span class="co">##    [4]         TRUE chr10:63858881-63859000   5.99257   1.77724 4.03575e-07</span></span>
<span><span class="co">##    [5]         TRUE chr10:63859601-63859720   5.99366   1.84777 6.37775e-07</span></span>
<span><span class="co">##    ...          ...                     ...       ...       ...         ...</span></span>
<span><span class="co">##   [17]         TRUE chr10:79307441-79307560   6.45271   1.15247 0.000563495</span></span>
<span><span class="co">##   [18]         TRUE chr10:74015201-74015320   6.14440   1.26391 0.000592892</span></span>
<span><span class="co">##   [19]         TRUE chr10:90615361-90615480   5.97947   1.14308 0.000804490</span></span>
<span><span class="co">##   [20]         TRUE chr10:63697121-63697240   6.00058   1.11473 0.001074487</span></span>
<span><span class="co">##   [21]         TRUE chr10:74878201-74878320   6.19539   1.05769 0.001437350</span></span>
<span><span class="co">##            hmp_fdr      status       tss               gene_id       gene_name</span></span>
<span><span class="co">##          &lt;numeric&gt; &lt;character&gt; &lt;logical&gt;       &lt;CharacterList&gt; &lt;CharacterList&gt;</span></span>
<span><span class="co">##    [1] 4.46331e-10   Increased     FALSE ENSG00000156113.25_17          KCNMA1</span></span>
<span><span class="co">##    [2] 1.21750e-06   Increased     FALSE ENSG00000198915.12_14        RASGEF1A</span></span>
<span><span class="co">##    [3] 1.21750e-06   Increased     FALSE   ENSG00000166295.9_6         ANAPC16</span></span>
<span><span class="co">##    [4] 7.25426e-05   Increased     FALSE ENSG00000150347.17_12          ARID5B</span></span>
<span><span class="co">##    [5] 9.17121e-05   Increased     FALSE ENSG00000150347.17_12          ARID5B</span></span>
<span><span class="co">##    ...         ...         ...       ...                   ...             ...</span></span>
<span><span class="co">##   [17]   0.0236827   Increased     FALSE ENSG00000156113.25_17          KCNMA1</span></span>
<span><span class="co">##   [18]   0.0236827   Increased     FALSE   ENSG00000289506.2_2 ENSG00000289506</span></span>
<span><span class="co">##   [19]   0.0304436   Increased     FALSE   ENSG00000152766.6_8         ANKRD22</span></span>
<span><span class="co">##   [20]   0.0386278   Increased     FALSE ENSG00000150347.17_12          ARID5B</span></span>
<span><span class="co">##   [21]   0.0492121   Increased     FALSE  ENSG00000166321.14_7          NUDT13</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<p>So now we have our regions showing differential signal, along with
which gene they are most likely to be impacting, and whether they
directly overlap the TSS. This can all be summarised into a single
MA-plot for presentation of final results.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">status_colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Unchanged <span class="op">=</span> <span class="st">"grey70"</span>, Increased <span class="op">=</span> <span class="st">"red3"</span>, Decreased <span class="op">=</span> <span class="st">"royalblue"</span><span class="op">)</span></span>
<span><span class="va">tmm_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logCPM</span>, <span class="va">logFC</span>, colour <span class="op">=</span> <span class="va">status</span>, shape <span class="op">=</span> <span class="va">tss</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>,</span>
<span>    data <span class="op">=</span> <span class="va">.</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">hmp</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        label  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">gene_name</span>, <span class="va">paste</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"; "</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>          <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_trunc.html" class="external-link">str_trunc</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>    fill <span class="op">=</span> <span class="st">"white"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">status_colours</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">21</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-ma-mapped-1.png" alt="*MA-plot with the top 20 regions by significance labelled according to the most likely gene influenced by the detected signal. Regions which directly overlap a TSS are shown as empty circles*" width="384"><p class="caption">
<em>MA-plot with the top 20 regions by significance labelled according
to the most likely gene influenced by the detected signal. Regions which
directly overlap a TSS are shown as empty circles</em>
</p>
</div>
</div>
<div class="section level3">
<h3 id="mapping-to-regions">Mapping to Regions<a class="anchor" aria-label="anchor" href="#mapping-to-regions"></a>
</h3>
<p>In addition to which gene is likely to be directly impacted by the
detected signal, knowing which type of regulatory region signal was
observed in is another question commonly asked of ChIP-Seq data. The
function <code>deinfeRegions()</code> takes, gene-, transcript- and
exon-level information from an annotation source and creates unique
region annotations across the genome. These are for in hierarchical
order as: 1) Promoters, 2) Upstream Promoters, 3) Exons, 4) Introns, 5)
Proximal Intergenic and, 6 Distal Intergenic regions.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/defineRegions.html">defineRegions</a></span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="va">gencode</span><span class="op">$</span><span class="va">gene</span>, transcripts <span class="op">=</span> <span class="va">gencode</span><span class="op">$</span><span class="va">transcript</span>, exons <span class="op">=</span> <span class="va">gencode</span><span class="op">$</span><span class="va">exon</span></span>
<span><span class="op">)</span></span>
<span><span class="va">regions</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 6:</span></span>
<span><span class="co">## $promoter</span></span>
<span><span class="co">## GRanges object with 1491 ranges and 5 metadata columns:</span></span>
<span><span class="co">##          seqnames            ranges strand |                region</span></span>
<span><span class="co">##             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |           &lt;character&gt;</span></span>
<span><span class="co">##      [1]    chr10 42644498-42647497      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##      [2]    chr10 42678287-42681286      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##      [3]    chr10 42702938-42705937      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##      [4]    chr10 42729924-42732923      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##      [5]    chr10 42735669-42738668      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##      ...      ...               ...    ... .                   ...</span></span>
<span><span class="co">##   [1487]    chr10 99695536-99698535      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##   [1488]    chr10 99770595-99773594      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##   [1489]    chr10 99789879-99793085      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##   [1490]    chr10 99891881-99894913      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##   [1491]    chr10 99920140-99923543      * | Promoter (-2500/+500)</span></span>
<span><span class="co">##                        gene_id       gene_name</span></span>
<span><span class="co">##                &lt;CharacterList&gt; &lt;CharacterList&gt;</span></span>
<span><span class="co">##      [1]   ENSG00000229485.1_5          KSR1P1</span></span>
<span><span class="co">##      [2]   ENSG00000237592.2_5     IGKV1OR10-1</span></span>
<span><span class="co">##      [3]   ENSG00000271650.1_7 ENSG00000271650</span></span>
<span><span class="co">##      [4]     ENSG00000264398.1      AL031601.1</span></span>
<span><span class="co">##      [5]   ENSG00000290458.1_2 ENSG00000290458</span></span>
<span><span class="co">##      ...                   ...             ...</span></span>
<span><span class="co">##   [1487] ENSG00000095713.14_13          CRTAC1</span></span>
<span><span class="co">##   [1488] ENSG00000095713.14_13          CRTAC1</span></span>
<span><span class="co">##   [1489] ENSG00000095713.14_13          CRTAC1</span></span>
<span><span class="co">##   [1490]  ENSG00000166024.14_8         R3HCC1L</span></span>
<span><span class="co">##   [1491]  ENSG00000166024.14_8         R3HCC1L</span></span>
<span><span class="co">##                                                            transcript_id</span></span>
<span><span class="co">##                                                          &lt;CharacterList&gt;</span></span>
<span><span class="co">##      [1]                                             ENST00000446298.1_2</span></span>
<span><span class="co">##      [2]                                             ENST00000442306.2_2</span></span>
<span><span class="co">##      [3]                                             ENST00000605702.1_2</span></span>
<span><span class="co">##      [4]                                               ENST00000580460.1</span></span>
<span><span class="co">##      [5]                                             ENST00000622823.4_3</span></span>
<span><span class="co">##      ...                                                             ...</span></span>
<span><span class="co">##   [1487]                                             ENST00000413387.5_4</span></span>
<span><span class="co">##   [1488]                                             ENST00000309155.3_5</span></span>
<span><span class="co">##   [1489]                        ENST00000370597.8_11,ENST00000370591.6_9</span></span>
<span><span class="co">##   [1490]                         ENST00000370584.7_7,ENST00000298999.8_8</span></span>
<span><span class="co">##   [1491] ENST00000370586.6_5,ENST00000613938.4_5,ENST00000612478.4_2,...</span></span>
<span><span class="co">##                                  transcript_name</span></span>
<span><span class="co">##                                  &lt;CharacterList&gt;</span></span>
<span><span class="co">##      [1]                              KSR1P1-201</span></span>
<span><span class="co">##      [2]                         IGKV1OR10-1-201</span></span>
<span><span class="co">##      [3]                         ENST00000605702</span></span>
<span><span class="co">##      [4]                          AL031601.1-201</span></span>
<span><span class="co">##      [5]                         ENST00000622823</span></span>
<span><span class="co">##      ...                                     ...</span></span>
<span><span class="co">##   [1487]                              CRTAC1-204</span></span>
<span><span class="co">##   [1488]                              CRTAC1-201</span></span>
<span><span class="co">##   [1489]                   CRTAC1-203,CRTAC1-202</span></span>
<span><span class="co">##   [1490]                 R3HCC1L-203,R3HCC1L-201</span></span>
<span><span class="co">##   [1491] R3HCC1L-204,R3HCC1L-206,R3HCC1L-205,...</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;5 more elements&gt;</span></span></code></pre>
<p>We can now use the function <code><a href="../reference/bestOverlap-methods.html">bestOverlap()</a></code> to assign each
window with ChIP-Seq signal to the region which it shows the greatest
overlap, setting this as a factor which reflects the hierarchy with
which the regions were defined.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmm_results</span><span class="op">$</span><span class="va">bestOverlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bestOverlap-methods.html">bestOverlap</a></span><span class="op">(</span><span class="va">tmm_results</span>, <span class="va">regions</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tmm_results</span>, <span class="va">hmp_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">bestOverlap</span> <span class="op">==</span> <span class="st">"promoter"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 14 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand | n_windows      n_up    n_down</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">##   [1]    chr10 76779561-76780200      * |        14         4         0</span></span>
<span><span class="co">##   [2]    chr10 79623321-79635960      * |       314        13         0</span></span>
<span><span class="co">##       overlaps_ref            keyval_range    logCPM     logFC         hmp</span></span>
<span><span class="co">##          &lt;logical&gt;               &lt;GRanges&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]         TRUE chr10:76779721-76779840   6.21638  1.244849 0.000194483</span></span>
<span><span class="co">##   [2]         TRUE chr10:79630321-79630440   7.71543  0.906496 0.000100434</span></span>
<span><span class="co">##          hmp_fdr      status       tss</span></span>
<span><span class="co">##        &lt;numeric&gt; &lt;character&gt; &lt;logical&gt;</span></span>
<span><span class="co">##   [1] 0.00932221   Increased     FALSE</span></span>
<span><span class="co">##   [2] 0.00515801   Increased      TRUE</span></span>
<span><span class="co">##                                         gene_id            gene_name</span></span>
<span><span class="co">##                                 &lt;CharacterList&gt;      &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]                     ENSG00000156650.14_17                KAT6B</span></span>
<span><span class="co">##   [2] ENSG00000204049.1_8,ENSG00000151208.17_11 ENSG00000204049,DLG5</span></span>
<span><span class="co">##       bestOverlap</span></span>
<span><span class="co">##          &lt;factor&gt;</span></span>
<span><span class="co">##   [1]    promoter</span></span>
<span><span class="co">##   [2]    promoter</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences from GRCh37 genome</span></span></code></pre>
<p>Given the region column contains nicely formateed output, we can also
use this</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reg_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">regions</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">region</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tmm_results</span><span class="op">$</span><span class="va">bestOverlap</span> <span class="op">&lt;-</span> <span class="va">tmm_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/bestOverlap-methods.html">bestOverlap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span>, var <span class="op">=</span> <span class="st">"region"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="va">reg_levels</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualisation-of-results">Visualisation of Results<a class="anchor" aria-label="anchor" href="#visualisation-of-results"></a>
</h2>
<div class="section level3">
<h3 id="pie-charts">Pie Charts<a class="anchor" aria-label="anchor" href="#pie-charts"></a>
</h3>
<p>Now that we’ve assigned each window to a genomic regions, we can use
<code><a href="../reference/plotPie-methods.html">plotPie()</a></code> to view the distribution of ChIP-Seq signal
within these regions. First we’ll define a consistent colour palette for
these regions in all plots.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_colours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">reg_levels</span><span class="op">)</span>, <span class="st">"Viridis"</span>, rev <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">region_colours</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">reg_levels</span></span></code></pre></div>
<p>Labels produced using <code><a href="../reference/plotPie-methods.html">plotPie()</a></code> use the
<code><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue()</a></code> syntax and are able to access the summarised data
using <code>N</code> as the overall total number of sites. as well as
<code>.data[[fill]]</code> to access data in the column selected to fill
the pie segments. Totals and percentages within each segment are
accessible using <code>n</code> and <code>p</code> respectively as these
are columns within the internal <code>data.frame</code> formed when
creating the plot. External formatting functions such as
<code><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma()</a></code>, <code><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent()</a></code> or
<code><a href="https://stringr.tidyverse.org/reference/case.html" class="external-link">str_to_title()</a></code> can also be put to good use here.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPie-methods.html">plotPie</a></span><span class="op">(</span></span>
<span>  <span class="va">tmm_results</span>, fill <span class="op">=</span> <span class="st">"bestOverlap"</span>, min_p <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  cat_glue <span class="op">=</span> <span class="st">"{.data[[fill]]}\n{n}\n({percent(p, 0.1)})"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">region_colours</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-pie-labels-1.png" alt="*Pie chart showing customised labels. Here the regions have been modified to use title case in the labels, but without modifying the underlying data.*" width="384"><p class="caption">
<em>Pie chart showing customised labels. Here the regions have been
modified to use title case in the labels, but without modifying the
underlying data.</em>
</p>
</div>
<p>In addition to highly customisable labels, segments can be scaled by
width to produce segments representing the overlaps in kb instead of
counts, by setting <code>scale_by = "width"</code></p>
</div>
<div class="section level3">
<h3 id="split-donut-charts">Split Donut Charts<a class="anchor" aria-label="anchor" href="#split-donut-charts"></a>
</h3>
<p>Pie charts can be further extended into Split-Donut charts, with data
in two rings. The default settings can be used to add the status of each
window to the overlapping regions.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSplitDonut-methods.html">plotSplitDonut</a></span><span class="op">(</span></span>
<span>  <span class="va">tmm_results</span>, inner <span class="op">=</span> <span class="st">"bestOverlap"</span>, outer <span class="op">=</span> <span class="st">"status"</span>,</span>
<span>  inner_palette <span class="op">=</span> <span class="va">region_colours</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-donut-default-1.png" alt="*Split-Donut chart showing status with overlapping region*" width="384"><p class="caption">
<em>Split-Donut chart showing status with overlapping region</em>
</p>
</div>
<p>A key advantage of <code><a href="../reference/plotSplitDonut-methods.html">plotSplitDonut()</a></code> is the ability to
provide separate colour palettes for the inner &amp; outer rings. Once
again these are heavily customisable using the <code><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue()</a></code>
syntax for labels and exploding key segments of interest.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSplitDonut-methods.html">plotSplitDonut</a></span><span class="op">(</span></span>
<span>  <span class="va">tmm_results</span>, inner <span class="op">=</span> <span class="st">"bestOverlap"</span>, outer <span class="op">=</span> <span class="st">"status"</span>, </span>
<span>  inner_palette <span class="op">=</span> <span class="va">region_colours</span>, outer_palette <span class="op">=</span> <span class="va">status_colours</span>,</span>
<span>  inner_glue <span class="op">=</span> <span class="st">"{str_replace_all(.data[[inner]], ' ', '\n')}\nn = {comma(n)}\n{percent(p,0.1)}"</span>,</span>
<span>  outer_glue <span class="op">=</span> <span class="st">"{.data[[outer]]}\n{str_replace_all(.data[[inner]], ' ', '\n')}\nn = {n}"</span>,</span>
<span>  explode_outer <span class="op">=</span> <span class="st">"(In|De)creased"</span>, explode_r <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>  outer_label <span class="op">=</span> <span class="st">"text"</span>, outer_min_p <span class="op">=</span> <span class="fl">0</span>, outer_max_p <span class="op">=</span> <span class="fl">0.02</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-donut-customised-1.png" alt="*Split-Donut chart exploding key regions and customising labels.*" width="384"><p class="caption">
<em>Split-Donut chart exploding key regions and customising labels.</em>
</p>
</div>
</div>
<div class="section level3">
<h3 id="coverage-plots">Coverage Plots<a class="anchor" aria-label="anchor" href="#coverage-plots"></a>
</h3>
<p>In order to show our changed signal in context we can show the
coverage using a <code>BigWigFileList</code> and the function
<code><a href="../reference/plotHFGC.html">plotHFGC()</a></code>, which stands for plot <strong>H</strong>iC,
<strong>F</strong>eatures, <strong>G</strong>enes and
<strong>C</strong>overage, with tracks appearing in that order.
<code><a href="../reference/plotHFGC.html">plotHFGC()</a></code> wraps multiple calls to the package
<code>Gviz</code> (Hahne and Ivanek 2016) to simplify repetitive plot
generation but retain significant flexibility. All tracks are options,
with the minimal requirements being cytogenetic bands along with a
specified GenomicRange. However, this does restrict plotting to the
“Standard Chromosomes”.</p>
<p>Coverage is produced using BigWig Files, passed to
<code><a href="../reference/plotHFGC.html">plotHFGC()</a></code> as either a <code>BigWigFileList</code> (for
separate tracks) or a list of <code>BigWigFileList</code> objects, for
multiple tracks with individual samples overlaid.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bwfl</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>  <span class="st">"data"</span>, <span class="st">"H3K27ac"</span>, <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{levels(samples$treatment)}_cov_chr10.bw"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFileList</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can also include cytogenetic bands for these coverage plots, some
of which are provided with <code>extraChIPs</code>.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"grch37.cytobands"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grch37.cytobands</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chrom chromStart chromEnd   name gieStain</span></span>
<span><span class="co">## 1  chr1          0  2300000 p36.33     gneg</span></span>
<span><span class="co">## 2  chr1    2300000  5400000 p36.32   gpos25</span></span>
<span><span class="co">## 3  chr1    5400000  7200000 p36.31     gneg</span></span>
<span><span class="co">## 4  chr1    7200000  9200000 p36.23   gpos25</span></span>
<span><span class="co">## 5  chr1    9200000 12700000 p36.22     gneg</span></span>
<span><span class="co">## 6  chr1   12700000 16200000 p36.21   gpos50</span></span></code></pre>
<p>Let’s check the coverage across the most highly-ranked region with
default settings.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">tmm_results</span>, <span class="va">hmp_fdr</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">bestOverlap</span>, <span class="st">'^Prom'</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<p>The most important option to consider when showing coverage is
whether to display tracks individually or as overlapping tracks. Passing
a BigWigFileList will plot each file on a separate track by default,
with y-axis limits able to be specified as consistent across all tracks.
Alternatively, by passing the coverage as a list of BigWigFileList
objects, each list element will be drawn as a single track with coverage
overlaid. For one or more overlaid tracks, colours will need to now be
provided in a list with matching structure.</p>
<p>The code for separate tracks would be: (not shown)</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span><span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, coverage <span class="op">=</span> <span class="va">bwfl</span><span class="op">)</span></span></code></pre></div>
<p>Overlaying coverage onto a single track can instead be highly
informative and enables the visualisation of multiple ChIP targets.
(Plot not shown in vignette)</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cov_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>H3K27ac <span class="op">=</span> <span class="va">bwfl</span><span class="op">)</span></span>
<span><span class="va">cov_colour</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>H3K27ac <span class="op">=</span> <span class="va">treat_colours</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">samples</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span></span>
<span>  <span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, </span>
<span>  coverage <span class="op">=</span>  <span class="va">cov_list</span>, linecol <span class="op">=</span> <span class="va">cov_colour</span>, zoom <span class="op">=</span> <span class="fl">50</span>, rotation.title <span class="op">=</span> <span class="fl">90</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>By default, the specified range will also be outlined
(i.e. highlighted) with a blue rectangle which is able to be turned off
and customised as the user sees fit.</p>
<div class="section level4">
<h4 id="adding-annotations-to-coverage">Adding Annotations To Coverage<a class="anchor" aria-label="anchor" href="#adding-annotations-to-coverage"></a>
</h4>
<p>An indication of which regions are associated with increased or
decreased ChIP signal can also be a useful annotation to add to plots
such as the above. Using the status from our differential signal
testing, we can create a <code>GRangesList</code> with the various
regions annotated as Increased, Unchanged etc,</p>
<p>Similar to the features track, where the names of
<code>GRangesList</code> elements denote the different feature types,
able to then assigned a colour, coverage annotation tracks follow these
same rules. For each coverage track being annotated, a
<code>GRangesList</code> object can denote the ranges which can be
assigned different colours. Annotation tracks should be passed to
<code><a href="../reference/plotHFGC.html">plotHFGC()</a></code> using the same structure as the coverage tracks,
so here if we pass the annotations as a list with the single element
<code>H3K27ac</code> these annotations will be added directly above the
coverage track for H3K27ac signal.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cov_annot</span> <span class="op">&lt;-</span> <span class="fu">splitAsList</span><span class="op">(</span><span class="va">tmm_results</span>, <span class="va">tmm_results</span><span class="op">$</span><span class="va">status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">endoapply</span><span class="op">(</span><span class="va">granges</span><span class="op">)</span> </span>
<span><span class="va">cov_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>H3K27ac <span class="op">=</span> <span class="va">cov_annot</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="displaying-genes-and-features">Displaying Genes and Features<a class="anchor" aria-label="anchor" href="#displaying-genes-and-features"></a>
</h4>
<p>Next we might like to add gene models to provide the regulatory
context. We’ll use the exons element of our <code>gencode</code> object
to create the format used by the defaults of the
<code>GeneRegionTrack()</code> function, with all exons and transcripts
annotated.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_models</span> <span class="op">&lt;-</span> <span class="va">gencode</span><span class="op">$</span><span class="va">exon</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">type</span>, gene <span class="op">=</span> <span class="va">gene_id</span>, exon <span class="op">=</span> <span class="va">exon_id</span>, transcript <span class="op">=</span> <span class="va">transcript_id</span>, </span>
<span>    symbol <span class="op">=</span> <span class="va">gene_name</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Another useful track to add might be some key features such as
promoters and other annotated regions, as we have formed earlier in the
workflow. Features must <strong>always</strong> be a
<code>GRangesList</code>, with each element defining a different type of
feature, as we already have in our <code>regions</code> object. For
multiple feature tracks, a list of <code>GRangesList</code> objects can
be passed.</p>
<p>In our final plot, we’ll zoom out to the entire region covered by
<em>KAT6B</em>, showing all transcripts, with out original region
highlighted in blue. All merged windows for which a differential signal
test was performed will be annotated with colours indicating the result
for that window. As can be seen multiple H3K27ac signal regions appear
to be responding to DHT treatment for <em>KAT6B</em> with a possible
intronic enhancer and alternative promoter being most likely impacted,
including the neighbouring region to our initial range. In contrast, the
primary promoter appears unaffected by DHT treatment.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">region_colours</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHFGC.html">plotHFGC</a></span><span class="op">(</span></span>
<span>  <span class="va">gr</span>, cytobands <span class="op">=</span> <span class="va">grch37.cytobands</span>, </span>
<span>  features <span class="op">=</span> <span class="va">regions</span>, featcol <span class="op">=</span> <span class="va">region_colours</span>, featstack <span class="op">=</span> <span class="st">"dense"</span>,</span>
<span>  coverage <span class="op">=</span>  <span class="va">cov_list</span>, linecol <span class="op">=</span> <span class="va">cov_colour</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">gene_models</span>, genecol <span class="op">=</span> <span class="st">"wheat"</span>, collapseTranscripts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  zoom <span class="op">=</span> <span class="fl">350</span>, shift <span class="op">=</span> <span class="op">-</span><span class="fl">1e5</span>,</span>
<span>  rotation.title <span class="op">=</span> <span class="fl">90</span>, covsize <span class="op">=</span> <span class="fl">10</span>, genesize <span class="op">=</span> <span class="fl">1</span>, featsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">cov_annot</span>, annotcol <span class="op">=</span> <span class="va">status_colours</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Unchanged"</span>, <span class="st">"Increased"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="differential_signal_sliding_files/figure-gfm/plot-hfgc-with-annot-1.png" alt="The addition of an annotation track for the coverage tracks shows which regions were retained during the analysis, as well as those which were considered as showing changed or unchanged signal" width="384"><p class="caption">
The addition of an annotation track for the coverage tracks shows which
regions were retained during the analysis, as well as those which were
considered as showing changed or unchanged signal
</p>
</div>
<p>Plots are able to be tweaked considerably further via multiple
parameters, however these basic approaches cover the core functionality
of <code><a href="../reference/plotHFGC.html">plotHFGC()</a></code> for enabling simple &amp; reproducible
plotting across regions for multiple sites within a larger
experiment.</p>
<p>If long-range interactions are available, these can also be provided
as a GenomicInteractions object, completing all available options for
the HFGC components.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] splines   stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggrepel_0.9.3               cqn_1.46.0                 </span></span>
<span><span class="co">##  [3] quantreg_5.95               SparseM_1.81               </span></span>
<span><span class="co">##  [5] preprocessCore_1.62.1       nor1mix_1.3-0              </span></span>
<span><span class="co">##  [7] mclust_6.0.0                quantro_1.34.0             </span></span>
<span><span class="co">##  [9] magrittr_2.0.3              here_1.0.1                 </span></span>
<span><span class="co">## [11] glue_1.6.2                  scales_1.2.1               </span></span>
<span><span class="co">## [13] plyranges_1.20.0            extraChIPs_1.5.6           </span></span>
<span><span class="co">## [15] ggside_0.2.2                patchwork_1.1.2            </span></span>
<span><span class="co">## [17] edgeR_3.42.4                limma_3.56.2               </span></span>
<span><span class="co">## [19] rtracklayer_1.60.0          BiocParallel_1.34.2        </span></span>
<span><span class="co">## [21] csaw_1.34.0                 SummarizedExperiment_1.30.2</span></span>
<span><span class="co">## [23] Biobase_2.60.0              MatrixGenerics_1.12.2      </span></span>
<span><span class="co">## [25] matrixStats_1.0.0           Rsamtools_2.16.0           </span></span>
<span><span class="co">## [27] Biostrings_2.68.1           XVector_0.40.0             </span></span>
<span><span class="co">## [29] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        </span></span>
<span><span class="co">## [31] IRanges_2.34.1              S4Vectors_0.38.1           </span></span>
<span><span class="co">## [33] BiocGenerics_0.46.0         lubridate_1.9.2            </span></span>
<span><span class="co">## [35] forcats_1.0.0               stringr_1.5.0              </span></span>
<span><span class="co">## [37] dplyr_1.1.2                 purrr_1.0.1                </span></span>
<span><span class="co">## [39] readr_2.1.4                 tidyr_1.3.0                </span></span>
<span><span class="co">## [41] tibble_3.2.1                ggplot2_3.4.2              </span></span>
<span><span class="co">## [43] tidyverse_2.0.0             BiocStyle_2.28.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.2                   ProtGenerics_1.32.0       </span></span>
<span><span class="co">##   [3] bitops_1.0-7               httr_1.4.6                </span></span>
<span><span class="co">##   [5] RColorBrewer_1.1-3         doParallel_1.0.17         </span></span>
<span><span class="co">##   [7] InteractionSet_1.28.1      doRNG_1.8.6               </span></span>
<span><span class="co">##   [9] tools_4.3.0                backports_1.4.1           </span></span>
<span><span class="co">##  [11] utf8_1.2.3                 R6_2.5.1                  </span></span>
<span><span class="co">##  [13] HDF5Array_1.28.1           lazyeval_0.2.2            </span></span>
<span><span class="co">##  [15] Gviz_1.44.0                rhdf5filters_1.12.1       </span></span>
<span><span class="co">##  [17] GetoptLong_1.0.5           withr_2.5.0               </span></span>
<span><span class="co">##  [19] prettyunits_1.1.1          gridExtra_2.3             </span></span>
<span><span class="co">##  [21] base64_2.0.1               VennDiagram_1.7.3         </span></span>
<span><span class="co">##  [23] cli_3.6.1                  textshaping_0.3.6         </span></span>
<span><span class="co">##  [25] formatR_1.14               sass_0.4.6                </span></span>
<span><span class="co">##  [27] genefilter_1.82.1          askpass_1.1               </span></span>
<span><span class="co">##  [29] pkgdown_2.0.7.9000         systemfonts_1.0.4         </span></span>
<span><span class="co">##  [31] foreign_0.8-84             siggenes_1.74.0           </span></span>
<span><span class="co">##  [33] illuminaio_0.42.0          dichromat_2.0-0.1         </span></span>
<span><span class="co">##  [35] scrime_1.3.5               BSgenome_1.68.0           </span></span>
<span><span class="co">##  [37] rstudioapi_0.14            RSQLite_2.3.1             </span></span>
<span><span class="co">##  [39] generics_0.1.3             shape_1.4.6               </span></span>
<span><span class="co">##  [41] BiocIO_1.10.0              Matrix_1.5-4.1            </span></span>
<span><span class="co">##  [43] interp_1.1-4               futile.logger_1.4.3       </span></span>
<span><span class="co">##  [45] fansi_1.0.4                lifecycle_1.0.3           </span></span>
<span><span class="co">##  [47] yaml_2.3.7                 rhdf5_2.44.0              </span></span>
<span><span class="co">##  [49] BiocFileCache_2.8.0        grid_4.3.0                </span></span>
<span><span class="co">##  [51] blob_1.2.4                 crayon_1.5.2              </span></span>
<span><span class="co">##  [53] lattice_0.21-8             ComplexUpset_1.3.3        </span></span>
<span><span class="co">##  [55] GenomicFeatures_1.52.1     annotate_1.78.0           </span></span>
<span><span class="co">##  [57] KEGGREST_1.40.0            pillar_1.9.0              </span></span>
<span><span class="co">##  [59] knitr_1.43                 ComplexHeatmap_2.16.0     </span></span>
<span><span class="co">##  [61] beanplot_1.3.1             metapod_1.8.0             </span></span>
<span><span class="co">##  [63] rjson_0.2.21               codetools_0.2-19          </span></span>
<span><span class="co">##  [65] data.table_1.14.8          vctrs_0.6.3               </span></span>
<span><span class="co">##  [67] png_0.1-8                  gtable_0.3.3              </span></span>
<span><span class="co">##  [69] cachem_1.0.8               xfun_0.39                 </span></span>
<span><span class="co">##  [71] S4Arrays_1.0.4             survival_3.5-5            </span></span>
<span><span class="co">##  [73] iterators_1.0.14           GenomicInteractions_1.34.0</span></span>
<span><span class="co">##  [75] nlme_3.1-162               bit64_4.0.5               </span></span>
<span><span class="co">##  [77] progress_1.2.2             filelock_1.0.2            </span></span>
<span><span class="co">##  [79] rprojroot_2.0.3            bslib_0.5.0               </span></span>
<span><span class="co">##  [81] rpart_4.1.19               colorspace_2.1-0          </span></span>
<span><span class="co">##  [83] DBI_1.1.3                  Hmisc_5.1-0               </span></span>
<span><span class="co">##  [85] nnet_7.3-19                tidyselect_1.2.0          </span></span>
<span><span class="co">##  [87] bit_4.0.5                  compiler_4.3.0            </span></span>
<span><span class="co">##  [89] curl_5.0.1                 htmlTable_2.4.1           </span></span>
<span><span class="co">##  [91] xml2_1.3.4                 desc_1.4.2                </span></span>
<span><span class="co">##  [93] DelayedArray_0.26.5        bookdown_0.34             </span></span>
<span><span class="co">##  [95] checkmate_2.2.0            quadprog_1.5-8            </span></span>
<span><span class="co">##  [97] rappdirs_0.3.3             digest_0.6.32             </span></span>
<span><span class="co">##  [99] rmarkdown_2.22             GEOquery_2.68.0           </span></span>
<span><span class="co">## [101] htmltools_0.5.5            pkgconfig_2.0.3           </span></span>
<span><span class="co">## [103] jpeg_0.1-10                base64enc_0.1-3           </span></span>
<span><span class="co">## [105] sparseMatrixStats_1.12.1   highr_0.10                </span></span>
<span><span class="co">## [107] dbplyr_2.3.2               fastmap_1.1.1             </span></span>
<span><span class="co">## [109] ensembldb_2.24.0           rlang_1.1.1               </span></span>
<span><span class="co">## [111] GlobalOptions_0.1.2        htmlwidgets_1.6.2         </span></span>
<span><span class="co">## [113] DelayedMatrixStats_1.22.1  EnrichedHeatmap_1.30.0    </span></span>
<span><span class="co">## [115] farver_2.1.1               jquerylib_0.1.4           </span></span>
<span><span class="co">## [117] jsonlite_1.8.7             VariantAnnotation_1.46.0  </span></span>
<span><span class="co">## [119] RCurl_1.98-1.12            Formula_1.2-5             </span></span>
<span><span class="co">## [121] GenomeInfoDbData_1.2.10    Rhdf5lib_1.22.0           </span></span>
<span><span class="co">## [123] munsell_0.5.0              Rcpp_1.0.10               </span></span>
<span><span class="co">## [125] stringi_1.7.12             zlibbioc_1.46.0           </span></span>
<span><span class="co">## [127] MASS_7.3-60                bumphunter_1.42.0         </span></span>
<span><span class="co">## [129] plyr_1.8.8                 minfi_1.46.0              </span></span>
<span><span class="co">## [131] parallel_4.3.0             deldir_1.0-9              </span></span>
<span><span class="co">## [133] multtest_2.56.0            hms_1.1.3                 </span></span>
<span><span class="co">## [135] circlize_0.4.15            locfit_1.5-9.8            </span></span>
<span><span class="co">## [137] igraph_1.5.0               rngtools_1.5.2            </span></span>
<span><span class="co">## [139] biomaRt_2.56.1             futile.options_1.0.1      </span></span>
<span><span class="co">## [141] XML_3.99-0.14              evaluate_0.21             </span></span>
<span><span class="co">## [143] latticeExtra_0.6-30        biovizBase_1.48.0         </span></span>
<span><span class="co">## [145] lambda.r_1.2.4             BiocManager_1.30.21       </span></span>
<span><span class="co">## [147] tzdb_0.4.0                 foreach_1.5.2             </span></span>
<span><span class="co">## [149] tweenr_2.0.2               MatrixModels_0.5-1        </span></span>
<span><span class="co">## [151] openssl_2.0.6              polyclip_1.10-4           </span></span>
<span><span class="co">## [153] reshape_0.8.9              clue_0.3-64               </span></span>
<span><span class="co">## [155] ggforce_0.4.1              xtable_1.8-4              </span></span>
<span><span class="co">## [157] broom_1.0.5                restfulr_0.0.15           </span></span>
<span><span class="co">## [159] AnnotationFilter_1.24.0    ragg_1.2.5                </span></span>
<span><span class="co">## [161] memoise_2.0.1              AnnotationDbi_1.62.1      </span></span>
<span><span class="co">## [163] GenomicAlignments_1.36.0   cluster_2.1.4             </span></span>
<span><span class="co">## [165] timechange_0.2.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-edgeR2016" class="csl-entry">
Chen, Yunshun, Aaron A T Lun, and Gordon K Smyth. 2016. “From Reads to
Genes to Pathways: Differential Expression Analysis of RNA-Seq
Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”
<em>F1000Research</em> 5: 1438. <a href="https://doi.org/10.12688/f1000research.8987.2" class="external-link uri">https://doi.org/10.12688/f1000research.8987.2</a>.
</div>
<div id="ref-Gandolfo2018-oc" class="csl-entry">
Gandolfo, Luke C, and Terence P Speed. 2018. “RLE Plots: Visualizing
Unwanted Variation in High Dimensional Data.” <em>PLoS One</em> 13 (2):
e0191629.
</div>
<div id="ref-gviz" class="csl-entry">
Hahne, Florian, and Robert Ivanek. 2016. “Statistical Genomics: Methods
and Protocols.” In, edited by Ewy Mathé and Sean Davis, 335–51. New
York, NY: Springer New York. <a href="https://doi.org/10.1007/978-1-4939-3578-9_16" class="external-link uri">https://doi.org/10.1007/978-1-4939-3578-9_16</a>.
</div>
<div id="ref-Hicks2015-ee" class="csl-entry">
Hicks, Stephanie C, and Rafael A Irizarry. 2015. “Quantro: A Data-Driven
Approach to Guide the Choice of an Appropriate Normalization Method.”
<em>Genome Biol.</em> 16 (1): 117.
</div>
<div id="ref-Law2014-xq" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. “Voom:
Precision Weights Unlock Linear Model Analysis Tools for <span class="nocase">RNA-seq</span> Read Counts.” <em>Genome Biol.</em> 15
(2): R29.
</div>
<div id="ref-DESeq22014" class="csl-entry">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. “Moderated
Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”
<em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link uri">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-csaw2016" class="csl-entry">
Lun, Aaron T L, and Gordon K Smyth. 2016. “Csaw: A Bioconductor Package
for Differential Binding Analysis of ChIP-Seq Data Using Sliding
Windows.” <em>Nucleic Acids Res.</em> 44 (5): e45.
</div>
<div id="ref-Lund2012-xo" class="csl-entry">
Lund, Steven P, Dan Nettleton, Davis J McCarthy, and Gordon K Smyth.
2012. “Detecting Differential Expression in <span class="nocase">RNA-sequence</span> Data Using Quasi-Likelihood with
Shrunken Dispersion Estimates.” <em>Stat. Appl. Genet. Mol. Biol.</em>
11 (5).
</div>
<div id="ref-McCarthy2009-qf" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. “Testing Significance
Relative to a Fold-Change Threshold Is a TREAT.” <em>Bioinformatics</em>
25 (6): 765–71.
</div>
<div id="ref-Molder2021-mo" class="csl-entry">
Mölder, Felix, Kim Philipp Jablonski, Brice Letcher, Michael B Hall,
Christopher H Tomkins-Tinch, Vanessa Sochat, Jan Forster, et al. 2021.
“Sustainable Data Analysis with Snakemake.” <em>F1000Res.</em> 10
(January): 33.
</div>
<div id="ref-Robinson2010-qp" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization
Method for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.” <em>Genome Biol.</em> 11 (3): R25.
</div>
<div id="ref-DiffBind2012" class="csl-entry">
Ross-Innes, Caryn S., Rory Stark, Andrew E. Teschendorff, Kelly A.
Holmes, H. Raza Ali, Mark J. Dunning, Gordon D. Brown, et al. 2012.
“Differential Oestrogen Receptor Binding Is Associated with Clinical
Outcome in Breast Cancer.” <em>Nature</em> 481: –4. <a href="http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html" class="external-link uri">http://www.nature.com/nature/journal/v481/n7381/full/nature10730.html</a>.
</div>
<div id="ref-Wilson2019-ln" class="csl-entry">
Wilson, Daniel J. 2019. “The Harmonic Mean <em>p</em>-Value for
Combining Dependent Tests.” <em>Proc. Natl. Acad. Sci. U. S. A.</em> 116
(4): 1195–1200.
</div>
</div>
<p><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephen Pederson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
