<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>extraChIPs: Range-Based operations • extraChIPs</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="extraChIPs: Range-Based operations">
<meta property="og:description" content="extraChIPs">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">extraChIPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.5.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/range_based_functions.html">Range-based Operations</a>
    </li>
    <li>
      <a href="../articles/differential_signal_sliding.html">Differential Signal Using Sliding Windows</a>
    </li>
    <li>
      <a href="../articles/differential_signal_fixed.html">Differential Signal Using Fixed-Width Windows</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/smped/extraChIPs" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>extraChIPs: Range-Based operations</h1>
                        <h4 data-toc-skip class="author">Stevie
Pederson</h4>
            <address class="author_afil">
      Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide,
AustraliaDame Roma Mitchell Cancer Researc Laboratories, University of
AdelaideJohn Curtin School of Medical Research, Australian National
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:stephen.pederson.au@gmail.com" class="email">stephen.pederson.au@gmail.com</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/smped/extraChIPs/blob/HEAD/vignettes/range_based_functions.Rmd" class="external-link"><code>vignettes/range_based_functions.Rmd</code></a></small>
      <div class="hidden name"><code>range_based_functions.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This package is designed to enable the Gene Regulatory Analysis using
Variable IP (<a href="https://github.com/smped/GRAVI" class="external-link">GRAVI</a>)
workflow, as a method for detecting differential binding in ChIP-Seq
datasets. As a workflow focussed on data integration, most functions
provided by the package <code>extraChIPs</code> are designed to enable
comparison across datasets. This vignette looks primarily at functions
which work with <code>GenomicRanges</code> objects.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>In order to use the package <code>extraChIPs</code> and follow this
vignette, we recommend using the package <code>BiocManager</code> hosted
on CRAN. Once this is installed, the additional packages required for
this vignette (<code>tidyverse</code>, <code>plyranges</code> and
<code>Gviz</code>) can also be installed.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tidyverse"</span>, <span class="st">"plyranges"</span>, <span class="st">"Gviz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"extraChIPs"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="coercion">Coercion<a class="anchor" aria-label="anchor" href="#coercion"></a>
</h2>
<p>The advent of the <code>tidyverse</code> has led to
<code>tibble</code> objects becoming a common alternative to
<code>data.frame</code> or <code>DataFrame</code> objects. Simple
functions within <code>extraChIP</code> enable coercion from
<code>GRanges</code>, <code>GInteractions</code> and
<code>DataFrame</code> objects to tibble objects, with list columns
correctly handled. By default these coercion functions will coerce
<code>GRanges</code> elements to a character vector. Similarly,
<code>GRanges</code> represented as a character column can be coerced to
the ranges element of a <code>GRanges</code> object.</p>
<p>First let’s coerce from a <code>tibble</code> (or S3
<code>data.frame</code>) to a <code>GRanges</code></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smped/extraChIPs" class="external-link">extraChIPs</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">73</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1-10:+"</span>, <span class="st">"chr1:5-10:+"</span>, <span class="st">"chr1:5-6:+"</span><span class="op">)</span>,</span>
<span>  gene_id <span class="op">=</span> <span class="st">"gene1"</span>,</span>
<span>  tx_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"transcript"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">df</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span></span>
<span><span class="co">##   range       gene_id tx_id        score</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1:1-10:+ gene1   transcript1 0.442 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr1:5-10:+ gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr1:5-6:+  gene1   transcript3 0.615</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colToRanges-methods.html">colToRanges</a></span><span class="op">(</span><span class="va">df</span>, <span class="st">"range"</span><span class="op">)</span></span>
<span><span class="va">gr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id       tx_id     score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1      1-10      + |       gene1 transcript1 0.4423369</span></span>
<span><span class="co">##   [2]     chr1      5-10      + |       gene1 transcript2 0.0831099</span></span>
<span><span class="co">##   [3]     chr1       5-6      + |       gene1 transcript3 0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Coercion back to a <code>tibble</code> will place the ranges as a
character column by default. However, this can be turned off and the
conventional coercion from <code>as.data.frame</code> will instead be
applied, internally wrapped in <code><a href="../reference/as_tibble.html">as_tibble()</a></code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span></span>
<span><span class="co">##   range       gene_id tx_id        score</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1:1-10:+ gene1   transcript1 0.442 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr1:5-10:+ gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr1:5-6:+  gene1   transcript3 0.615</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">gr</span>, rangeAsChar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 8</span></span></span>
<span><span class="co">##   seqnames start   end width strand gene_id tx_id        score</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1         1    10    10 +      gene1   transcript1 0.442 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr1         5    10     6 +      gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr1         5     6     2 +      gene1   transcript3 0.615</span></span></code></pre>
<p>A simple feature which may be useful for printing gene names using
<code>rmarkdown</code> is contained in <code><a href="../reference/collapseGenes.html">collapseGenes()</a></code>.
Here a character vector of gene names is collapsed into a
<code>glue</code> object of length `, with gene names rendered in
italics by default.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/collapseGenes.html">collapseGenes</a></span><span class="op">(</span><span class="va">gn</span><span class="op">)</span></span></code></pre></div>
<p><em>Gene1</em>, <em>Gene2</em> and <em>Gene3</em></p>
</div>
<div class="section level2">
<h2 id="formation-of-consensus-peaks">Formation of Consensus Peaks<a class="anchor" aria-label="anchor" href="#formation-of-consensus-peaks"></a>
</h2>
<p>The formation of consensus peaks incorporating ranges from multiple
replicates is a key part of many ChIP-Seq analyses. A common format
returned by tools such as <code>mcas2 callpeak</code> is the
<code>narrowPeak</code> (or <code>broadPeak</code>) format. Sets of
these across multiple replicates can imported using the function
<code><a href="../reference/importPeaks.html">importPeaks()</a></code>, which returns a
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList()</a></code></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"extdata/ER_1.narrowPeak"</span>, <span class="st">"extdata/ER_2.narrowPeak"</span><span class="op">)</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"extraChIPs"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/importPeaks.html">importPeaks</a></span><span class="op">(</span><span class="va">fl</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ER_1"</span>, <span class="st">"ER_2"</span><span class="op">)</span></span></code></pre></div>
<p>In the above we have loaded the peaks from two replicates from the
Estrogen Receptor (ER) in T-47D cells. To form consensus peaks we can
place a restriction on the number of replicates an overlapping peak
needs to appear in. By default, the function
<code><a href="../reference/makeConsensus.html">makeConsensus()</a></code> sets the proportion to be zero
<code>(p = 0)</code> so all peaks are retained. Note that a logical
vector/column is returned for each replicate, along with the number of
replicates the consensus peak is derived from.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 14 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames          ranges strand |      ER_1      ER_2         n</span></span>
<span><span class="co">##           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;logical&gt; &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##    [1]     chr1   856458-856640      * |      TRUE     FALSE         1</span></span>
<span><span class="co">##    [2]     chr1   868541-868839      * |     FALSE      TRUE         1</span></span>
<span><span class="co">##    [3]     chr1 1008550-1010075      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##    [4]     chr1 1014770-1016015      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##    [5]     chr1 1051307-1051918      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##    ...      ...             ...    ... .       ...       ...       ...</span></span>
<span><span class="co">##   [10]     chr1 1608098-1608346      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [11]     chr1 1690460-1690641      * |     FALSE      TRUE         1</span></span>
<span><span class="co">##   [12]     chr1 1790733-1790975      * |      TRUE     FALSE         1</span></span>
<span><span class="co">##   [13]     chr1 1878927-1879257      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [14]     chr1 1900588-1900902      * |      TRUE     FALSE         1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>However, we may wish for peaks to appear in all replicates, so we can
set the argument <code>p = 1</code> (or any other value <span class="math inline">\(0 \leq p \leq 1\)</span>).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span><span class="va">peaks</span>, p <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames          ranges strand |      ER_1      ER_2         n</span></span>
<span><span class="co">##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;logical&gt; &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1 1008550-1010075      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [2]     chr1 1014770-1016015      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [3]     chr1 1051307-1051918      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [4]     chr1 1368372-1369009      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [5]     chr1 1608098-1608346      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   [6]     chr1 1878927-1879257      * |      TRUE      TRUE         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>In addition, we may wish to keep the values from the
<code>mcols</code> as part of our consensus peaks, such as the
<code>qValue</code>. This can be specified using the <code>var</code>
argument, and will return a <code>CompressedList</code> as returned by
<code><a href="../reference/reduceMC.html">reduceMC()</a></code>, seen below.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span><span class="va">peaks</span>, p <span class="op">=</span> <span class="fl">1</span>, var <span class="op">=</span> <span class="st">"qValue"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames          ranges strand |          qValue      ER_1      ER_2</span></span>
<span><span class="co">##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;NumericList&gt; &lt;logical&gt; &lt;logical&gt;</span></span>
<span><span class="co">##   [1]     chr1 1008550-1010075      * | 635.748,605.040      TRUE      TRUE</span></span>
<span><span class="co">##   [2]     chr1 1014770-1016015      * | 95.1223,78.3953      TRUE      TRUE</span></span>
<span><span class="co">##   [3]     chr1 1051307-1051918      * | 65.5538,95.9677      TRUE      TRUE</span></span>
<span><span class="co">##   [4]     chr1 1368372-1369009      * | 22.5483,29.1231      TRUE      TRUE</span></span>
<span><span class="co">##   [5]     chr1 1608098-1608346      * | 72.3235,64.8439      TRUE      TRUE</span></span>
<span><span class="co">##   [6]     chr1 1878927-1879257      * | 33.8900,14.4909      TRUE      TRUE</span></span>
<span><span class="co">##               n</span></span>
<span><span class="co">##       &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]         2</span></span>
<span><span class="co">##   [2]         2</span></span>
<span><span class="co">##   [3]         2</span></span>
<span><span class="co">##   [4]         2</span></span>
<span><span class="co">##   [5]         2</span></span>
<span><span class="co">##   [6]         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>We could even identify the peak centre and pass these to the set of
consensus peaks for downstream peak re-centreing.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">endoapply</span><span class="op">(</span><span class="va">mutate</span>, centre <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">peak</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">1</span>, var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames          ranges strand |          centre      ER_1      ER_2</span></span>
<span><span class="co">##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;NumericList&gt; &lt;logical&gt; &lt;logical&gt;</span></span>
<span><span class="co">##   [1]     chr1 1008550-1010075      * | 1009212,1009212      TRUE      TRUE</span></span>
<span><span class="co">##   [2]     chr1 1014770-1016015      * | 1014981,1014949      TRUE      TRUE</span></span>
<span><span class="co">##   [3]     chr1 1051307-1051918      * | 1051565,1051634      TRUE      TRUE</span></span>
<span><span class="co">##   [4]     chr1 1368372-1369009      * | 1368613,1368767      TRUE      TRUE</span></span>
<span><span class="co">##   [5]     chr1 1608098-1608346      * | 1608247,1608247      TRUE      TRUE</span></span>
<span><span class="co">##   [6]     chr1 1878927-1879257      * | 1879087,1879094      TRUE      TRUE</span></span>
<span><span class="co">##               n</span></span>
<span><span class="co">##       &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]         2</span></span>
<span><span class="co">##   [2]         2</span></span>
<span><span class="co">##   [3]         2</span></span>
<span><span class="co">##   [4]         2</span></span>
<span><span class="co">##   [5]         2</span></span>
<span><span class="co">##   [6]         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>We could then find the mean of the peak centres for an averaged
centre.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peaks</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">endoapply</span><span class="op">(</span><span class="va">mutate</span>, centre <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">peak</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="../reference/makeConsensus.html">makeConsensus</a></span><span class="op">(</span>p <span class="op">=</span> <span class="fl">1</span>, var <span class="op">=</span> <span class="st">"centre"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>centre <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">centre</span>, <span class="va">mean</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames          ranges strand |    centre      ER_1      ER_2         n</span></span>
<span><span class="co">##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;logical&gt; &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1 1008550-1010075      * |   1009212      TRUE      TRUE         2</span></span>
<span><span class="co">##   [2]     chr1 1014770-1016015      * |   1014965      TRUE      TRUE         2</span></span>
<span><span class="co">##   [3]     chr1 1051307-1051918      * |   1051600      TRUE      TRUE         2</span></span>
<span><span class="co">##   [4]     chr1 1368372-1369009      * |   1368690      TRUE      TRUE         2</span></span>
<span><span class="co">##   [5]     chr1 1608098-1608346      * |   1608247      TRUE      TRUE         2</span></span>
<span><span class="co">##   [6]     chr1 1878927-1879257      * |   1879090      TRUE      TRUE         2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="simple-operations-retaining-mcols">Simple Operations Retaining <code>mcols()</code><a class="anchor" aria-label="anchor" href="#simple-operations-retaining-mcols"></a>
</h2>
<p>The standard set operations implemented in the package
<code>GenomicRanges</code> will always drop the <code>mcols</code>
element by default. The <code>extraChIPs</code> functions
<code><a href="../reference/reduceMC.html">reduceMC()</a></code>, <code><a href="../reference/setoptsMC-methods.html">setdiffMC()</a></code>,
<code><a href="../reference/setoptsMC-methods.html">intersectMC()</a></code> and <code><a href="../reference/setoptsMC-methods.html">unionMC()</a></code> all produce the
same output as their similarly-named functions, however, the
<code>mcols()</code> elements in the query object are also returned.
Where required, columns are coerced into <code>CompressedList</code>
columns. This can particularly useful when needed to propagate the
information contained in the initial ranges through to subsequent
analytic steps</p>
<div class="section level3">
<h3 id="simplifying-single-granges-objects">Simplifying single <code>GRanges</code> objects<a class="anchor" aria-label="anchor" href="#simplifying-single-granges-objects"></a>
</h3>
<p>The classical approach to defining TSS regions for a set of
transcripts would be to use the function <code>resize</code>()`, setting
the width as 1.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span><span class="va">gr</span>, width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tss</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id       tx_id     score</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      + |       gene1 transcript1 0.4423369</span></span>
<span><span class="co">##   [2]     chr1         5      + |       gene1 transcript2 0.0831099</span></span>
<span><span class="co">##   [3]     chr1         5      + |       gene1 transcript3 0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>As we can see, two transcripts start at position 5, so we may choose
to reduce this, which would lose the <code>mcols</code> element. The
alternative <code><a href="../reference/reduceMC.html">reduceMC()</a></code> will retain all
<code>mcols</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      +</span></span>
<span><span class="co">##   [2]     chr1         5      +</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reduceMC.html">reduceMC</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id                   tx_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      + |       gene1             transcript1</span></span>
<span><span class="co">##   [2]     chr1         5      + |       gene1 transcript2,transcript3</span></span>
<span><span class="co">##                     score</span></span>
<span><span class="co">##             &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1]            0.442337</span></span>
<span><span class="co">##   [2] 0.0831099,0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>By default, this function will attempt to coerce <code>mcols</code>
to a new <code>mcol</code> of the appropriate type, however, when
multiple values are inevitable such as for the <code>tx_id</code> column
above, these will be coerced to a <code>CompressedList</code>. The
simplification of the multiple values seen in the <code>gene_id</code>
can also be turned off if desired should repeated values be important
for downstream analysis.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reduceMC.html">reduceMC</a></span><span class="op">(</span><span class="va">tss</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |         gene_id                   tx_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;         &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      + |           gene1             transcript1</span></span>
<span><span class="co">##   [2]     chr1         5      + |     gene1,gene1 transcript2,transcript3</span></span>
<span><span class="co">##                     score</span></span>
<span><span class="co">##             &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1]            0.442337</span></span>
<span><span class="co">##   [2] 0.0831099,0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>This allows for simple integration with <code>tidyverse</code>
nesting strategies.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reduceMC.html">reduceMC</a></span><span class="op">(</span><span class="va">tss</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/unnest.html" class="external-link">unnest</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 4</span></span></span>
<span><span class="co">##   range    gene_id tx_id        score</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> chr1:1:+ gene1   transcript1 0.442 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> chr1:5:+ gene1   transcript2 0.083<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> chr1:5:+ gene1   transcript3 0.615</span></span></code></pre>
<p>Whilst <code>reduceMC</code> relies on the range-reduction as
implemented in <code><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">GenomicRanges::reduce()</a></code>, some alternative
approaches are included, such as <code><a href="../reference/chopMC.html">chopMC()</a></code>, which finds
identical ranges and nests the mcols element as
<code>CompressedList</code> objects.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/chopMC.html">chopMC</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id                   tx_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      + |       gene1             transcript1</span></span>
<span><span class="co">##   [2]     chr1         5      + |       gene1 transcript2,transcript3</span></span>
<span><span class="co">##                     score</span></span>
<span><span class="co">##             &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1]            0.442337</span></span>
<span><span class="co">##   [2] 0.0831099,0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>In the case of the object <code>tss</code>, this output is identical
to <code><a href="../reference/reduceMC.html">reduceMC()</a></code>, however, given that there are no identical
ranges in <code>gr</code> the two functions would behave very
differently on that object.</p>
<p>A final operation for simplifying <code>GRanges</code> objects would
be <code><a href="../reference/distinctMC.html">distinctMC()</a></code> which is a wrapper to
<code>dplyr]::distinct</code> incorporating both the range and
<code>mcols</code>. The columns to search can be called using
<code>&lt;data-masking&gt;</code> approaches as detailed in the
manual.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/distinctMC.html">distinctMC</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      +</span></span>
<span><span class="co">##   [2]     chr1         5      +</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/distinctMC.html">distinctMC</a></span><span class="op">(</span><span class="va">tss</span>, <span class="va">gene_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1         1      + |       gene1</span></span>
<span><span class="co">##   [2]     chr1         5      + |       gene1</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="set-operations-with-two-granges-objects">Set Operations with Two <code>GRanges</code> Objects<a class="anchor" aria-label="anchor" href="#set-operations-with-two-granges-objects"></a>
</h3>
<p>Whilst <code>reduce/reduceMC</code> is applied to a single
<code>GRanges</code> object, the set operation functions
<code>intersect</code>, <code>setdiff</code> and <code>union</code> are
valuable approaches for comparing ranges. Using the <code>*MC()</code>
functions will retain <code>mcols</code> elements from the <em>query
range</em>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1-5"</span>, <span class="st">"chr1:9-12:*"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">peaks</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"peak1"</span>, <span class="st">"peak2"</span><span class="op">)</span></span>
<span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1       1-5      *</span></span>
<span><span class="co">##   [2]     chr1      9-10      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setoptsMC-methods.html">intersectMC</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1       1-5      * |       gene1</span></span>
<span><span class="co">##   [2]     chr1      9-10      * |       gene1</span></span>
<span><span class="co">##                                     tx_id                         score</span></span>
<span><span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span></span>
<span><span class="co">##   [2]             transcript1,transcript2           0.4423369,0.0831099</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setoptsMC-methods.html">setdiffMC</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 1 range and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1       6-8      * |       gene1</span></span>
<span><span class="co">##                                     tx_id                         score</span></span>
<span><span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setoptsMC-methods.html">unionMC</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 1 range and 3 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     gene_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1      1-12      * |       gene1</span></span>
<span><span class="co">##                                     tx_id                         score</span></span>
<span><span class="co">##                           &lt;CharacterList&gt;                 &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1] transcript1,transcript2,transcript3 0.4423369,0.0831099,0.6146112</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>There is a performance overhead to preparation of mcols as
<code>CompressedList</code> objects and all <code>mcols</code> in the
query object will be returned. If only wishing to retain a subset of
<code>mcols</code>, these should be selected prior to passing to these
functions.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">tx_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/setoptsMC-methods.html">intersectMC</a></span><span class="op">(</span><span class="va">peaks</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames    ranges strand |                               tx_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;CharacterList&gt;</span></span>
<span><span class="co">##   [1]     chr1       1-5      * | transcript1,transcript2,transcript3</span></span>
<span><span class="co">##   [2]     chr1      9-10      * |             transcript1,transcript2</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="overlapping-proportions">Overlapping proportions<a class="anchor" aria-label="anchor" href="#overlapping-proportions"></a>
</h3>
<p>Whilst the functions <code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps()</a></code> and
<code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">overlapsAny()</a></code> are extremely useful, the addition of
<code><a href="../reference/propOverlap-methods.html">propOverlap()</a></code> returns a numeric vector with the proportion
of each query range (<code>x</code>) which overlaps any range in the
subject range (<code>y</code>)</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/propOverlap-methods.html">propOverlap</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.7 0.5 0.5</span></span></code></pre>
<p>This is also extended to enable comparison across multiple features
and to classify each peak by which features that it overlaps the most
strongly.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bestOverlap-methods.html">bestOverlap</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span>, var <span class="op">=</span> <span class="st">"peak_id"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "peak1" "peak2" "peak1"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="more-complex-operations">More Complex Operations<a class="anchor" aria-label="anchor" href="#more-complex-operations"></a>
</h2>
<div class="section level3">
<h3 id="range-partitioning">Range Partitioning<a class="anchor" aria-label="anchor" href="#range-partitioning"></a>
</h3>
<p>In addition to standard set operations, one set of ranges can be used
to partition another set of ranges, returning <code>mcols</code> from
both ranges. Ranges from the query range (<code>x</code>) are returned
after being partitioned by the ranges in the subject range
(<code>y</code>). Subject ranges used for partitioning <em>must be
non-overlapping</em>, and if overlapping ranges are provided, these will
be reduced prior to partitioning.</p>
<p>This enables the identification of specific ranges from the query
range (<code>x</code>) which overlap ranges from the subject range
(<code>y</code>) Under this approach, <code>mcols</code> from both
<code>query</code> and <code>subject</code> ranges will be returned to
enable the clear ranges which are common and distinct within the two
sets of ranges.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/partitionRanges-methods.html">partitionRanges</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 5 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     peak_id     gene_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1       1-5      + |       peak1       gene1</span></span>
<span><span class="co">##   [2]     chr1         5      + |       peak1       gene1</span></span>
<span><span class="co">##   [3]     chr1         6      + |        &lt;NA&gt;       gene1</span></span>
<span><span class="co">##   [4]     chr1       6-8      + |        &lt;NA&gt;       gene1</span></span>
<span><span class="co">##   [5]     chr1      9-10      + |       peak2       gene1</span></span>
<span><span class="co">##                         tx_id               score</span></span>
<span><span class="co">##               &lt;CharacterList&gt;       &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1]             transcript1            0.442337</span></span>
<span><span class="co">##   [2] transcript2,transcript3 0.0831099,0.6146112</span></span>
<span><span class="co">##   [3]             transcript3            0.614611</span></span>
<span><span class="co">##   [4] transcript1,transcript2 0.4423369,0.0831099</span></span>
<span><span class="co">##   [5] transcript1,transcript2 0.4423369,0.0831099</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Whilst this shares some similarity with <code><a href="../reference/setoptsMC-methods.html">intersectMC()</a></code>
the additional capabilities provide greater flexibility.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/partitionRanges-methods.html">partitionRanges</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">peaks</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">peak_id</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 4 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand |     peak_id     gene_id</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1         6      + |        &lt;NA&gt;       gene1</span></span>
<span><span class="co">##   [2]     chr1       6-8      + |        &lt;NA&gt;       gene1</span></span>
<span><span class="co">##                         tx_id               score</span></span>
<span><span class="co">##               &lt;CharacterList&gt;       &lt;NumericList&gt;</span></span>
<span><span class="co">##   [1]             transcript3            0.614611</span></span>
<span><span class="co">##   [2] transcript1,transcript2 0.4423369,0.0831099</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="stitching-ranges">Stitching Ranges<a class="anchor" aria-label="anchor" href="#stitching-ranges"></a>
</h3>
<p>Using the function <code><a href="../reference/stitchRanges.html">stitchRanges()</a></code> we are able to join
together sets of nearby ranges, but with the option of placing clear
barriers between ranges, across which ranges cannot be joined. This may
be useful if joining enhancers to form putative super-enhancers, but
explicitly omitting defined promoter regions.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr1:1-10"</span>, <span class="st">"chr1:101-110"</span>, <span class="st">"chr1:181-200"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">prom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1:150:+"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stitchRanges.html">stitchRanges</a></span><span class="op">(</span><span class="va">enh</span>, exclude <span class="op">=</span> <span class="va">prom</span>, maxgap <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 2 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames    ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1     1-110      *</span></span>
<span><span class="co">##   [2]     chr1   181-200      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>As a visualisation (below) ranges within 100bp were stitched
together, however the region defined as a ‘promoter’ acted as a barrier
and ranges were not stitched together across this barrier.</p>
<p><img src="stitched.png" width="694"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] plyranges_1.20.0            extraChIPs_1.5.7           </span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.30.2 Biobase_2.60.0             </span></span>
<span><span class="co">##  [5] MatrixGenerics_1.12.2       matrixStats_1.0.0          </span></span>
<span><span class="co">##  [7] ggside_0.2.2                GenomicRanges_1.52.0       </span></span>
<span><span class="co">##  [9] GenomeInfoDb_1.36.1         IRanges_2.34.1             </span></span>
<span><span class="co">## [11] S4Vectors_0.38.1            BiocGenerics_0.46.0        </span></span>
<span><span class="co">## [13] BiocParallel_1.34.2         lubridate_1.9.2            </span></span>
<span><span class="co">## [15] forcats_1.0.0               stringr_1.5.0              </span></span>
<span><span class="co">## [17] dplyr_1.1.2                 purrr_1.0.1                </span></span>
<span><span class="co">## [19] readr_2.1.4                 tidyr_1.3.0                </span></span>
<span><span class="co">## [21] tibble_3.2.1                ggplot2_3.4.2              </span></span>
<span><span class="co">## [23] tidyverse_2.0.0             BiocStyle_2.28.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] BiocIO_1.10.0              bitops_1.0-7              </span></span>
<span><span class="co">##   [3] filelock_1.0.2             polyclip_1.10-4           </span></span>
<span><span class="co">##   [5] XML_3.99-0.14              rpart_4.1.19              </span></span>
<span><span class="co">##   [7] lifecycle_1.0.3            edgeR_3.42.4              </span></span>
<span><span class="co">##   [9] doParallel_1.0.17          rprojroot_2.0.3           </span></span>
<span><span class="co">##  [11] MASS_7.3-60                lattice_0.21-8            </span></span>
<span><span class="co">##  [13] ensembldb_2.24.0           backports_1.4.1           </span></span>
<span><span class="co">##  [15] magrittr_2.0.3             limma_3.56.2              </span></span>
<span><span class="co">##  [17] Hmisc_5.1-0                sass_0.4.6                </span></span>
<span><span class="co">##  [19] rmarkdown_2.23             jquerylib_0.1.4           </span></span>
<span><span class="co">##  [21] yaml_2.3.7                 metapod_1.8.0             </span></span>
<span><span class="co">##  [23] Gviz_1.44.0                DBI_1.1.3                 </span></span>
<span><span class="co">##  [25] RColorBrewer_1.1-3         zlibbioc_1.46.0           </span></span>
<span><span class="co">##  [27] AnnotationFilter_1.24.0    biovizBase_1.48.0         </span></span>
<span><span class="co">##  [29] RCurl_1.98-1.12            nnet_7.3-19               </span></span>
<span><span class="co">##  [31] tweenr_2.0.2               VariantAnnotation_1.46.0  </span></span>
<span><span class="co">##  [33] rappdirs_0.3.3             circlize_0.4.15           </span></span>
<span><span class="co">##  [35] GenomeInfoDbData_1.2.10    ggrepel_0.9.3             </span></span>
<span><span class="co">##  [37] pkgdown_2.0.7.9000         codetools_0.2-19          </span></span>
<span><span class="co">##  [39] DelayedArray_0.26.6        ggforce_0.4.1             </span></span>
<span><span class="co">##  [41] xml2_1.3.4                 tidyselect_1.2.0          </span></span>
<span><span class="co">##  [43] shape_1.4.6                futile.logger_1.4.3       </span></span>
<span><span class="co">##  [45] farver_2.1.1               ComplexUpset_1.3.3        </span></span>
<span><span class="co">##  [47] BiocFileCache_2.8.0        base64enc_0.1-3           </span></span>
<span><span class="co">##  [49] GenomicAlignments_1.36.0   jsonlite_1.8.7            </span></span>
<span><span class="co">##  [51] GetoptLong_1.0.5           Formula_1.2-5             </span></span>
<span><span class="co">##  [53] iterators_1.0.14           systemfonts_1.0.4         </span></span>
<span><span class="co">##  [55] foreach_1.5.2              tools_4.3.0               </span></span>
<span><span class="co">##  [57] progress_1.2.2             ragg_1.2.5                </span></span>
<span><span class="co">##  [59] Rcpp_1.0.10                glue_1.6.2                </span></span>
<span><span class="co">##  [61] gridExtra_2.3              xfun_0.39                 </span></span>
<span><span class="co">##  [63] withr_2.5.0                formatR_1.14              </span></span>
<span><span class="co">##  [65] BiocManager_1.30.21        fastmap_1.1.1             </span></span>
<span><span class="co">##  [67] latticeExtra_0.6-30        fansi_1.0.4               </span></span>
<span><span class="co">##  [69] digest_0.6.32              timechange_0.2.0          </span></span>
<span><span class="co">##  [71] R6_2.5.1                   textshaping_0.3.6         </span></span>
<span><span class="co">##  [73] colorspace_2.1-0           jpeg_0.1-10               </span></span>
<span><span class="co">##  [75] dichromat_2.0-0.1          biomaRt_2.56.1            </span></span>
<span><span class="co">##  [77] RSQLite_2.3.1              utf8_1.2.3                </span></span>
<span><span class="co">##  [79] generics_0.1.3             data.table_1.14.8         </span></span>
<span><span class="co">##  [81] rtracklayer_1.60.0         prettyunits_1.1.1         </span></span>
<span><span class="co">##  [83] InteractionSet_1.28.1      httr_1.4.6                </span></span>
<span><span class="co">##  [85] htmlwidgets_1.6.2          S4Arrays_1.0.4            </span></span>
<span><span class="co">##  [87] pkgconfig_2.0.3            gtable_0.3.3              </span></span>
<span><span class="co">##  [89] blob_1.2.4                 ComplexHeatmap_2.16.0     </span></span>
<span><span class="co">##  [91] XVector_0.40.0             htmltools_0.5.5           </span></span>
<span><span class="co">##  [93] bookdown_0.34              ProtGenerics_1.32.0       </span></span>
<span><span class="co">##  [95] clue_0.3-64                scales_1.2.1              </span></span>
<span><span class="co">##  [97] png_0.1-8                  knitr_1.43                </span></span>
<span><span class="co">##  [99] lambda.r_1.2.4             rstudioapi_0.14           </span></span>
<span><span class="co">## [101] tzdb_0.4.0                 rjson_0.2.21              </span></span>
<span><span class="co">## [103] checkmate_2.2.0            curl_5.0.1                </span></span>
<span><span class="co">## [105] cachem_1.0.8               GlobalOptions_0.1.2       </span></span>
<span><span class="co">## [107] parallel_4.3.0             foreign_0.8-84            </span></span>
<span><span class="co">## [109] AnnotationDbi_1.62.2       restfulr_0.0.15           </span></span>
<span><span class="co">## [111] desc_1.4.2                 pillar_1.9.0              </span></span>
<span><span class="co">## [113] grid_4.3.0                 vctrs_0.6.3               </span></span>
<span><span class="co">## [115] dbplyr_2.3.2               cluster_2.1.4             </span></span>
<span><span class="co">## [117] htmlTable_2.4.1            evaluate_0.21             </span></span>
<span><span class="co">## [119] VennDiagram_1.7.3          EnrichedHeatmap_1.30.0    </span></span>
<span><span class="co">## [121] GenomicFeatures_1.52.1     cli_3.6.1                 </span></span>
<span><span class="co">## [123] locfit_1.5-9.8             compiler_4.3.0            </span></span>
<span><span class="co">## [125] futile.options_1.0.1       Rsamtools_2.16.0          </span></span>
<span><span class="co">## [127] rlang_1.1.1                crayon_1.5.2              </span></span>
<span><span class="co">## [129] interp_1.1-4               fs_1.6.2                  </span></span>
<span><span class="co">## [131] stringi_1.7.12             deldir_1.0-9              </span></span>
<span><span class="co">## [133] munsell_0.5.0              Biostrings_2.68.1         </span></span>
<span><span class="co">## [135] lazyeval_0.2.2             csaw_1.34.0               </span></span>
<span><span class="co">## [137] Matrix_1.5-4.1             BSgenome_1.68.0           </span></span>
<span><span class="co">## [139] hms_1.1.3                  patchwork_1.1.2           </span></span>
<span><span class="co">## [141] bit64_4.0.5                KEGGREST_1.40.0           </span></span>
<span><span class="co">## [143] highr_0.10                 igraph_1.5.0              </span></span>
<span><span class="co">## [145] broom_1.0.5                memoise_2.0.1             </span></span>
<span><span class="co">## [147] bslib_0.5.0                bit_4.0.5                 </span></span>
<span><span class="co">## [149] GenomicInteractions_1.34.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephen Pederson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
